---
title: "Oral microbiome relationship to oral cancer"
author: "Tommy Tran"
date: "2023-03-05"
output:
  word_document: default
  html_document: default
bibliography: references.bib
csl: https://raw.githubusercontent.com/citation-style-language/styles/master/research-institute-for-nature-and-forest.csl
---

# Introduction

The prevalence of oral cavity or pharyngeal cancer was estimated to be 410,000 people living with it in the United States. Worldwide the prevalence could be even higher with the lack of importance on oral health. After being diagnosed with oral cancer only 68% of individuals will survive 5 or more years. The rate of new cases and death rate have remained relatively unchanged for the past 45 years. This shows the need for better fundamental and mechanistic understanding of oral cancer to formulate an effective remedy on prevention and treatment plans.

There are only a few large-scale case-control studies to help interpret the genomic microbiome and microbiota inhabiting the oral cavity. Studies have shown there are certain taxa that will increase or decrease in abundance. *Fusobacterium* seems to be a major candidate that thrives in the environment of oral cancer. Identification of metabolic functions shifting, revealed differences between diseased and healthy groups. This metabolic change generally favored anaerobic and less oxidative like oral environments. Microbial signatures associated with oral cancers have been predicted that may help future diagnoses and treatment.

# Oral microbiota composition is most heavily affected by total teeth loss

The oral microbiome in a larger of cohort cancer cases and controls displayed changes in community structure and metabolic pathways. Different factors such as; tooth loss, tumor HPV status, peridontal disease, tobacco usage, and alcohol consumption were assessed to predict whether they affect the oral microbiota and their relationship to oral cancer. The most significant dysbioses occurs when no natural teeth remained and was shown to be the highest risk factor to oral cancer. This caused a higher abundance of *Dialister* which is a Gram-negative bacteria and has been associated with worse peridontal status and endodontic infections [@Kumar2005; @Ribeiro2011; @Drago2013; @Börnigen2017].

In another study bacteria were profiled from patients within the surface of an oral squamous cell carcinoma (OSCC) lesion and on its adjacent healthy tissue. Taxonomic signatures revealed a profusion of *Fusobacterium nucleatum CTI-2* and a decrease in *Streptococcus pneumoniae* of buccal mucosa cancers[@Su2021; @Al-Hebshi2017].

# Microbiota metabolic functions change in the oral cancer cancers

Prediction of oral microbiome functionality revealed bacterial genes that were altered in association to oral cancer. Specifically, pathways involved with terpenoids and polyketides, lipopolysaccharides (LPS), and vitamins and cofactors [@Börnigen2017; @Su2021]. Terpenoids and polyketides function to produce tumor-suppression metabolites, however a decrease of this function was observed in oral cancer cohorts [@Su2021]. LPS transport decreased, which are usually used in the TCA cycle for oxidation energy. Thus, may gear the microbial environment towards anaerobic microbes metabolism during oral cancer [@Börnigen2017]. Microbiol dysbioses also lead to functional dysbioses during oral cancer as there was a predicted increase in synthesis and transport of vitamins and cofactors. These specific KEGG modules were M00123: Biotin biosynthesis, pimeloyl-ACP/CoA =\> biotin, M00127: Thiamine biosynthesis, AIR =\> thiamine-P/thiamine-2P). Also, an increased in iron uptake (M00190: Iron(III) transport, M00243: Manganese/iron transport), iron (M00243: Manganese/iron transport, M00248: ABC transporter), zinc, and manganese transport (M00244: Putative zinc/manganese transport), may explain the necessity of fewer cofacters in a lesser oxidative environment [@Börnigen2017]. Functional prediction also showed genes selecting for an "inflammatory bacteriome" is enriched is OSSC while genes complicitly involved in DNA repair and combination, purine metabolism, tyrosine and typtophan biosynthesis, glycolysis/gluconeogenesis, ribosome and ribosome biogenesis were significantly associated with healthy controls [@Al-Hebshi2017].

*P. aeruginosa* may play an important role in oral cancer as it is able to synthesize NO (nitric oxide). NO regulates different cancer-related factors such as apoptosis, cell cycle, angiogenesis, invasion, and metastasis. NO expression has been observed in various human cancers [@Korde2011]. Increased salivary NO concentrations are linked to patients with OSCC [@Bahar2007].

# Microbial signatures related to oropharyngeal and oral squamous cell carcinomas

It is important to acknowledge and begin studying the virome, bacteriome, mycobiome, and parasitic signatures in conjunction, to deepen our understanding of oral cancer. A study screening 100 oral cancer samples did whole genome and transcriptome amplification to predict the prominent signatures diseased samples. In their analysis they found the most viral sequences belonged to Papillomavirdae and this virus was detected in 98% of the cancer samples [@Banerjee2017]. There was an increase in abundance of Proteobacteria and Fusobacteria as stated earlier and an increase in gram-negative bacteria. However, the most prevalent species were Proteobacteria *Brevunidmonas* and Actinobacteria *Mobiluncus* 98% in cancer samples [@Banerjee2017].

The saliva of healthy co-horts and OSCC (oral squamous cell carcinomas) differed. It has been shown the microbiata taxanomic composition differed in abundance of certain species. In healthy controls an abundance of *Aggregatibacter*, *Lautropia*, *Haemophillus*, *Neisseria*, and *Leptotrichia* are significantly higher [@Chattopadhyay2019]. However, the saliva isolated from OSCC patients revealed *Capnocytophaga gingivalis*, *P. melaninogenica*, and *S. mitis* were highly abundant [@Mager2005; @Mager2006]. The taxonomic profile of healthy and OSCC saliva could potential serve as a diagnostic marker for oral cancer.

# Alcohol affects the microbiome and oral cancer susceptibility

Alcohol as an independent factor is known to increase the permeability of oral mucosa, which leads to easier penetration of carcinogens. The major metabolite of alcohol is acetaldehyde. Acetaldehyde causes DNA damage in cultured mammalian cells, and interferes with DNA synthesis and repair [@Homann2000]. There are studies showing alcohol's association with the oral microbiota. Alcohol has direct cytotoxic effects on bacteria by interupting salivia bacterium interactions, and provides ethanol as a substrate for bacterial metabolism [@Ingram1990; @Enberg2001; @Scannapieco1994]. In a large study of 1044 individuals testing alcoholic beverages in relationship to the oral microbiome showed oral dysbioses by 16S rRNA gene sequencing [@Fan2018]. Heavy alcoholic drinking may change bacterial composition. Thus, potential depletion of beneficial commensal bacteria and increased colonization of potentially pathogenic bacteria. These changes could contribute to alcohol-related diseases, such as periodontal disease, head and neck cancer, and gastrointestinal tract cancers. The improved understanding of the causes and health impacts of oral dysbiosis can edge microbiome-targeted therapeutic approaches.

# HPV infection and oral cancer

Human papillomavirus is highly prevalent in cervical cancer with woman in the United States. I has also been shown to be causative for the development of oral cancer [@Skaaby2014]. There are many types of HPV, not all are involved in cancer development. However, HPV 16 and HPV 18 are high risk types and are generally involved in human cancers, especially woman cervical cancer. HPV 16 was shown to be more correlated with oral squamous cell carcinomas rather than HPV 18 [@Anaya-Saavedra2008]. In a study, oropharyngeal cancer patients were HPV positive mostly because of high risk oral sexual behaviors [@Chow2020]. There is not strong evidence of the mechanisms HPV 16 might have on the oral microbiome, but correlation between HPV 16 infection and oral cancer is strongly supported across multiple studies.

# Future directions

Thousands of individuals suffer with oral cancer every year within the United States. Even with continued advancements in technologies and medicine the treatment and death rate of oral cancer remains unchanged for the last century. There are various directions oral microbiome research can delve into to strengthen the role the microbiome has in oral cancer.

Metabolic pathways such as LPS, iron production, and vitamin and cofactors could undergo gene deletion or altered gene expression to identify how important they are to oral cancer. A challenging, but rewarding model would be creating a sufficient human oral organoid to further test co-inoculation and whether certain taxonomic profiles can be deemed positive or negative to health. Further studies, to identify any metabolite that can prevent or mediate the effects of oral cancer are needed with *in vitro* then clinical trials. Higher throughput sequencing of the bacteriome, virome, and mycobiome to provide higher resolution of the total cross-talk of the oral cavity will strengthen the prediction of more precise therapies.

# References
