---
title: "Supplemental Methods: anvi'o"
author:
  - Tommy Tran, (KLemonLab) Tommy.Tran@bcm.edu
  - Ari Roberts, (KLemonLab) Ari.Roberts@bcm.edu
  - Isabel Escapa, (KLemonLab) Isabel.FernandezEscapa@bcm.edu
output:
  rmdformats::robobook:
    use_bookdown: true
    code_folding: show
bibliography: referencesAnvio.bib
---

```{r klippy, echo=FALSE, include=TRUE}
klippy::klippy(c('r', 'bash'))
```

This notebook contains analyses using anvi'o version 7.1 [@Eren2021]. For more in-depth information about anvi'o go to their official online site here: <https://anvio.org/>

# Functional annotations

We use as input (`path_i`) the anvi'o contigs database files (`.db`) that were generated in [[SupplementalMethods_Prokka_Annotations]{.underline}](SupplementalMethods_Prokka_Annotations.html).

## COG annotation

First we download and set up the NCBI's Clusters of Orthologous Groups database using `anvi-setup-ncbi-cogs`. We used the COG20 version.

Now the `anvi-run-ncbi-cogs` command can be used to annotate the .db genome files against the NCBI's COGs database [@Tatusov2001]: It uses DIAMOND [@Buchfink2015] "fast" as default, but it is recommended to use the "sensitive" option. This will take considerably much longer. Blastp can also be used instead of DIAMOND.

```{bash, eval = FALSE}
#conda activate anvio-7.1
path_i="analysis_Anvio7/Contigs_db"

for file in $path_i/*.db; do
    anvi-run-ncbi-cogs -T 8 --sensitive -c $file;
done
```

## KEGG annotation

First we download and set up the KEGG KOfam database using `anvi-setup-kegg-kofams`. We used the KEGG snapshot provided by Anvio 7.1.

The program `anvi-run-kegg-kofams` annotates genes in a given anvi'o contigs database with KEGG Orthology (KO) numbers via hits to the KEGG KOfam database (release 102.0) [@Kanehisa2000].

```{bash, eval = FALSE}
#conda activate anvio-7.1
path_i="analysis_Anvio7/Contigs_db"

for file in $path_i/*.db; do
    anvi-run-kegg-kofams -T 8 --log-bitscores -c $file;
done
```

The option `--log-bitscores` records the bit scores values. They were saved to the main folder for the repo, so we moved them to a new subfolder:

```{bash, eval = FALSE}
#conda activate anvio-7.1
mkdir "analysis_Anvio7/KEGGbitscores"
mv *_bitscores.txt analysis_Anvio7/KEGGbitscores/
```

## PFAM annotation

First we need to download and set up the PFAM database using `anvi-setup-pfams`. We used the 35.0 version.

The program `anvi-run-kegg-pfam` annotates genes in a given anvi'o contigs database with the PFAM database [@Mistry2021].

```{bash, eval = FALSE}
#conda activate anvio-7.1
path_i="analysis_Anvio7/Contigs_db"

for file in $path_i/*.db; do
    anvi-run-pfams -T 8 -c $file;
done
```

## Running HMMs

The following command adds the occurrence of bacterial single-copy genes across your contigs to the anvi'o contigs databases [@Eddy2011].

```{bash, eval = FALSE}
#conda activate anvio-7.1
path_i="analysis_Anvio7/Contigs_db"

for file in $path_i/*.db; do
    anvi-run-hmms -T 8 -c $file --just-do-it;
done
```

## Annotation Summaries

This generates basic stats on the annotations:

```{bash, eval = FALSE}
#conda activate anvio-7.1
mkdir -p "analysis_Anvio7/Contigs_stats"

path_i="analysis_Anvio7/Contigs_db"
path_o="analysis_Anvio7/Contigs_stats"

for file in $path_i/*.db; do
    FILENAME=`basename ${file%.*}`
    anvi-display-contigs-stats $file --report-as-text -o $path_o/stats_$FILENAME.txt;
done
```

This provides basic info on the contigs databases:

```{bash, eval = FALSE}
#conda activate anvio-7.1
path_i="analysis_Anvio7/Contigs_db"

for file in $path_i/*.db; do
    anvi-db-info $file;
done
```

Terminal output saved to `analysis_Anvio7/anvi-db-info_log.txt`.

# Pangenome Analysis

## Generating an anvi'o genomes storage database

The program `anvi-gen-genomes-storage` requires a `.txt` file with the names and path locations of the genomes to be included in the genome storage database. The storage database generated with `CorPGA_genome_list_PangenomeAnalysis.txt` includes 104 genomes, and was used to run the pangenome analysis for each of the four *Corynebacterium* species.

```{bash, eval = FALSE}
#conda activate anvio-7.1
mkdir -p "analysis_Anvio7/Pangenomic_Analysis"

anvi-gen-genomes-storage -e "analysis_Anvio7/CorPGA_genome_list_PangenomeAnalysis.txt" -o "analysis_Anvio7/Pangenomic_Analysis/CorPGA-All4Cor-GENOMES.db" --gene-caller Prodigal
```

This provides basic info on the generated genomes storage database:

```{bash, eval = FALSE}
#conda activate anvio-7.1
anvi-db-info "analysis_Anvio7/Pangenomic_Analysis/CorPGA-All4Cor-GENOMES.db"
```

## Running the pangenomic analyses

In order to analyze the pangenome individually for each of the four *Corynebacterium* species we created a `.txt` file including the list of genomes for each species and ran `anvi-pan-genome` with the `-G` flag:

```{bash, eval = FALSE}
#conda activate anvio-7.1
anvi-pan-genome -g "analysis_Anvio7/Pangenomic_Analysis/CorPGA-All4Cor-GENOMES.db" \
                -G "analysis_Anvio7/CorPGA_genome_list_PangenomeCpr.txt" \
                -o "analysis_Anvio7/Pangenomic_Analysis/Only_Cpr" -n PAN_Cpr \
                --use-ncbi-blast --mcl-inflation 10  --num-threads 8
                
anvi-pan-genome -g "analysis_Anvio7/Pangenomic_Analysis/CorPGA-All4Cor-GENOMES.db" \
                -G "analysis_Anvio7/CorPGA_genome_list_PangenomeCps.txt" \
                -o "analysis_Anvio7/Pangenomic_Analysis/Only_Cps" -n PAN_Cps \
                --use-ncbi-blast --mcl-inflation 10  --num-threads 8                
                
anvi-pan-genome -g "analysis_Anvio7/Pangenomic_Analysis/CorPGA-All4Cor-GENOMES.db" \
                -G "analysis_Anvio7/CorPGA_genome_list_PangenomeCac.txt" \
                -o "analysis_Anvio7/Pangenomic_Analysis/Only_Cac" -n PAN_Cac \
                --use-ncbi-blast --mcl-inflation 10  --num-threads 8

anvi-pan-genome -g "analysis_Anvio7/Pangenomic_Analysis/CorPGA-All4Cor-GENOMES.db" \
                -G "analysis_Anvio7/CorPGA_genome_list_PangenomeCtu.txt" \
                -o "analysis_Anvio7/Pangenomic_Analysis/Only_Ctu" -n PAN_Ctu \
                --use-ncbi-blast --mcl-inflation 10  --num-threads 8
```

## Displaying the pangenomes

This will display the pangenome in your default internet browser. Anvi'o says google chrome is the most compatible with the anvi''o interactive interface. You need to pick one of the `-p` inputs for each PAN.db file:

```{bash, eval = FALSE}
#conda activate anvio-7.1
anvi-display-pan -p "analysis_Anvio7/Pangenomic_Analysis/Cps/PAN_Cps-PAN.db" -g "analysis_Anvio7/Pangenomic_Analysis/CorPGA-All4Cor-GENOMES.db"

anvi-display-pan -p "analysis_Anvio7/Pangenomic_Analysis/Cpr/PAN_Cpr-PAN.db" -g "analysis_Anvio7/Pangenomic_Analysis/CorPGA-All4Cor-GENOMES.db"

anvi-display-pan -p "analysis_Anvio7/Pangenomic_Analysis/Cac/PAN_Cac-PAN.db" -g "analysis_Anvio7/Pangenomic_Analysis/CorPGA-All4Cor-GENOMES.db"

anvi-display-pan -p "analysis_Anvio7/Pangenomic_Analysis/Ctu/PAN_Ctu-PAN.db" -g "analysis_Anvio7/Pangenomic_Analysis/CorPGA-All4Cor-GENOMES.db"
```

If you are using Windows and it cannot display the interactive page then try copy and pasting either 127.0.0.1:8080 or <http://localhost:8080/> into your browser. Do not use sudo in the front of the command either.

Once it opens in the browser, you can adjust image settings and click the "DRAW" icon to view it. For help using the interactive interface see <http://merenlab.org/2016/02/27/the-anvio-interactive-interface/>

We manually reordered the layers to match the phylogenomic trees per species using the click and drag option on the interactive interface. The order names is inverse to the layers in the concentric diagram. When all the layers were re-ordered to match the phylogenomic tree we clicked the draw button and saved the svg file.

## Creating bin collections

### Core vs. Accessory

In order to define the Core vs Accessory pangenome for each of the four species level pangenomes we use anvi'o interactive interface. In the "Bins" tab we created the corresponding bins. Using the "Search" tab we used "Search gene clusters using filters" and "Append splits to selected bin" to create bins. The following tables summarizes how the bins in the collection named **CorevsAccessory** were created (last column indicates the criteria used on each search):

#### *C. propinquum* pangenome

| Pangenome Bin    | \# of Gene Clusters/Pangenome | Percentage                    | Gene Cluster occurs in \# of genomes |
|------------------|------------------|------------------|-------------------|
| Core             | 1824/3108                     | `r round(100*(1824/3108),1)`% | 19                                   |
| Single Copy Core | 1715/3108                     | `r round(100*(1715/3108),1)`% | 19                                   |
| Multicopy Core   | 109/3108                      | `r round(100*(109/3108),1)`%  | 19                                   |
| Soft Core        | 39/3108                       | `r round(100*(39/3108),1)`%   | 18                                   |
| Shell            | 689/3108                      | `r round(100*(689/3108),1)`%  | 3-17                                 |
| Cloud            | 556/3108                      | `r round(100*(556/3108),1)`%  | 1-2                                  |

#### *C. pseudodiphtheriticum* pangenome

| Pangenome Bin    | \# of Gene Clusters/Pangenome | Percentage                    | Gene Cluster occurs in \# of genomes |
|------------------|------------------|------------------|-------------------|
| Core             | 1714/3590                     | `r round(100*(1714/3590),1)`% | 43                                   |
| Single Copy Core | 1488/3590                     | `r round(100*(1488/3590),1)`% | 43                                   |
| Multicopy Core   | 226/3590                      | `r round(100*(226/3590),1)`%  | 43                                   |
| Soft Core        | 46/3590                       | `r round(100*(46/3590),1)`%   | 40-42                                |
| Shell            | 972/3590                      | `r round(100*(972/3590),1)`%  | 3-39                                 |
| Cloud            | 858/3590                      | `r round(100*(858/3590),1)`%  | 1-2                                  |

#### *C. accolens* pangenome

| Pangenome Bin    | \# of Gene Clusters/Pangenome | Percentage                    | Gene Cluster occurs in \# of genomes |
|------------------|------------------|------------------|-------------------|
| Core             | 1904/3427                     | `r round(100*(1904/3427),1)`% | 34                                   |
| Single Copy Core | 1734/3427                     | `r round(100*(1734/3427),1)`% | 34                                   |
| Multicopy Core   | 170/3427                      | `r round(100*(170/3427),1)`%  | 34                                   |
| Soft Core        | 77/3427                       | `r round(100*(77/3427),1)`%   | 32-33                                |
| Shell            | 676/3427                      | `r round(100*(676/3427),1)`%  | 3-31                                 |
| Cloud            | 770/3427                      | `r round(100*(770/3427),1)`%  | 1-2                                  |

#### *C. tuberculostearicum* pangenome

| Pangenome Bin    | \# of Gene Clusters/Pangenome | Percentage                    | Gene Cluster occurs in \# of genomes |
|------------------|------------------|------------------|-------------------|
| Core             | 1915/2907                     | `r round(100*(1915/2907),1)`% | 8                                    |
| Single Copy Core | 1860/2907                     | `r round(100*(1860/2907),1)`% | 8                                    |
| Multicopy Core   | 55/2907                       | `r round(100*(55/2907),1)`%   | 8                                    |
| Soft Core        | 47/2907                       | `r round(100*(47/2907),1)`%   | 7                                    |
| Shell            | 270/2907                      | `r round(100*(270/2907),1)`%  | 3-6                                  |
| Cloud            | 671/2907                      | `r round(100*(671/2907),1)`%  | 1-2                                  |

### PPanGGOLiN Partitions

PPanGGOLiN partitions were imported as a collection of bins named **PPanGGOLiNpartitions** as described in [PPanGGOLiN analysis](https://klemonlab.github.io/CorPGA_Pangenomics/SupplementalMethods_PPanGGOLiN.html).

## Summarizing the pangenome files

Using the program `anvi-summarize` we exported summary tables for each of the four species level pangenomes.

The resulting summaries contain a `gene_clusters_summary.txt.gz` file that links each gene to gene clusters, genomes, functions, layers, and bins selected from the interface. We used the flag `-C` to selected the **CorevsAccessory** collection described above:

```{bash, eval = FALSE}
#conda activate anvio-7.1
anvi-summarize -p "analysis_Anvio7/Pangenomic_Analysis/Cps/PAN_Cps-PAN.db" \
               -g "analysis_Anvio7/Pangenomic_Analysis/CorPGA-All4Cor-GENOMES.db" \
               -o "analysis_Anvio7" \
               -C CorevsAccessory
               
anvi-summarize -p "analysis_Anvio7/Pangenomic_Analysis/Cpr/PAN_Cpr-PAN.db" \
               -g "analysis_Anvio7/Pangenomic_Analysis/CorPGA-All4Cor-GENOMES.db" \
               -o "analysis_Anvio7" \
               -C CorevsAccessory
               
anvi-summarize -p "analysis_Anvio7/Pangenomic_Analysis/Ctu/PAN_Ctu-PAN.db" \
               -g "analysis_Anvio7/Pangenomic_Analysis/CorPGA-All4Cor-GENOMES.db" \
               -o "analysis_Anvio7" \
               -C CorevsAccessory
               
anvi-summarize -p "analysis_Anvio7/Pangenomic_Analysis/Cac/PAN_Cac-PAN.db" \
               -g "analysis_Anvio7/Pangenomic_Analysis/CorPGA-All4Cor-GENOMES.db" \
               -o "analysis_Anvio7" \
               -C CorevsAccessory
```

The output files were used for the COG functional analysis based on PPanGGOLiN partitions.

# Metabolic Enrichment Analysis

## Generating an anvi'o genomes storage database

A new anvi'o storage database was created with the genomes included in the metabolic enrichment analysis. The file `CorPGA_genome_list_MetabolicAnalysis.txt` includes 102 *Corynebacterium* genomes plus 27 *Dolosigranulum pigrum* genomes described in [@floresramos2021] (KPL3086 was not included in the analysis due to being highly similar to KPL3065).

```{bash, eval = FALSE}
#conda activate anvio-7.1
mkdir -p "analysis_Anvio7/Metabolic_Analysis"

anvi-gen-genomes-storage -e "analysis_Anvio7/CorPGA_genome_list_MetabolicAnalysis.txt" -o "analysis_Anvio7/Metabolic_Analysis/CorPGA-All4CorDpi-GENOMES.db" --gene-caller Prodigal
```

This provides basic info on the generated genomes storage database:

```{bash, eval = FALSE}
#conda activate anvio-7.1
anvi-db-info "analysis_Anvio7/Metabolic_Analysis/CorPGA-All4CorDpi-GENOMES.db"
```

## Running the pangenomic analyses

We ran `anvi-pan-genome` to generate several new pangenome files:

-   One combining all 102 *Corynebacterium* species and 27 *D. pigrum* genomes included in the new genome storage database. This file will be used to compare the metabolic capabilities between the *Corynebacterium* and *D. pigrum* strains.
    -   --mcl-inflation 10 generated 9,714 gene clusters
    -   --mcl-inflation 2 generated 9,210  gene clusters
-   Another one using the `-G` flag to subset the clustering to only the 102 *Corynebacterium* genomes listed in `CorPGA_genome_list_MetabolicAnalysis4Cor.txt`.
-   A final one using the `-G` flag to subset the clustering to only 41 *C. pseudodiphtheriticum* genomes selected to study metabolic different across geographic locations and listed in `CorPGA_genome_list_MetabolicAnalysisCps.txt.txt`.

```{bash, eval = FALSE}
#conda activate anvio-7.1

path_i="analysis_Anvio7/Metabolic_Analysis"

anvi-pan-genome -g $path_i/CorPGA-All4CorDpi-GENOMES.db \
                -o $path_i/PanAll4CorDpi_2 -n PAN_All4CorDpi \
                --use-ncbi-blast --mcl-inflation 2  --num-threads 8
                
anvi-pan-genome -g $path_i/CorPGA-All4CorDpi-GENOMES.db \
                -o $path_i/PanAll4CorDpi_10 -n PAN_All4CorDpi \
                --use-ncbi-blast --mcl-inflation 10  --num-threads 8

anvi-pan-genome -g $path_i/CorPGA-All4CorDpi-GENOMES.db \
                -G "analysis_Anvio7/CorPGA_genome_list_MetabolicAnalysis4Cor.txt" \
                -o $path_i/PanAll4Cor -n PAN_All4Cor \
                --use-ncbi-blast --mcl-inflation 10  --num-threads 8
                
anvi-pan-genome -g $path_i/CorPGA-All4CorDpi-GENOMES.db \
                -G "analysis_Anvio7/CorPGA_genome_list_MetabolicAnalysisCps.txt" \
                -o $path_i/PanCps -n PAN_Cps \
                --use-ncbi-blast --mcl-inflation 10  --num-threads 8
```

## Enrichment Analysis at the gene cluster level

We first define the groups of genomes that we want to compare the metabolic capabilities against each other and import them in the pangenome files using `anvi-import-misc-data` to import each corresponding `.txt` file:

```{bash, eval = FALSE}
#conda activate anvio-7.1

path_i="analysis_Anvio7/Metabolic_Analysis"

anvi-import-misc-data "analysis_Anvio7/CorPGA_groups_MetabolicAnalysisAll.txt" \
                      -p $path_i/PanAll4CorDpi_2/PAN_All4CorDpi-PAN.db \
                      --target-data-table layers
                      
anvi-import-misc-data "analysis_Anvio7/CorPGA_groups_MetabolicAnalysisAll.txt" \
                      -p $path_i/PanAll4CorDpi_10/PAN_All4CorDpi-PAN.db \
                      --target-data-table layers

anvi-import-misc-data "analysis_Anvio7/CorPGA_groups_MetabolicAnalysis4Cor.txt" \
                      -p $path_i/PanAll4Cor/PAN_All4Cor-PAN.db\
                      --target-data-table layers
                      
anvi-import-misc-data "analysis_Anvio7/CorPGA_groups_MetabolicAnalysisCps.txt" \
                      -p $path_i/PANCps/PAN_Cps-PAN.db \
                      --target-data-table layers
```

And then run `anvi-compute-functional-enrichment-in-pan` for each of the pangenomes based on the imported layers:

```{bash, eval = FALSE}
#conda activate anvio-7.1

mkdir -p "analysis_Anvio7/Metabolic_Analysis/GeneClusterLevel"
path_i="analysis_Anvio7/Metabolic_Analysis"
path_o="analysis_Anvio7/Metabolic_Analysis/GeneClusterLevel"

                                          
anvi-compute-functional-enrichment-in-pan -p $path_i/PanAll4CorDpi_2/PAN_All4CorDpi-PAN.db \
                                          -g $path_i/CorPGA-All4CorDpi-GENOMES.db \
                                          --category group \
                                          --annotation-source KEGG_Class \
                                          -o $path_o/CorPGA_enriched_All4CorDpi2_KEGG_Class.txt 
                                          
anvi-compute-functional-enrichment-in-pan -p $path_i/PanAll4CorDpi_2/PAN_All4CorDpi-PAN.db \
                                          -g $path_i/CorPGA-All4CorDpi-GENOMES.db \
                                          --category group \
                                          --annotation-source KEGG_Module \
                                          -o $path_o/CorPGA_enriched_All4CorDpi2_KEGG_Module.txt  
                                          
                                          
anvi-compute-functional-enrichment-in-pan -p $path_i/PanAll4CorDpi_10/PAN_All4CorDpi-PAN.db \
                                          -g $path_i/CorPGA-All4CorDpi-GENOMES.db \
                                          --category group \
                                          --annotation-source KEGG_Class \
                                          -o $path_o/CorPGA_enriched_All4CorDpi10_KEGG_Class.txt 
                                          
anvi-compute-functional-enrichment-in-pan -p $path_i/PanAll4CorDpi_10/PAN_All4CorDpi-PAN.db \
                                          -g $path_i/CorPGA-All4CorDpi-GENOMES.db \
                                          --category group \
                                          --annotation-source KEGG_Module \
                                          -o $path_o/CorPGA_enriched_All4CorDpi10_KEGG_Module.txt  
                                          
                                          
anvi-compute-functional-enrichment-in-pan -p $path_i/PanAll4Cor/PAN_All4Cor-PAN.db \
                                          -g $path_i/CorPGA-All4CorDpi-GENOMES.db \
                                          --category group \
                                          --annotation-source KEGG_Class \
                                          -o $path_o/CorPGA_enriched_All4Cor_KEGG_Class.txt 
                                          
anvi-compute-functional-enrichment-in-pan -p $path_i/PanAll4Cor/PAN_All4Cor-PAN.db \
                                          -g $path_i/CorPGA-All4CorDpi-GENOMES.db \
                                          --category group \
                                          --annotation-source KEGG_Module \
                                          -o $path_o/CorPGA_enriched_All4Cor_KEGG_Module.txt  
                                          
                                          
anvi-compute-functional-enrichment-in-pan -p $path_i/PANCps/PAN_Cps-PAN.db \
                                          -g $path_i/CorPGA-All4CorDpi-GENOMES.db \
                                          --category group \
                                          --annotation-source KEGG_Class \
                                          -o $path_o/CorPGA_enriched_Cps_KEGG_Class.txt 
                                          
anvi-compute-functional-enrichment-in-pan -p $path_i/PANCps/PAN_Cps-PAN.db \
                                          -g $path_i/CorPGA-All4CorDpi-GENOMES.db \
                                          --category group \
                                          --annotation-source KEGG_Module \
                                          -o $path_o/CorPGA_enriched_Cps_KEGG_Module.txt  
```

## Enrichment Analysis at the gene level

```{bash, eval = FALSE}
#conda activate anvio-7.1

mkdir -p "analysis_Anvio7/Metabolic_Analysis/GeneLevel"
path_o="analysis_Anvio7/Metabolic_Analysis/GeneLevel"

anvi-estimate-metabolism -e "analysis_Anvio7/CorPGA_genome_list_MetabolicAnalysis.txt" \
                         -O $path_o/CorPGA_AllStrains --matrix-format --include-metadata
                         
anvi-estimate-metabolism -e "analysis_Anvio7/CorPGA_genome_list_MetabolicAnalysis.txt" \
                         -O $path_o/CorPGA_AllStrains --kegg-output-modes modules,kofam_hits
                         
```

```{bash, eval = FALSE}
#conda activate anvio-7.1

path_i="analysis_Anvio7/Metabolic_Analysis/GeneLevel"

anvi-compute-metabolic-enrichment -M $path_i/CorPGA_AllStrains_modules.txt \
                                  -G "analysis_Anvio7/CorPGA_groups_MetabolicAnalysisAll.txt"\
                                  -o $path_i/CorPGA_enriched_modules_4CorDpi.txt

anvi-compute-metabolic-enrichment -M $path_i/CorPGA_AllStrains_modules.txt \
                                  -G "analysis_Anvio7/CorPGA_groups_MetabolicAnalysis4Cor.txt" \
                                  -o $path_i/CorPGA_enriched_modules_4Cor.txt
                                  
anvi-compute-metabolic-enrichment -M $path_i/CorPGA_AllStrains_modules.txt \
                                  -G "analysis_Anvio7/CorPGA_groups_MetabolicAnalysisCps.txt" \
                                  -o $path_i/CorPGA_enriched_modules_Cps.txt
```

<img src="images/Department-of-Molecular-Virology-&amp;-Microbiologyy-Horz-GRAY.png" align="left" width="200" height="90"/>

------------------------------------------------------------------------

------------------------------------------------------------------------

------------------------------------------------------------------------

------------------------------------------------------------------------

<center>

<h3>[REFERENCES]{.underline}</h3>

</center>
