---
title: "SupplementalMethods_GENOMES"
author: 
- Tommy Tran, (KLemonLab) Tommy.Tran@bcm.edu
output:
  rmdformats::robobook:
    use_bookdown: true
    code_folding: show
  github_document:
    toc: true
    toc_depth: 4
bibliography: referencesGENOMES.bib
---

```{r klippy, echo=FALSE, include=TRUE}
klippy::klippy(c('r', 'bash'))
```

# Genome Sourcing

Initially 212 *Corynebacterium* assemblies were collected and analyzed using PanACoTa [@Perrin2021a] in order to filter low quality assemblies (maximum value of L90 allowed = 100 & maximum number of contigs allowed to keep a genome = 999).

1.  **93 JASHHK genomes:** 94 KPL isolates were sequenced at NIH in collaboration with Julia Segre & Heidi Kong. This is follow up work on Wei Gao work with the CSF isolates. Only 1 out of 94 genomes was discarded because of low quality of the assembly (bovis_3648.spades.pilon.fasta, gsize=2626186, nb_conts=776 & L90=368). Several assemblies were identified as redundant pairs, but kept for follow up analysis.

2.  **16 BROAD genomes:** 16 KPL isolates were deposited at NCBI by the Broad as *Corynebacterium* sp. KPL#. None of the genomes were discarded because of low quality of the assembly. Several assemblies were identified as redundant pairs, but kept for follow up analysis.

3.  **32 RefSeq genomes:** We downloaded from NCBI reference assemblies using `PanACoTA prepare -s` subcommand. Default parameters were used, therefore genomes with mash distances that was not between 1e-4 and 0.06 were discarded to avoid redundant pairs or mislabeled assemblies.

4.  **71 BWA genomes:** 71 nasal isolates from Botswana were provided by Dr. Matthew Kelly at Duke. None of the genomes were discarded due to quality. However, 6 discarded due to redundancy. We choose the strain with higher quality to keep between the pairs.

Download NCBI reference assemblies with `PanACoTa prepare -s`:

```{bash, eval=FALSE}
    PanACoTA prepare -s "Corynebacterium accolens" -o "GENOMES/Genomes_NCBI/C_accolens"
    PanACoTA prepare -s "Corynebacterium macginleyi" -o "GENOMES/Genomes_NCBI/C_macginleyi"
    PanACoTA prepare -s "Corynebacterium propinquum" -o "GENOMES/Genomes_NCBI/C_propinquum"
    PanACoTA prepare -s "Corynebacterium pseudodiphtheriticum" -o "GENOMES/Genomes_NCBI/C_pseudodiphtheriticum"
    PanACoTA prepare -s "Corynebacterium tuberculostearicum" -o "GENOMES/Genomes_NCBI/C_tuberculostearicum"
```

# Finalize Genome Selection

## Quality Control and Mash Redundancy

We used PanACoTa in order to filter low quality assemblies with the default setting of a maximum L90 allowed to be = 100 and maximum number of contigs allowed to keep a genome = 999.

All *Corynebacterium* assemblies we have of these four species were batched into one folder.

I first use --max 1 and --min 0 in order to keep all strains from all species:

```{bash, eval=FALSE}
PanACoTA prepare --norefseq -d "/Users/tommytran/Downloads/CorPGA_genomes_final" -o "/Users/tommytran/Downloads/MASH_CorPGA_Final" --max 1 --min 0 
```

I run again the MASH [@Ondov2016]analysis with --max 1 and default --min 1e-4. With this step we can identify assemblies that are very similar and might be redundant:

```{bash, eval=FALSE}
PanACoTA prepare --norefseq -d "/Users/tommytran/Downloads/CorPGA_genomes_final" -o "/Users/tommytran/Downloads/MASH_CorPGA_Final_MinMash" --max 1 -M --info "/Users/tommytran/Downloads/MASH_CorPGA_Final/LSTINFO-NA-filtered-0.0_1.0.txt"
```

## All by All MASH table

```{r, eval=TRUE, echo=FALSE, warning=FALSE, message=FALSE}
library(readr)
CorPGA_ALLxALL_Mash <- read_csv("CorPGA_ALLxALL_Mash.csv")
```

```{r, eval=TRUE, echo=FALSE, message = FALSE, warning=FALSE}
library(knitr)
library(kableExtra)
```

```{r, echo=FALSE}
kbl(cbind(CorPGA_ALLxALL_Mash, CorPGA_ALLxALL_Mash)) %>%
  kable_styling(fixed_thead = T,  font_size = 12) %>%
  kable_paper() %>%
  scroll_box(width = "100%", height = "500px")
```

## Eliminating Genomes with **\<** 95% Identity to the Type Strain

We created ANI (Average Nucleotide Identity) matrices with GET_HOMOLOGUES [@Contreras-Moreira2013] for each of the four *Corynebacterium* species. These ANIs were generated using blastn and which means they were generated at the nucleotide level. We decided to look at all shared CDS (CoDing Sequences) regions and eliminated strains that did not meet a threshold of 95% of higher to the type strain of the concerning species. This left us with a total of 107 *Corynebacterium* strain genomes to proceed with.

<img src="images/Department-of-Molecular-Virology-&amp;-Microbiologyy-Horz-GRAY.png" align="left" width="200" height="90"/>

------------------------------------------------------------------------

------------------------------------------------------------------------

------------------------------------------------------------------------

------------------------------------------------------------------------

<center>

<h3>[REFERENCES]{.underline}</h3>

</center>
