---
title: "SupplementalMethods_GET_HOMOLOGUES"
author: "Tommy Tran (KLemonLab)"
date: 'Last updated: 09/20/2022'
output:
  rmdformats::robobook:
    use_bookdown: true
    code_folding: show
  github_document:
    toc: true
    toc_depth: 4
---

# GET_HOMOLOGUES

This notebook contains analyses using the genomic software GET_HOMOLOGUES version xxx. If you want to know more in-depth information about GET_HOMOLOGUES you can go to their official online manual here: <http://eead-csic-compbio.github.io/get_homologues/manual/>

We will want to use the gbk (genebank) file for each genome from the prokka output. We isolated the gbk files in new folders.

```{bash, eval=FALSE}
cd /directory/holding/prokka/output
mv *.gbk /new/directory/only/containing/gbk/files

We will want to use the gbk file for each genome from the prokka output. We isolated the gbk files in new folders.

cd /directory/holding/prokka/output
mv *.gbk /new/directory/only/containing/gbk/files
```

# ANI (Average Nucleotide Identity)

There will be **two** ANI matrices at the nucleotide level for each of the four species using OMCL (OrthoMuscle). The first one is to calculate ANI percentages from the core CDS (coding sequences) and the second is to calculate ANI values from all shared CDS regions. You must use gbk files outputted from prokka annotation.

## Generate ANI tab files

*C. propinquum* core CDS ANI tab file:

```{bash, eval=FALSE}
cd get_homologues
./get_homologues.pl -d "/Users/klemonlab/Downloads/Prokka_out_107_CorPGA_Anvio/108_gbf/Cpr" -A -t 19 -M -n 8 -C 90
```

*C. propinquum* all shared CDS regions ANI tab file:

```{bash, eval=FALSE}
cd get_homologues
./get_homologues.pl -d "/Users/klemonlab/Downloads/Prokka_out_107_CorPGA_Anvio/108_gbf/Cpr" -A -t 0 -M -n 8 -C 90
```

*C. pseudodiphtheriticum* core CDS ANI tab file:

```{bash, eval=FALSE}
cd get_homologues
./get_homologues.pl -d "/Users/klemonlab/Downloads/Prokka_out_107_CorPGA_Anvio/108_gbf/Cps" -A -t 43 -M -n 8 -C 90
```

*C. pseudodiphtheriticum* all shared CDS regions ANI tab file:

```{bash, eval=FALSE}
cd get_homologues
./get_homologues.pl -d "/Users/klemonlab/Downloads/Prokka_out_107_CorPGA_Anvio/108_gbf/Cps" -A -t 0 -M -n 8 -C 90
```

*C. tuberculostearicum* core CDS ANI tab file:

```{bash, eval=FALSE}
cd get_homologues
./get_homologues.pl -d "/Users/klemonlab/Downloads/Prokka_out_107_CorPGA_Anvio/108_gbf/Ctu" -A -t 8 -M -n 8 -C 90   
```

*C. tuberculostearicum* all shared CDS regions ANI tab file:

```{bash, eval=FALSE}
cd get_homologues
./get_homologues.pl -d "/Users/klemonlab/Downloads/Prokka_out_107_CorPGA_Anvio/108_gbf/Ctu" -A -t 0 -M -n 8 -C 90
```

*C. accolens* core CDS ANI tab file:

```{bash, eval=FALSE}
cd get_homologues
./get_homologues.pl -d "/Users/klemonlab/Downloads/Prokka_out_107_CorPGA_Anvio/108_gbf/Cac" -A -t 34 -M -n 8 -C 90
```

*C. accolens* all shared CDS regions ANI tab file:

```{bash, eval=FALSE}
cd get_homologues
./get_homologues.pl -d "/Users/klemonlab/Downloads/Prokka_out_107_CorPGA_Anvio/108_gbf/Cac" -A -t 0 -M -n 8 -C 90
```

## Convert ANI Tab Files into Editable svgs

*C. propinquum* core CDS ANI tab file:

```{bash, eval=FALSE}
cd get_homologues
./plot_matrix_heatmap.sh -i "/Users/klemonlab/get_homologues/Cpr_homologues/CprBWA290_f0_19taxa_algOMCL_e0_C90_Avg_identity.tab" -d 1 -t "Cpr Core CDS ANI" -H 9 -W 13
```

*C. propinquum* all shared CDS regions ANI tab file:

```{bash, eval=FALSE}
cd get_homologues
./plot_matrix_heatmap.sh -i "/Users/klemonlab/get_homologues/Cpr_homologues/CprBWA290_f0_0taxa_algOMCL_e0_C90_Avg_identity.tab" -d 1 -t "Cpr All Shared CDS Regions ANI" -H 9 -W 13
```

*C. pseudodiphtheriticum* core CDS ANI tab file:

```{bash, eval=FALSE}
cd get_homologues
./plot_matrix_heatmap.sh -i "/Users/klemonlab/get_homologues/Cps_homologues/CpsDU080_f0_43taxa_algOMCL_e0_C90_Avg_identity.tab" -d 1 -t "Cps Core CDS ANI" -H 16 -W 23
```

*C. pseudodiphtheriticum* all shared CDS regions ANI tab file:

```{bash, eval=FALSE}
cd get_homologues
./plot_matrix_heatmap.sh -i "/Users/klemonlab/get_homologues/Cps_homologues/CpsDU080_f0_0taxa_algOMCL_e0_C90_Avg_identity.tab" -d 1 -t "Cps All Shared CDS Regions ANI" -H 16 -W 23
```

*C. tuberculostearicum* core CDS ANI tab file:

```{bash, eval=FALSE}
cd get_homologues
./plot_matrix_heatmap.sh -i "/Users/klemonlab/get_homologues/Ctu_homologues/CtuBWA207_f0_8taxa_algOMCL_e0_C90_Avg_identity.tab" -d 1 -t "Ctu Core CDS ANI" -H 7 -W 8
```

*C. tuberculostearicum* all shared CDS regions ANI tab file:

```{bash, eval=FALSE}
cd get_homologues
./plot_matrix_heatmap.sh -i "/Users/klemonlab/get_homologues/Ctu_homologues/CtuBWA207_f0_0taxa_algOMCL_e0_C90_Avg_identity.tab" -d 1 -t "Ctu All Shared CDS Regions ANI" -H 7 -W 8
```

*C. accolens* core CDS ANI tab file:

```{bash, eval=FALSE}
cd get_homologues
./plot_matrix_heatmap.sh -i "/Users/klemonlab/get_homologues/Cac_homologues/CacKPL3832_f0_34taxa_algOMCL_e0_C90_Avg_identity.tab" -d 1 -t "Cac Core CDS ANI" -H 15 -W 19
```

*C. accolens* all shared CDS regions ANI tab file:

```{bash, eval=FALSE}
cd get_homologues
./plot_matrix_heatmap.sh -i "/Users/klemonlab/get_homologues/Cac_homologues/CacKPL3832_f0_0taxa_algOMCL_e0_C90_Avg_identity.tab" -d 1 -t "Cac All Shared CDS Regions ANI" -H 15 -W 19
```

## Edit ANI svg files

Once all eight svg files were created, they were each opened with adobe illustrator and edited to look like our figures.

# Conservative Core and Phylogeny of All 107 *Corynebacterium* Genomes

The core has to be determined by three seperate clustering algorithms. These algorithms include BDBH (bidirectional best hits), COGS (Clusters of Orthologous Genes), and OMCL (OrthoMCL). Once the core has been determined for each algorithm we will generate a venn diagram of the shared conservative gene clusters between them. This will be the number in the middle of the venn diagram where all three algorithms intersect.

## Running BDBH, COGS, & OMCL with 107 *Corynebacterium* Genomes

107 *Corynebacterium* genomes clustered with BDBH:

```{bash, eval=FALSE}
cd get_homologues
./get_homologues.pl -d "/Users/klemonlab/Downloads/Prokka_out_107_CorPGA/108_gbk/All_107_Coryne" -n 8 -t 107 -C 90
```

107 *Corynebacterium* genomes clustered with COGS:

```{bash, eval=FALSE}
cd get_homologues
./get_homologues.pl -d "/Users/klemonlab/Downloads/Prokka_out_107_CorPGA/108_gbk/All_107_Coryne" -n 8 -t 0 -C 90
```

107 *Corynebacterium* genomes clustered with OMCL:

```{bash, eval=FALSE}
cd get_homologues
./get_homologues.pl -d "/Users/klemonlab/Downloads/Prokka_out_107_CorPGA/108_gbk/All_107_Coryne" -n 8 -t 0 -C 90
```

## Creating a Conservative Core Venn Diagram

From my experience you will run into less issues here if you move the three cluster folders and cluster list files into a new directory. Then use the paths to those cluster folders after the `-d` flag seperated by commas.

Create conservative core venn diagram intersection to output faa files:

```{bash, eval=FALSE}
cd get_homologues
./compare_clusters.pl -o "/Users/klemonlab/Downloads/CorPGA_107_intersect" -d "/Users/klemonlab/Downloads/CorPGA_107_cluster_files/CpsDU080_f0_107taxa_algBDBH_e0_C90_","/Users/klemonlab/Downloads/CorPGA_107_cluster_files/CpsDU080_f0_0taxa_algOMCL_e0_C90_","/Users/klemonlab/Downloads/CorPGA_107_cluster_files/CpsDU080_f0_0taxa_algCOG_e0_C90_" -t 107
```

Create conservative core venn diagram intersection to output fna files:

```{bash, eval=FALSE}
cd get_homologues
./compare_clusters.pl -o "/Users/klemonlab/Downloads/CorPGA_107_intersect" -d "/Users/klemonlab/Downloads/CorPGA_107_cluster_files/CpsDU080_f0_107taxa_algBDBH_e0_C90_","/Users/klemonlab/Downloads/CorPGA_107_cluster_files/CpsDU080_f0_0taxa_algOMCL_e0_C90_","/Users/klemonlab/Downloads/CorPGA_107_cluster_files/CpsDU080_f0_0taxa_algCOG_e0_C90_" -t 107 -n
```

## Generate a Phylogenomic Tree of four *Corynebacterium* species from the Conservative Core

At this point we have created faa and faa files of the core gene clusters from 107 *Corynebacterium* genomes. To begin generating a phylogenomic tree we begin by isolating only the faa and fna files into one folder.

1.  At this point we have created faa and faa files of the core gene clusters from 107 *Corynebacterium* genomes. To begin generating a phylogenomic tree we begin by isolating only the faa and fna files into one folder.

Move all files with the extension .faa and fna to new destination folder:

```{bash, eval=FALSE}
cd /path/where/the/faa/and/fna/files/are # this should be the output folder from creating the venn diagram intersection
mv *.faa /Users/klemonlab/Downloads/CorPGA_107_faa_fna_only
mv *.fna /Users/klemonlab/Downloads/CorPGA_107_faa_fna_only
```

Now we use GET_PHYLOMARKERS version xxx to generate concatenated alignment file for each single copy core gene cluster. Delete Lines

1.  Now we use GET_PHYLOMARKERS version xxx to generate concatenated alignment file for each single copy core gene cluster. Delete Lines

Concatenate and align the faa and fna files to generate codon alignment files:

```{bash, eval=FALSE}
cd /new/path/where/only/faa/fna/files/are # go to directory holding all the .fna and .faa intersection of Corynebacterium
/Users/klemonlab/get_phylomarkers/run_get_phylomarkers_pipeline.sh -R 1 -t DNA -k 0.7 -m 0.7
```

Create a folder in IQTree v2.1.3 directory and move the codon alignment files outputted from GET_PHYLOMARKERS into that folder.

1.  Create a folder in IQTree v2.1.3 directory and move the codon alignment files outputted from GET_PHYLOMARKERS into that folder.

Make a directory to hold codon alignment files and move codon alignment files into that directory:

```{bash, eval=FALSE}
cd IQTree2
mkdir -p CorPGA_107_codon_alignment_files
cd /path/to/where/codon/alignment/files/are
mv *.fasta /Users/klemonlab/Documents/iqtree-2.1.3-MacOSX/CorPGA_107_codon_alignment_files
```

This will create a maximum likely hood tree and a consensus tree. **This tree is being generated from the gene clustering.** The .treefile was used because it is the best fit maximum likely hood tree. It was viewed through iTOL tree editor on google chrome browser.

The `-p` flag performs edge-linked proportional partition model for each of the individual clusters. Fast model selection for each cluster was determined by ModelFinder.

Run IQTree2.1.3 with the codon alignment files in the new folder:

```{bash, eval=FALSE}
cd IQTree2
bin/iqtree2 -p "CorPGA_107_codon_alignment_files" --prefix CorPGA_107_codon_alignment_files -alrt 1000 -B 1000 -T 8
```

The treefile generated from IQTree2 was then opened in iTol to scale and assign strain names from the `tree_labels.list` file created from GET_PHYLOMARKERS. After scaling and assigning strain names the svg version was exported from iTol and further edited in Adobe Illustrator.

# Generate *Corynebacterium* species Specific Phylogenomic Trees with an Outgroup

We are able to create phylogenies with an addition of an outgroup without losing the gene clusters not shared by the outgroup to the species of interest.

## Setting Up Two Datasets for Each of the Four Species

1.  First, we created a directory for each of the four species and only have the gbk (genebank) files for each of the four species in their corresponding directory.
2.  Secondly, we created another directory with the same four folders with the addition of an outgroup we chose for each species.

There was **eight** total directories. The first four only contain n (# of genomes of that specific species) in each of its directory. The latter four directories have n+1 genomes due to the addition of the outgroup we chose.

## Running BDBH, COGS, and OMCL on Each Directory

Running BDBH, COGS, and OMCL on only *C. propinquum*

```{bash, eval=FALSE}
cd get_homologues
./get_homologues.pl -d "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cpr" -n 8 -t 19 -C 90 # Runs BDBH
./get_homologues.pl -d "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cpr" -n 8 -t 0 -C 90 -M # Runs COGS
./get_homologues.pl -d "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cpr" -n 8 -t 0 -C 90 -G # Runs OMCL
```

Running BDBH, COGS, and OMCL on *C. propinquum* with an Outgroup

```{bash, eval=FALSE}
cd get_homologues
./get_homologues.pl -d "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cpr" -n 8 -t 20 -C 90 # Runs BDBH
./get_homologues.pl -d "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cpr" -n 8 -t 0 -C 90 -M # Runs COGS
./get_homologues.pl -d "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cpr" -n 8 -t 0 -C 90 -G # Runs OMCL
```

Running BDBH, COGS, and OMCL on only *C. pseudodiphtheriticum*

```{bash, eval=FALSE}
cd get_homologues
./get_homologues.pl -d "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cps" -n 8 -t 43 -C 90 # Runs BDBH
./get_homologues.pl -d "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cps" -n 8 -t 0 -C 90 -M # Runs COGS
./get_homologues.pl -d "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cps" -n 8 -t 0 -C 90 -G # Runs OMCL
```

Running BDBH, COGS, and OMCL on *C. pseudodiphtheriticum* with an Outgroup

```{bash, eval=FALSE}
cd get_homologues
./get_homologues.pl -d "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cps" -n 8 -t 44 -C 90 # Runs BDBH
./get_homologues.pl -d "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cps" -n 8 -t 0 -C 90 -M # Runs COGS
./get_homologues.pl -d "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cps" -n 8 -t 0 -C 90 -G # Runs OMCL
```

Running BDBH, COGS, and OMCL on only *C. tuberculostearicum*

```{bash, eval=FALSE}
cd get_homologues
./get_homologues.pl -d "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Ctu" -n 8 -t 8 -C 90 # Runs BDBH
./get_homologues.pl -d "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Ctu" -n 8 -t 0 -C 90 -M # Runs COGS
./get_homologues.pl -d "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Ctu" -n 8 -t 0 -C 90 -G # Runs OMCL
```

Running BDBH, COGS, and OMCL on *C. tuberculostearicum* with an Outgroup

```{bash, eval=FALSE}
cd get_homologues
./get_homologues.pl -d "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Ctu" -n 8 -t 9 -C 90 # Runs BDBH
./get_homologues.pl -d "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Ctu" -n 8 -t 0 -C 90 -M # Runs COGS
./get_homologues.pl -d "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Ctu" -n 8 -t 0 -C 90 -G # Runs OMCL
```

Running BDBH, COGS, and OMCL on only *C. accolens*

```{bash, eval=FALSE}
cd get_homologues
./get_homologues.pl -d "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cac" -n 8 -t 34 -C 90 # Runs BDBH
./get_homologues.pl -d "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cac" -n 8 -t 0 -C 90 -M # Runs COGS
./get_homologues.pl -d "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cac" -n 8 -t 0 -C 90 -G # Runs OMCL
```

Running BDBH, COGS, and OMCL on *C. accolens* with an Outgroup

```{bash, eval=FALSE}
cd get_homologues
./get_homologues.pl -d "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cac" -n 8 -t 35 -C 90 # Runs BDBH
./get_homologues.pl -d "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cac" -n 8 -t 0 -C 90 -M # Runs COGS
./get_homologues.pl -d "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cac" -n 8 -t 0 -C 90 -G # Runs OMCL
```

## Creating a Conservative Core of Only the Species of Interest and with the Additional Outgroup per Species

From my experience you will run into less issues here if you move the three cluster folders and cluster list files into a new directory. Then use the paths to those cluster folders after the `-d` flag seperated by commas.

Create conservative core venn diagram intersection to output faa and fna files for *C. propinquum*:

```{bash, eval=FALSE}
cd get_homologues
./compare_clusters.pl -o "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cpr/Cpr_intersection_19" -d "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cpr/Cpr_clusters_19/CprBWA290_f0_0taxa_algCOG_e0_C90_","/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cpr/Cpr_clusters_19/CprBWA290_f0_0taxa_algOMCL_e0_C90_","/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cpr/Cpr_clusters_19/CprBWA290_f0_19taxa_algBDBH_e0_C90_" -t 19 # Creates a conservative core with the output of faa files
./compare_clusters.pl -o "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cpr/Cpr_intersection_19" -d "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cpr/Cpr_clusters_19/CprBWA290_f0_0taxa_algCOG_e0_C90_","/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cpr/Cpr_clusters_19/CprBWA290_f0_0taxa_algOMCL_e0_C90_","/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cpr/Cpr_clusters_19/CprBWA290_f0_19taxa_algBDBH_e0_C90_" -t 19 -n # Creates a conservative core with the output of fna files from the additional `-n` flag
```

Create conservative core venn diagram intersection to output faa and fna files for *C. propinquum* with an outgroup:

```{bash, eval=FALSE}
cd get_homologues
./compare_clusters.pl -o "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cpr/Cpr_intersection_outgroup_20" -d "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cpr/Cpr_clusters_outgroup_20/CpsDSM44287_f0_0taxa_algCOG_e0_C90_","/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cpr/Cpr_clusters_outgroup_20/CpsDSM44287_f0_0taxa_algOMCL_e0_C90_","/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cpr/Cpr_clusters_outgroup_20/CpsDSM44287_f0_20taxa_algBDBH_e0_C90_" -t 20  # Creates a conservative core with the output of faa files
./compare_clusters.pl -o "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cpr/Cpr_intersection_outgroup_20" -d "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cpr/Cpr_clusters_outgroup_20/CpsDSM44287_f0_0taxa_algCOG_e0_C90_","/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cpr/Cpr_clusters_outgroup_20/CpsDSM44287_f0_0taxa_algOMCL_e0_C90_","/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cpr/Cpr_clusters_outgroup_20/CpsDSM44287_f0_20taxa_algBDBH_e0_C90_" -t 20 -n # Creates a conservative core with the output of fna files from the additional `-n` flag
```

Create conservative core venn diagram intersection to output faa and fna files for *C. pseudodiphtheriticum*:

```{bash, eval=FALSE}
cd get_homologues
./compare_clusters.pl -o "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cps/Cps_intersection_43" -d "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cps/Cps_clusters_43/CpsDU080_f0_43taxa_algBDBH_e0_C90_","/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cps/Cps_clusters_43/CpsDU080_f0_43taxa_algCOG_e0_C90_","/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cps/Cps_clusters_43/CpsDU080_f0_43taxa_algOMCL_e0_C90_" -t 43 # Creates a conservative core with the output of faa files
./compare_clusters.pl -o "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cps/Cps_intersection_43" -d "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cps/Cps_clusters_43/CpsDU080_f0_43taxa_algBDBH_e0_C90_","/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cps/Cps_clusters_43/CpsDU080_f0_43taxa_algCOG_e0_C90_","/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cps/Cps_clusters_43/CpsDU080_f0_43taxa_algOMCL_e0_C90_" -t 43 -n # Creates a conservative core with the output of fna files from the additional `-n` flag
```

Create conservative core venn diagram intersection to output faa and fna files for *C. pseudodiphtheriticum* with an outgroup:

```{bash, eval=FALSE}
cd get_homologues
./compare_clusters.pl -o "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cps/Cps_intersection_outgroup_44" -d "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cps/Cps_clusters_outgroup_44/CpsDU080_f0_0taxa_algCOG_e0_C90_","/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cps/Cps_clusters_outgroup_44/CpsDU080_f0_0taxa_algOMCL_e0_C90_","/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cps/Cps_clusters_outgroup_44/CpsDU080_f0_44taxa_algBDBH_e0_C90_" -t 44  # Creates a conservative core with the output of faa files
./compare_clusters.pl -o "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cps/Cps_intersection_outgroup_44" -d "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cps/Cps_clusters_outgroup_44/CpsDU080_f0_0taxa_algCOG_e0_C90_","/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cps/Cps_clusters_outgroup_44/CpsDU080_f0_0taxa_algOMCL_e0_C90_","/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cps/Cps_clusters_outgroup_44/CpsDU080_f0_44taxa_algBDBH_e0_C90_" -t 44 -n # Creates a conservative core with the output of fna files from the additional `-n` flag
```

Create conservative core venn diagram intersection to output faa and fna files for *C. tuberculostearicum*:

```{bash, eval=FALSE}
cd get_homologues
./compare_clusters.pl -o "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Ctu/Ctu_intersection_8" -d "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Ctu/Ctu_clusters_8/CtuBWA207_f0_0taxa_algCOG_e0_C90_","/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Ctu/Ctu_clusters_8/CtuBWA207_f0_0taxa_algOMCL_e0_C90_","/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Ctu/Ctu_clusters_8/CtuBWA207_f0_8taxa_algBDBH_e0_C90_" -t 8 # Creates a conservative core with the output of faa files
./compare_clusters.pl -o "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Ctu/Ctu_intersection_8" -d "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Ctu/Ctu_clusters_8/CtuBWA207_f0_0taxa_algCOG_e0_C90_","/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Ctu/Ctu_clusters_8/CtuBWA207_f0_0taxa_algOMCL_e0_C90_","/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Ctu/Ctu_clusters_8/CtuBWA207_f0_8taxa_algBDBH_e0_C90_" -t 8 -t 8 -n # Creates a conservative core with the output of fna files from the additional `-n` flag
```

Create conservative core venn diagram intersection to output faa and fna files for *C. tuberculostearicum* with an outgroup:

```{bash, eval=FALSE}
cd get_homologues
./compare_clusters.pl -o "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Ctu/Ctu_intersection_outgroup_9" -d "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Ctu/Ctu_clusters_outgroup_9/CtuBWA207_f0_0taxa_algCOG_e0_C90_","/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Ctu/Ctu_clusters_outgroup_9/CtuBWA207_f0_0taxa_algOMCL_e0_C90_","/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Ctu/Ctu_clusters_outgroup_9/CtuBWA207_f0_9taxa_algBDBH_e0_C90_" -t 9  # Creates a conservative core with the output of faa files
./compare_clusters.pl -o "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Ctu/Ctu_intersection_outgroup_9" -d "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Ctu/Ctu_clusters_outgroup_9/CtuBWA207_f0_0taxa_algCOG_e0_C90_","/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Ctu/Ctu_clusters_outgroup_9/CtuBWA207_f0_0taxa_algOMCL_e0_C90_","/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Ctu/Ctu_clusters_outgroup_9/CtuBWA207_f0_9taxa_algBDBH_e0_C90_" -t 9 -n # Creates a conservative core with the output of fna files from the additional `-n` flag
```

Create conservative core venn diagram intersection to output faa and fna files for *C. accolens*:

```{bash, eval=FALSE}
cd get_homologues
./compare_clusters.pl -o "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cac/Cac_intersection_34" -d "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cac/Cac_clusters_34/CacKPL3832_f0_0taxa_algCOG_e0_C90_","/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cac/Cac_clusters_34/CacKPL3832_f0_0taxa_algOMCL_e0_C90_","/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cac/Cac_clusters_34/CacKPL3832_f0_34taxa_algBDBH_e0_C90_" -t 34 # Creates a conservative core with the output of faa files
./compare_clusters.pl -o "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cac/Cac_intersection_34" -d "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cac/Cac_clusters_34/CacKPL3832_f0_0taxa_algCOG_e0_C90_","/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cac/Cac_clusters_34/CacKPL3832_f0_0taxa_algOMCL_e0_C90_","/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cac/Cac_clusters_34/CacKPL3832_f0_34taxa_algBDBH_e0_C90_" -t 34 -n # Creates a conservative core with the output of fna files from the additional `-n` flag
```

Create conservative core venn diagram intersection to output faa and fna files for *C. accolens* with an outgroup:

```{bash, eval=FALSE}
cd get_homologues
./compare_clusters.pl -o "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cac/Cac_intersection_outgroup_35" -d "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cac/Cac_clusters_outgroup_35/CacKPL3832_f0_35taxa_algBDBH_e0_C90_","/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cac/Cac_clusters_outgroup_35/CacKPL3832_f0_0taxa_algOMCL_e0_C90_","/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cac/Cac_clusters_outgroup_35/CacKPL3832_f0_0taxa_algCOG_e0_C90_" -t 35  # Creates a conservative core with the output of faa files
./compare_clusters.pl -o "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cac/Cac_intersection_outgroup_35" -d "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cac/Cac_clusters_outgroup_35/CacKPL3832_f0_35taxa_algBDBH_e0_C90_","/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cac/Cac_clusters_outgroup_35/CacKPL3832_f0_0taxa_algOMCL_e0_C90_","/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cac/Cac_clusters_outgroup_35/CacKPL3832_f0_0taxa_algCOG_e0_C90_" -t 35 -n # Creates a conservative core with the output of fna files from the additional `-n` flag
```

## Isolate fna and faa Files Together into a New Directory

Isolate the faa and fna files that make of the Conservative Core for *C. propinquum*:

```{bash, eval=FALSE}
cd "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cpr/Cpr_intersection_19" # Go to the directory holding the faa and fna files
mv *.faa "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cpr/Cpr_faa_fna_19" # Moves all faa files into the new folder holding only faa and fna files
mv *.fna "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cpr/Cpr_faa_fna_19" # Moves all fna files into the new folder holding only faa and fna files
```

Isolate the faa and fna files that make of the Conservative Core for *C. propinquum* with an Outgroup:

```{bash, eval=FALSE}
cd "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cpr/Cpr_intersection_outgroup_20" # Go to the directory holding the faa and fna files
mv *.faa "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cpr/Cpr_faa_fna_20" # Moves all faa files into the new folder holding only faa and fna files
mv *.fna "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cpr/Cpr_faa_fna_20" # Moves all fna files into the new folder holding only faa and fna files
```

Isolate the faa and fna files that make of the Conservative Core for *C. pseudodiphtheriticum*:

```{bash, eval=FALSE}
cd "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cps/Cps_intersection_43" # Go to the directory holding the faa and fna files
mv *.faa "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cps/Cps_faa_fna_43" # Moves all faa files into the new folder holding only faa and fna files
mv *.fna "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cps/Cps_faa_fna_43" # Moves all fna files into the new folder holding only faa and fna files
```

Isolate the faa and fna files that make of the Conservative Core for *C. pseudodiptheriticum* with an Outgroup:

```{bash, eval=FALSE}
cd "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cps/Cps_intersection_outgroup_44" # Go to the directory holding the faa and fna files
mv *.faa "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cps/Cps_faa_fna_44" # Moves all faa files into the new folder holding only faa and fna files
mv *.fna "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cps/Cps_faa_fna_44" # Moves all fna files into the new folder holding only faa and fna files
```

Isolate the faa and fna files that make of the Conservative Core for *C. tuberculostearicum*:

```{bash, eval=FALSE}
cd "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Ctu/Ctu_intersection_8" # Go to the directory holding the faa and fna files
mv *.faa "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Ctu/Ctu_faa_fna_8" # Moves all faa files into the new folder holding only faa and fna files
mv *.fna "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Ctu/Ctu_faa_fna_8" # Moves all fna files into the new folder holding only faa and fna files
```

Isolate the faa and fna files that make of the Conservative Core for *C. tuberculostearicum* with an Outgroup:

```{bash, eval=FALSE}
cd "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Ctu/Ctu_intersection_outgroup_9" # Go to the directory holding the faa and fna files
mv *.faa "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Ctu/Ctu_faa_fna_9" # Moves all faa files into the new folder holding only faa and fna files
mv *.fna "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Ctu/Ctu_faa_fna_9" # Moves all fna files into the new folder holding only faa and fna files
```

Isolate the faa and fna files that make of the Conservative Core for *C. accolens*:

```{bash, eval=FALSE}
cd "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cac/Cac_intersection_34" # Go to the directory holding the faa and fna files
mv *.faa "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cac/Cac_faa_fna_34" # Moves all faa files into the new folder holding only faa and fna files
mv *.fna "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cac/Cac_faa_fna_34" # Moves all fna files into the new folder holding only faa and fna files
```

Isolate the faa and fna files that make of the Conservative Core for *C. accolens* with an Outgroup:

```{bash, eval=FALSE}
cd "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cac/Cac_intersection_outgroup_35" # Go to the directory holding the faa and fna files
mv *.faa "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cac/Cac_faa_fna_35" # Moves all faa files into the new folder holding only faa and fna files
mv *.fna "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cac/Cac_faa_fna_35" # Moves all fna files into the new folder holding only faa and fna files
```

## Comparing the Conservative Core of the Species of Interest to the Conservative Core with an Additional Outgroup

This will compare the two conservative core of only the specis of interest to the conservative of the species of interest plus an outgroup. The addition of an outgroup causes the conservative core to decrase in value due to less gene clusters being shared with the outgroup. We are able to generate a list of which gene clusters are not shared with the outgroup from the strict conservative core. We can use this list to recover the lost gene clusters back into our analysis when making phylogenies.

*C. propinquum* Conservative Core vs *C. propinquum* Conservative Core + Outgroup:

```{bash, eval=FALSE}
cd get_homologues
./compare_clusters.pl -o "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cpr/Cpr_compare_clusters_19_20" -d "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cpr/Cpr_faa_fna_19","/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cpr/Cpr_faa_fna_20" -r 
```

*C. pseudodiphtheriticum* Conservative Core vs *C. pseudodiphtheriticum* Conservative Core + Outgroup:

```{bash, eval=FALSE}
cd get_homologues
./compare_clusters.pl -o "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cps/Cps_compare_clusters_43_44" -d "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cps/Cps_faa_fna_43","/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cps/Cps_faa_fna_44" -r 
```

*C. tuberculostearicum* Conservative Core vs *C. tuberculostearicum* Conservative Core + Outgroup:

```{bash, eval=FALSE}
cd get_homologues
./compare_clusters.pl -o "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Ctu/Ctu_compare_clusters_8_9" -d "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Ctu/Ctu_faa_fna_8","/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Ctu/Ctu_faa_fna_9" -r 
```

*C. accolens* Conservative Core vs *C. accolens* Conservative Core + Outgroup:

```{bash, eval=FALSE}
cd get_homologues
./compare_clusters.pl -o "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cac/Cac_compare_34_35_clusters" -d "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cac/Cac_faa_fna_34","/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cac/Cac_faa_fna_35" -r 
```

## Recover Lost Gene Clusters Due to the Addition of an Outgroup

Now in the output folder from the section above there will be a text file listing the gene clusters that the outgroup did NOT share with the species of interest. Find and move this text file to the directory holding only the faa and fna files of the conservative core of the specis of interest. We will use this text file to isolate and add these gene clusters into the directory holding the faa and fna files of the conservative core from the species of interest plus the outgroup.

Recover lost gene clusters due to addition of an outgroup *C. propinquum*:

```{bash, eval=FALSE}
cd "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cpr/Cpr_compare_clusters_19_20"
mv "unique_Cpr_faa_fna_19.venn_t0.txt" "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cpr/Cpr_faa_fna_19"
cd "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cpr/Cpr_faa_fna_19"
mkdir -p Recovered_GCs_Cpr
cat "unique_Cpr_faa_fna_19.venn_t0.txt" | xargs -J % cp % "Recovered_GCs_Cpr"
---
# We also need the fna files of these gene clusters. So, open the txt file with a text editor we use BBEdit. Search up ".faa" and replace all with ".fna". Then re-run the cat and xargs pipe. 
cat "unique_Cpr_faa_fna_19.venn_t0.txt" | xargs -J % cp % "Recovered_GCs_Cpr" # second run for fna files
cd "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cpr/Cpr_faa_fna_19/Recovered_GCs_Cpr" # Go inside recovered GCs folder
mv *.f* "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cpr/Cpr_faa_fna_20" # Move all the recovered GCs into the conservative  core + outgroup faa and fna directory
```

Recover lost gene clusters due to addition of an outgroup *C. pseudodiphtheriticum*:

```{bash, eval=FALSE}
cd "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cps/Cps_compare_clusters_43_44"
mv "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cps/Cps_compare_clusters_43_44/unique_Cps_faa_fna_43.venn_t0.txt" "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cps/Cps_faa_fna_43"
cd "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cps/Cps_faa_fna_43"
mkdir -p Recovered_GCs_Cps
cat "unique_Cps_faa_fna_43.venn_t0.txt" | xargs -J % cp % "Recovered_GCs_Cps"
---
# We also need the fna files of these gene clusters. So, open the txt file with a text editor we use BBEdit. Search up ".faa" and replace all with ".fna". Then re-run the cat and xargs pipe. 
cat "unique_Cps_faa_fna_43.venn_t0.txt" | xargs -J % cp % "Recovered_GCs_Cps" # second run for fna files
cd "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cps/Cps_faa_fna_43/Recovered_GCs_Cps" # Go inside recovered GCs folder
mv *.f* "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cps/Cps_faa_fna_44" # Move all the recovered GCs into the conservative  core + outgroup faa and fna directory
```

Recover lost gene clusters due to addition of an outgroup *C. tuberculostearicum*:

```{bash, eval=FALSE}
cd  "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Ctu/Ctu_compare_clusters_8_9"
mv "unique_Ctu_faa_fna_8.venn_t0.txt" "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Ctu/Ctu_faa_fna_8"
cd "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Ctu/Ctu_faa_fna_8"
mkdir -p Recovered_GCs_Ctu
cat "unique_Ctu_faa_fna_8.venn_t0.txt" | xargs -J % cp % "Recovered_GCs_Ctu"
---
# We also need the fna files of these gene clusters. So, open the txt file with a text editor we use BBEdit. Search up ".faa" and replace all with ".fna". Then re-run the cat and xargs pipe. 
cat "unique_Ctu_faa_fna_8.venn_t0.txt" | xargs -J % cp % "Recovered_GCs_Ctu" # second run for fna files
cd "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Ctu/Ctu_faa_fna_8/Recovered_GCs_Ctu" # Go inside recovered GCs folder
mv *.f* "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Ctu/Ctu_faa_fna_9" # Move all the recovered GCs into the conservative  core + outgroup faa and fna directory
```

Recover lost gene clusters due to addition of an outgroup *C. accolens*:

```{bash, eval=FALSE}
cd "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cac/Cac_compare_34_35_clusters"
mv "unique_Cac_faa_fna_34.venn_t0.txt" "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cac/Cac_faa_fna_34"
cd "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cac/Cac_faa_fna_34"
mkdir -p Recovered_GCs_Cac
cat "unique_Cac_faa_fna_34.venn_t0.txt" | xargs -J % cp % "Recovered_GCs_Cac"
---
# We also need the fna files of these gene clusters. So, open the txt file with a text editor we use BBEdit. Search up ".faa" and replace all with ".fna". Then re-run the cat and xargs pipe. 
cat "unique_Cac_faa_fna_34.venn_t0.txt" | xargs -J % cp % "Recovered_GCs_Cac" # second run for fna files
cd "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cac/Cac_faa_fna_34/Recovered_GCs_Cac" # Go inside recovered GCs folder
mv *.f* "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cac/Cac_faa_fna_35" # Move all the recovered GCs into the conservative  core + outgroup faa and fna directory
```

## Concatenate and Align Files using GET_PHYLOMARKERS vxxx for Each of the Four *Corynebacterium* species

Find the file "run_get_phylomarkers_pipeline.sh" in the get_phylomarkers directory. Then delete lines 925-935.

*C. propinquum* concatenate and align the faa and fna files to generate codon alignment files:

```{bash, eval=FALSE}
cd "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cpr/Cpr_faa_fna_20" # go to directory holding all the .fna and .faa intersection of Corynebacterium
/Users/klemonlab/get_phylomarkers/run_get_phylomarkers_pipeline.sh -R 1 -t DNA -k 0.7 -m 0.7
```

*C. pseudodiphtheriticum* concatenate and align the faa and fna files to generate codon alignment files:

```{bash, eval=FALSE}
cd "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cps/Cps_faa_fna_44" # go to directory holding all the .fna and .faa intersection of Corynebacterium
/Users/klemonlab/get_phylomarkers/run_get_phylomarkers_pipeline.sh -R 1 -t DNA -k 0.7 -m 0.7
```

*C. tuberculostearicum* concatenate and align the faa and fna files to generate codon alignment files:

```{bash, eval=FALSE}
cd "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Ctu/Ctu_faa_fna_9" # go to directory holding all the .fna and .faa intersection of Corynebacterium
/Users/klemonlab/get_phylomarkers/run_get_phylomarkers_pipeline.sh -R 1 -t DNA -k 0.7 -m 0.7
```

*C. accolens* concatenate and align the faa and fna files to generate codon alignment files:

```{bash, eval=FALSE}
cd "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cac/Cac_faa_fna_35" # go to directory holding all the .fna and .faa intersection of Corynebacterium
/Users/klemonlab/get_phylomarkers/run_get_phylomarkers_pipeline.sh -R 1 -t DNA -k 0.7 -m 0.7
```

## Generate a Phylogeny treefile using IQTree v2.1.3 for Four *Corynebacterium* species with Their Corresponding Outgroups

*C.propinquum* make a directory to hold codon alignment files and move codon alignment files into that directory:

```{bash, eval=FALSE}
cd "/Users/klemonlab/Documents/iqtree-2.1.3-MacOSX"
mkdir -p CorPGA_20_Cpr_outgroup
cd "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cpr/Cpr_faa_fna_20/get_phylomarkers_run_AIR1tDNA_k0.7_m0.7_Tmedium_11Jun22/codon_alignments"
mv *.fasta "/Users/klemonlab/Documents/iqtree-2.1.3-MacOSX/CorPGA_20_Cpr_outgroup"
```

*C. propinquum* run IQTree v2.1.3 with the codon alignment files:

```{bash, eval=FALSE}
cd "/Users/klemonlab/Documents/iqtree-2.1.3-MacOSX"
bin/iqtree2 -p "CorPGA_20_Cpr_outgroup" --prefix CorPGA_20_Cpr_outgroup -alrt 1000 -B 1000 -T 8
```

*C. pseudodiphtheriticum* make a directory to hold codon alignment files and move codon alignment files into that directory:

```{bash, eval=FALSE}
cd "/Users/klemonlab/Documents/iqtree-2.1.3-MacOSX"
mkdir -p CorPGA_44_Cps_outgroup
cd "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cps/Cps_faa_fna_44/get_phylomarkers_run_AIR1tDNA_k0.7_m0.7_Tmedium_11Jun22/codon_alignments"
mv *.fasta "/Users/klemonlab/Documents/iqtree-2.1.3-MacOSX/CorPGA_44_Cps_outgroup"
```

*C. pseudodiphtheriticum* run IQTree v2.1.3 with the codon alignment files:

```{bash, eval=FALSE}
cd "/Users/klemonlab/Documents/iqtree-2.1.3-MacOSX"
bin/iqtree2 -p "CorPGA_44_Cps_outgroup" --prefix CorPGA_44_Cps_outgroup -alrt 1000 -B 1000 -T 8
```

*C. tuberculostearicum* make a directory to hold codon alignment files and move codon alignment files into that directory:

```{bash, eval=FALSE}
cd "/Users/klemonlab/Documents/iqtree-2.1.3-MacOSX"
mkdir -p CorPGA_9_Ctu_outgroup
cd "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Ctu/Ctu_faa_fna_9/get_phylomarkers_run_AIR1tDNA_k0.7_m0.7_Tmedium_11Jun22/codon_alignments"
mv *.fasta "/Users/klemonlab/Documents/iqtree-2.1.3-MacOSX/CorPGA_9_Ctu_outgroup"
```

*C. tuberculostearicum* run IQTree v2.1.3 with the codon alignment files:

```{bash, eval=FALSE}
cd "/Users/klemonlab/Documents/iqtree-2.1.3-MacOSX"
bin/iqtree2 -p "CorPGA_9_Ctu_outgroup" --prefix CorPGA_9_Ctu_outgroup -alrt 1000 -B 1000 -T 8
```

*C. accolens* make a directory to hold codon alignment files and move codon alignment files into that directory:

```{bash, eval=FALSE}
cd "/Users/klemonlab/Documents/iqtree-2.1.3-MacOSX"
mkdir -p CorPGA_35_Cac_outgroup
cd "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cac/Cac_faa_fna_35/get_phylomarkers_run_AIR1tDNA_k0.7_m0.7_Tmedium_11Jun22/codon_alignments"
mv *.fasta "/Users/klemonlab/Documents/iqtree-2.1.3-MacOSX/CorPGA_35_Cac_outgroup"
```

*C. accolens* run IQTree v2.1.3 with the codon alignment files:

```{bash, eval=FALSE}
cd "/Users/klemonlab/Documents/iqtree-2.1.3-MacOSX"
bin/iqtree2 -p "CorPGA_35_Cac_outgroup" --prefix CorPGA_35_Cac_outgroup -alrt 1000 -B 1000 -T 8
```

This will create a maximum likely hood tree and a consensus tree for each species. **The trees are generated from the gene clustering.** The .treefiles were used because they are the best fit maximum likely hood tree. They was viewed through iTOL tree editor on google chrome browser.

The `-p` flag performs edge-linked proportional partition model for each of the individual clusters. Fast model selection for each cluster was determined by ModelFinder.

# Rarefaction Curves of Four *Corynebacterium* species

Here we will be using `./get_homologues.pl` with the flag `-c` to report genome composition analysis at the nucleotide level using OMCL.

The file **lib/marfil_homology.pm**, lines 136-137, which control how a gene is compared to previously processed pangenome genes in order to call it novel were changed: (We keep it at the default value)

From:

```{bash, eval=FALSE}
$MIN_PERSEQID_HOM = 0;
$MIN_COVERAGE_HOM = 20;
```

To:

```{bash, eval=FALSE}
$MIN_PERSEQID_HOM = 50;
$MIN_COVERAGE_HOM = 50;
```

## Create Core-genome and Pangenome Tab Files

Create *C. propinquum* rarefaction core-genome and pangenome tab file:

```{bash, eval=FALSE}
cd get_homologues
./get_homologues.pl -d "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cpr/Cpr_gbk_19" -c -n 8 -M -C 90 
```

Create *C. pseudodiphtheriticum* rarefaction core-genome and pangenome tab file:

```{bash, eval=FALSE}
cd get_homologues
./get_homologues.pl -d "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cps/Cps_gbk_43" -c -n 8 -M -C 90 
```

Create *C. tuberculostearicum* rarefaction core-genome and pangenome tab file:

```{bash, eval=FALSE}
cd get_homologues
./get_homologues.pl -d "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Ctu/Ctu_gbk_8" -c -n 8 -M -C 90 
```

Create *C. accolens* rarefaction core-genome and pangenome tab file:

```{bash, eval=FALSE}
cd get_homologues
./get_homologues.pl -d "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cac/Cac_gbk_34" -c -n 8 -M -C 90 
```

## Plot Core-genome and Pangenome Tab Files

Plot *C. propinquum* rarefaction core-genome curve:

```{bash, eval=FALSE}
cd get_homologues
./plot_pancore_matrix.pl -i "/Users/klemonlab/get_homologues/Cpr_gbk_19_homologues/core_genome_algOMCL_C90.tab" -f core_both
```

Plot *C. propinquum* rarefaction pangenome curve:

```{bash, eval=FALSE}
cd get_homologues
./plot_pancore_matrix.pl -i "/Users/klemonlab/get_homologues/Cpr_gbk_19_homologues/pan_genome_algOMCL_C90.tab" -f pan
```

Plot *C. pseudodiphtheriticum* rarefaction core-genome curve:

```{bash, eval=FALSE}
cd get_homologues
./plot_pancore_matrix.pl -i "/Users/klemonlab/get_homologues/Cps_gbk_43_homologues/core_genome_algOMCL_C90.tab" -f core_both
```

Plot *C. pseudodiphtheriticum* rarefaction pangenome curve:

```{bash, eval=FALSE}
cd get_homologues
./plot_pancore_matrix.pl -i "/Users/klemonlab/get_homologues/Cps_gbk_43_homologues/pan_genome_algOMCL_C90.tab" -f pan
```

Plot *C. tuberculostearicum* rarefaction core-genome curve:

```{bash, eval=FALSE}
cd get_homologues
./plot_pancore_matrix.pl -i "/Users/klemonlab/get_homologues/Ctu_gbk_8_homologues/core_genome_algOMCL_C90.tab" -f core_both
```

Plot *C. tuberculostearicum* rarefaction pangenome curve:

```{bash, eval=FALSE}
cd get_homologues
./plot_pancore_matrix.pl -i "/Users/klemonlab/get_homologues/Ctu_gbk_8_homologues/pan_genome_algOMCL_C90.tab" -f pan
```

Plot *C. accolens* rarefaction core-genome curve:

```{bash, eval=FALSE}
cd get_homologues
./plot_pancore_matrix.pl -i "/Users/klemonlab/get_homologues/Cac_gbk_34_homologues/core_genome_algOMCL_C90.tab" -f core_both
```

Plot *C. accolens* rarefaction pangenome curve:

```{bash, eval=FALSE}
cd get_homologues
./plot_pancore_matrix.pl -i "/Users/klemonlab/get_homologues/Cac_gbk_34_homologues/pan_genome_algOMCL_C90.tab" -f pan
```

## Edit svg Rarefaction Files

The svg rarefaction curves were opened and edited in Adobe Illustrator to their final forms.

# Create a Pangenome Matrix for Each of the Four *Corynebacterium* species

This will involve using the command `./get_homologues.pl`. However, this time we will use the flag `-t 0` to get all the possible clusters. Also, this can only be performed with the **OrthoMCL and COGS clustering algorithms.**

We have already ran these previously, so if you want you can use the cluster output directory that was produced and move to the next step.

*C. propinquum* find all the possible clusters with the COG algorithm:

```{bash, eval=FALSE}
cd get_homologues
./get_homologues.pl -d "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cps/Cps_gbk_43" -n 8 -G -t 0 -C 90
```

*C. propinquum* find all the possible clusters with the OrthoMCL algorithm:

```{bash, eval=FALSE}
cd get_homologues
./get_homologues.pl -d "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cpr/Cpr_gbk_19s" -n 8 -M -t 0 -C 90
```

## Compile and Create Pangenome Matrix Tab File

This will compile the following data and make a pangenome matrix. Here we add the flag `-m` to create the pangenome matrices intersection. This will create a pangenome_matrix.tab file that will be used next.

*C. propinquum* intersection pangenomes of OMCL & COGS:

```{bash, eval=FALSE}
cd get_homologues
./compare_clusters.pl -o "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cpr/Cpr_pan_intersection" -m -d "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cpr/Cpr_clusters_19/CprBWA290_f0_0taxa_algOMCL_e0_C90_","/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cpr/Cpr_clusters_19/CprBWA290_f0_0taxa_algCOG_e0_C90_"
```

*C. pseudodiphtheriticum* intersection pangenomes of OMCL & COGS:

```{bash, eval=FALSE}
cd get_homologues
./compare_clusters.pl -o "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cps/Cps_pan_intersection" -m -d "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cps/Cps_clusters_43/CpsDU080_f0_0taxa_algOMCL_e0_C90_","/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cps/Cps_clusters_43/CpsDU080_f0_0taxa_algCOG_e0_C90_"
```

*C. accolens* intersection pangenomes of OMCL & COGS:

```{bash, eval=FALSE}
cd get_homologues
./compare_clusters.pl -o "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cac/Cac_pan_intersection" -m -d "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cac/Cac_clusters_34/CacKPL3832_f0_0taxa_algOMCL_e0_C90_","/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cac/Cac_clusters_34/CacKPL3832_f0_0taxa_algCOG_e0_C90_"
```

*C. tuberculostearicum* intersection pangenomes of OMCL & COGS:

```{bash, eval=FALSE}
cd get_homologues
./compare_clusters.pl -o "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Ctu/Ctu_pan_intersection" -m -d "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Ctu/Ctu_clusters_8/CtuBWA207_f0_0taxa_algOMCL_e0_C90_","/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Ctu/Ctu_clusters_8/CtuBWA207_f0_0taxa_algCOG_e0_C90_"
```

## Plot the Pangenome Matrix Tab File

*C. propinquum* plot pangenome:

```{bash, eval=FALSE}
./parse_pangenome_matrix.pl -m "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cpr/Cpr_pan_intersection/pangenome_matrix_t0.tab" -s -x
```

*C. pseudodiphtheriticum* plot pangenome:

```{bash, eval=FALSE}
./parse_pangenome_matrix.pl -m "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cps/Cps_pan_intersection/pangenome_matrix_t0.tab" -s -x
```

*C. accolens* plot pangenome:

```{bash, eval=FALSE}
./parse_pangenome_matrix.pl -m "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Cac/Cac_pan_intersection/pangenome_matrix_t0.tab" -s -x
```

*C. tuberculostearicum* plot pangenome:

```{bash, eval=FALSE}
./parse_pangenome_matrix.pl -m "/Users/klemonlab/Desktop/THT_CorPGA/CorPGA_Ctu/Ctu_pan_intersection/pangenome_matrix_t0.tab" -s -x
```

<img src="images/Department-of-Molecular-Virology-&amp;-Microbiologyy-Horz-GRAY.png" align="left" width="200" height="90"/>
