---
title: "SupplementalMethods_Installations"
author: "Tommy Tran (KLemonLab)"
output:
  rmdformats::robobook:
    use_bookdown: true
    code_folding: show
  github_document:
    toc: true
    toc_depth: 4
---

# INSTALLATIONS

## Installing Conda (MacOSX)

First you can check if you have conda on your system by opening terminal and typing `conda` then pressing enter. If the command was not found you need follow up with this installation. If you already have it installed then run the command `conda update conda`.

```{bash, eval=FALSE}
conda
conda update conda
```

These are the [instructions](https://www.codecademy.com/articles/install-python-data-analysis) for installing Miniconda, a mini version of Anaconda that includes just conda, its dependencies, and Python. You can also install the full anaconda following this [link](https://docs.anaconda.com/anaconda/install/).

Use this [link](https://docs.conda.io/en/latest/miniconda.html) to find the correct Miniconda you need to download for your system such as MacOSX, Windows, and Linux.

1.  Once you have downloaded it you will need to open terminal.
2.  Now that terminal is open you will need to use this command: `miniconda-filename.sh`. The "miniconda-filename.sh" should be replaced with the actual name of the file you downloaded. Your file name will look something like Miniconda2-latest-MacOSX-x86_64.sh
3.  Continue to press `enter`, and type `yes` when it is needed.
4.  When it is all completed I recommend running `conda update conda`.
5.  Then check which conda you have by running `conda --version`.

Use this command if you do not prefer conda's base environment to be activated on startup in terminal.

```{bash, eval=FALSE}
conda config --set auto_activate_base false
```

Here is a [link](https://docs.conda.io/projects/conda/en/4.6.0/_downloads/52a95608c49671267e40c689e0bc00ca/conda-cheatsheet.pdf) to some conda commands that may come in handy.

## Installing PanACoTa

We create a new Python environment on the Mac Terminal:

```{bash, eval=FALSE}
conda create -y --name PanACoTa python=3.6
conda activate PanACoTa
```

### Dependencies

I install all dependencies in the new PanACoTa environment on the Mac Terminal:

```{bash, eval=FALSE}
conda install -c conda-forge -c bioconda -c defaults prokka
conda install -c conda-forge -c bioconda mmseqs2
```

-   For prepare module: mash (to filter genomes) From <https://github.com/marbl/Mash/releases> I get mash-OSX64-v2.2.tar, unzip and move the 'mash' binary to /usr/local/bin
-   For annotate module: prokka and/or prodigal (to uniformly annotate your genomes)
-   For pangenome module: mmseqs (to generate pangenomes)
-   For align module: mafft (to align persistent genome) From <https://mafft.cbrc.jp/alignment/software/macstandard.html>. I download and install mafft-7.453-signed.pkg
-   For tree module: IQ Tree, FastTreeMP, FastME or Quicktree (to infer a phylogenetic tree) From <http://www.iqtree.org> I get the Stable release 1.6.12 (August 15, 2019) and move the 'iqtree' binary to /usr/local/bin

### PanACoTA

PanACoTA Documentation:

-   <https://github.com/gem-pasteur/PanACoTA>
-   <https://aperrin.pages.pasteur.fr/pipeline_annotation/html-doc/>

By cloning the GitHub repository, you will then be able to update the code to new versions very easily and quickly. Here is how to clone the repository and install PanACoTA from it on the Mac Terminal:

```{bash, eval=FALSE}
git clone https://github.com/gem-pasteur/PanACoTA.git
cd PanACoTA
./make --user
```

To update source code to the new version: Here we use Version 1.0.1

```{bash, eval=FALSE}
cd PanACoTA
git pull 
./make upgrade
```

## Installing anvi'o v7

Follow instructions here: <https://merenlab.org/2016/06/26/installation-v2/#3-install-anvio>

> If you are using Windows WSL2 Linux through Ubuntu 20.04 LTS and get to the step of doing the `anvi-self-test`, and the html site cannot be reached. Then try and use 127.0.0.1:8080 or <http://localhost:8080/> to get the interactive display to work.

### Updating anvi'o from GitHub

```{bash, eval=FALSE}
conda activate anvio-7
pip install git+https://github.com/merenlab/anvio.git
#pip install --upgrade git+git://github.com/merenlab/anvio.git
```

### anvi'o Help

-   Help <http://merenlab.org/2019/10/07/getting-help/>
-   All Programs and Scripts <http://merenlab.org/software/anvio/vignette/>
-   A course on microbial 'omics <http://merenlab.org/momics/>
-   Vocabulary <http://merenlab.org/vocabulary/>
-   Interactive interface <http://merenlab.org/2016/02/27/the-anvio-interactive-interface/>

## Installing Prokka

Guide to installing Prokka using this link: <https://github.com/tseemann/prokka>

Install using conda through terminal. You can try installing it into your specific anvi'o environment but that did not work for us. So, we installed Prokka on our base python environment.

Run the command:

```{bash, eval=FALSE}
`conda install -c conda-forge -c bioconda -c defaults prokka`
```

Then check if you have prokka installed by running the command `prokka -v`

## Installing PPanGGOLiN

We create a new Python environment on the Terminal:

```{bash, eval=FALSE}
conda create -y --name PPanGGOLiN python=3.6
conda activate PPanGGOLiN
```

For basic info on PPanGGOLiN go here: <https://github.com/labgem/PPanGGOLiN/wiki>

The public release doesn't include the "Regions of Genome Plasticity" analysis, so we want the version from GitHub.

First we install on the Terminal the required dependencies:

```{bash, eval=FALSE}
conda activate PPanGGOLiN
conda install --file other/requirements.txt
```

Also on the Terminal we install PPanGGOLiN:

```{bash, eval=FALSE}
pip install git+https://github.com/labgem/PPanGGOLiN.git#egg=PPanGGOLiN
#pip install --upgrade git+git://github.com/labgem/PPanGGOLiN.git
```

## Installing GET_HOMOLOGUES

This should be done in your MacOSX terminal or in your Linux WSL2 subsystem if you are on a Windows machine operating machine.

For more info on get_homologues go here: <http://eead-csic-compbio.github.io/get_homologues/manual/>

Run the following commands below to clone the current get_homologues software from Github onto your system:

```{bash, eval=FALSE}
cd # path of directory of where you want get_homologues to be
git clone https://github.com/eead-csic-compbio/get_homologues.git
```

Run this perl script and install and necessary packages it may ask for.

```{bash eval=FALSE}
perl install.pl
```

Once you have finished these steps you can update get_homologues anytime by:

```{bash, eval=FALSE}
cd # this should be the path to your get_homologues
git pull
```

Test if it is on your system by running:

```{bash, eval=FALSE}
./get_homologues.pl -v 
```

The output of this will be the version and credits of get_homologues.

## Installing GET_PHYLOMARKERS

This should be done in your MacOSX terminal or in your Linux WSL2 subsystem if you are on a Windows machine operating machine.

For more info on get_phylomarkers go here: <https://vinuesa.github.io/get_phylomarkers/>

Run the following commands below to clone the current get_pylomarkers software from Github onto your system:

```{bash, eval=FALSE}
cd # path of directory of where you want get_homologues to be
git clone https://github.com/vinuesa/get_phylomarkers.git
```

After get_phylomarkers has been installed and you are within it's directory use the command below. This will install Rpackages that get_phylomarkers needs. If you are getting errors then you may need to install xquartz and updated R cran packages manually. I had to do both on my MacOSX system. Go to [xquartz](https://www.xquartz.org "xquartz official website") or [cran-r](https://cran.r-project.org "cran.r project official website") websites by clinking on the hyperlinks and then download and install packages you need.

```{bash, eval=FALSE}
./install_R\_deps.R
```

Test get_phylomarkers:

```{bash, eval=FALSE}
cd # into a directory that holds .faa and corresponding .fna files
your/path/to/run_get_phylomarkers_pipeline.sh -R 1 -t DNA
```

It should run smoothly and create an output folder that has get_phylomarkers and the date as the title.

You may want to add get_phylomarkers to your \$PATH. I did this on MacOSX.

```{bash, eval=FALSE}
nano ~/.zshrc
export PATH=$PATH:/Users/tommytran/bin # For this line use your own path /Users/yourname/bin
```

After you added your path exit nano and then close terminal. Start a new window and check if your path was added by running.

```{bash, eval=FALSE}
echo $PATH
```

<img src="images/Department-of-Molecular-Virology-&amp;-Microbiologyy-Horz-GRAY.png" align="left" width="200" height="90"/>
