---
title: "SupplementalMethods_GENOMES"
author: 
- Tommy Tran, (KLemonLab) Tommy.Tran@bcm.edu
- Isabel Escapa, (KLemonLab) Isabel.FernandezEscapa@bcm.edu
output:
  rmdformats::robobook:
    use_bookdown: true
    code_folding: show
  github_document:
    toc: true
    toc_depth: 4
bibliography: referencesGENOMES.bib
---

```{r klippy, echo=FALSE, include=TRUE}
klippy::klippy(c('r', 'bash'))
```

```{r, eval=TRUE, echo=FALSE, warning=FALSE, message=FALSE}
library(readr)
library(knitr)
library(kableExtra)
```

# Genome Sourcing

Initially 212 *Corynebacterium* assemblies were collected from the following sources:

1.  16 genomes labelled as **KPL** that correspond to isolates originally sequenced as part of the **HMP**. These were deposited as *Corynebacterium* sp. prior to this study by the Broad Institute under PRJNA169439.

2.  94 genomes from strains labelled as **KPL** and isolated in **MA** as part of this study and with a putative 16S rRNA gene assignment to a *Corynebacterium* spp. Deposited at NHI databases under PRJNA842433.

3.  71 genomes from strains labelled as **BWA** and isolated in **Botswana** as part of this study. Deposited at NHI databases under XXXXX.

4.  31 reference assemblies for the nasal *Corynebacterium* species of interest. They were downloaded using using [PanACoTa](https://github.com/gem-pasteur/PanACoTA) with the `PanACoTA prepare -s` subcommand searching by the name of the species, as written by the NCBI. Default parameters were used in order to discard:

    -   low quality assemblies: maximum value of L90 allowed = 100 and maximum number of contigs allowed to keep a genome = 999.
    -   redundant pairs or mislabeled assemblies: min MASH [@Ondov2016] distance to the reference = 1e-4 and max MASH [@Ondov2016] distance to the reference = 0.06.

```{bash, eval=FALSE}
    PanACoTA prepare -s "Corynebacterium <SPECIES_NAME>" 
```

# Genome Selection

## Eliminating Low Quality and reduntant Assemblies

We used PanACoTa again in order to filter low quality assemblies with the default setting of a maximum L90 allowed to be = 100 and maximum number of contigs allowed to keep a genome = 999. The MASH [@Ondov2016] parameters used were `--max 1` and the default `--min 1e-4`. This allowed us to discard assemblies that are very similar and might be redundant:

```{bash, eval=FALSE}
PanACoTA prepare --norefseq -d "<GENOMES_FOLDER>" -o "<OUTPUT_FOLDER>" --max 1 -M 
```

## Eliminating Genomes with **\<** 95% ANI to the Type Strain

We created ANI (Average Nucleotide Identity) matrices with GET_HOMOLOGUES [@Contreras-Moreira2013] for each of the four *Corynebacterium* species. These ANIs were generated using blastn, which means they were generated at the nucleotide level. We decided to look at all shared CDS (CoDing Sequences) regions and eliminated strains that did not meet a threshold of 95% of higher to the type strain of the concerning species. This left us with a total of 107 *Corynebacterium* strain genomes to proceed with.

## All by All MASH table for the 107 selected *Corynebacterium* genomes

```{r, eval=TRUE, echo=FALSE, warning=FALSE, message=FALSE}
CorPGA_ALLxALL_Mash <- read_csv("CorPGA_ALLxALL_Mash.csv")

kbl(CorPGA_ALLxALL_Mash, col.names = c("Query-ID","Reference-ID","mash distance","P-value","Matching-hashes"), align = "c") %>%
  kable_styling(fixed_thead = T,  font_size = 16) %>%
  kable_paper() %>%
  scroll_box(width = "100%", height = "500px")
```

<img src="images/Department-of-Molecular-Virology-&amp;-Microbiologyy-Horz-GRAY.png" align="left" width="200" height="90"/>

------------------------------------------------------------------------

------------------------------------------------------------------------

------------------------------------------------------------------------

------------------------------------------------------------------------

<center>

<h3>[REFERENCES]{.underline}</h3>

</center>
