---
title: "Supplemental Methods: anvi'o"
author:
  - Tommy Tran, (KLemonLab) Tommy.Tran@bcm.edu
  - Ari Roberts, (KLemonLab) Ari.Roberts@bcm.edu
  - Isabel Escapa, (KLemonLab) Isabel.FernandezEscapa@bcm.edu
output:
  rmdformats::robobook:
    use_bookdown: true
    code_folding: show
bibliography: referencesAnvio.bib
---

```{r klippy, echo=FALSE, include=TRUE}
klippy::klippy(c('r', 'bash'))
```

This notebook contains analyses using anvi'o version 7.1 [@Eren2021]. For more in-depth information about anvi'o go to their official online site here: <https://anvio.org/>

# Import functional annotations

We will use as input (`path_i`) the anvi'o contigs database files (`.db`) that were generated in [[SupplementalMethods_Prokka_Annotations]{.underline}](SupplementalMethods_Prokka_Annotations.html).

## COG annotation

First you need to download and set up the NCBI's Clusters of Orthologous Groups database using `anvi-setup-ncbi-cogs`. We used the COG20 version.

Now the `anvi-run-ncbi-cogs` command can be used to annotate the .db genome files against the NCBI's COGs database [@Tatusov2001]: It uses DIAMOND [@Buchfink2015] "fast" as default, but it is recommended to use the "sensitive" option. This will take considerably much longer. Blastp can also be used instead of DIAMOND.

```{bash, eval = FALSE}
path_i="analysis_Anvio7/Contigs_db"

for file in $path_i/*.db; do
    anvi-run-ncbi-cogs -T 8 --sensitive -c $file;
done
```

## KEGG annotation

First you need to download and set up the KEGG KOfam database using `anvi-setup-kegg-kofams`. We used the KEGG snapshot provided by Anvio 7.1.

The program `anvi-run-kegg-kofams` annotates genes in a given anvi'o contigs database with KEGG Orthology (KO) numbers via hits to the KEGG KOfam database (release 102.0) [@Kanehisa2000].

```{bash, eval = FALSE}
path_i="analysis_Anvio7/Contigs_db"

for file in $path_i/*.db; do
    anvi-run-kegg-kofams -T 8 --log-bitscores -c $file;
done
```

The option `--log-bitscores` records the bit scores values. They were saved to the main folder for the repo, so we moved them to a new subfolder:

```{bash, eval = FALSE}
mkdir "analysis_Anvio7/KEGGbitscores"
mv *_bitscores.txt analysis_Anvio7/KEGGbitscores/
```

## PFAM annotation

First you need to download and set up the PFAM database using `anvi-setup-pfams`. We used the 35.0 version.

The program `anvi-run-kegg-pfam` annotates genes in a given anvi'o contigs database with the PFAM database [@Mistry2021].

```{bash, eval = FALSE}
path_i="analysis_Anvio7/Contigs_db"

for file in $path_i/*.db; do
    anvi-run-pfams -T 8 -c $file;
done
```

## Running HMMs

When you run the following command the occurrence of bacterial single-copy genes across your contigs is added to your contigs database [@Eddy2011].

```{bash, eval = FALSE}
path_i="analysis_Anvio7/Contigs_db"

for file in $path_i/*.db; do
    anvi-run-hmms -T 8 -c $file --just-do-it;
done
```

## Annotation Summaries

This generates basic stats on the annotations:

```{bash, eval = FALSE}
mkdir -p "analysis_Anvio7/Contigs_stats"

path_i="analysis_Anvio7/Contigs_db"
path_o="analysis_Anvio7/Contigs_stats"

for file in $path_i/*.db; do
    FILENAME=`basename ${file%.*}`
    anvi-display-contigs-stats $file --report-as-text -o $path_o/stats_$FILENAME.txt;
done
```

This provides basic info on the contigs databases:

```{bash, eval = FALSE}
path_i="analysis_Anvio7/Contigs_db"

for file in $path_i/*.db; do
    anvi-db-info $file;
done
```

Terminal output saved to `analysis_Anvio7/anvi-db-info_log.txt`.

# Pangenome Analysis

## Generating an anvi'o genomes storage

The program `anvi-gen-genomes-storage` requires you to provide names and locations of the genomes to be included in the genome storage database in a `.txt` file. We created the All_104_CorynesAnvio.txt file with the file names and paths. This file includes the name anvi'o will use internally to identify each genome. We added the prefix "RefSeq\_" to the reference genomes because some of them start with a digit and anvi'o only allows names that start with an ASCII letter. We made equivalent files for each of the four species.

We created a storage database with the 104 genomes that can be used to run the pangenomic analysis with all the strains at once or by selecting subsets of strains (i.e. by species, or by grouping related species together).

```{bash, eval = FALSE}
mkdir -p "analysis_Anvio7/Pangenomic_Results"

anvi-gen-genomes-storage -e "analysis_Anvio7/All_104_CorynesAnvio.txt" -o "analysis_Anvio7/Pangenomic_Results/Coryne-GENOMES.db" --gene-caller Prodigal
```

This provides basic info on the generated genomes storage database:

```{bash, eval = FALSE}
anvi-db-info "analysis_Anvio7/Pangenomic_Results/Coryne-GENOMES.db"
```

## Running the pangenomic analyses

We generated a `.txt` file for each of the four species including the list of genomes/paths for that group and run the pangenome analysis:

```{bash, eval = FALSE}
anvi-pan-genome -g analysis_Anvio7/Pangenomic_Results/Coryne-GENOMES.db -G Cac.txt -o analysis_Anvio7/Pangenomic_Results/Cac -n PAN_Cac --use-ncbi-blast --mcl-inflation 10  --num-threads 8

anvi-pan-genome -g analysis_Anvio7/Pangenomic_Results/Coryne-GENOMES.db -G Cpr.txt -o analysis_Anvio7/Pangenomic_Results/Cpr -n PAN_Cpr --use-ncbi-blast --mcl-inflation 10  --num-threads 8

anvi-pan-genome -g analysis_Anvio7/Pangenomic_Results/Coryne-GENOMES.db -G Ctu.txt -o analysis_Anvio7/Pangenomic_Results/Ctu -n PAN_Ctu --use-ncbi-blast --mcl-inflation 10  --num-threads 8 

anvi-pan-genome -g analysis_Anvio7/Pangenomic_Results/Coryne-GENOMES.db -G Cps.txt -o analysis_Anvio7/Pangenomic_Results/Cps -n PAN_Cps --use-ncbi-blast --mcl-inflation 10  --num-threads 8
```

## Displaying the pangenomes

This will display the pangenome in your default internet browser. Anvi'o says google chrome is the most compatible with the anvi''o interactive interface. You need to pick one of the `-p` inputs for each PAN.db file:

```{bash, eval = FALSE}
anvi-display-pan -p "/analysis_Anvio7/Pangenomic_Results/Cps/PAN_Cps-PAN.db" -g "/analysis_Anvio7/Pangenomic_Results/Coryne-GENOMES.db"

anvi-display-pan -p "/analysis_Anvio7/Pangenomic_Results/Cpr/PAN_Cpr-PAN.db" -g "/analysis_Anvio7/Pangenomic_Results/Coryne-GENOMES.db"

anvi-display-pan -p "/analysis_Anvio7/Pangenomic_Results/Cac/PAN_Cac-PAN.db" -g "/analysis_Anvio7/Pangenomic_Results/Coryne-GENOMES.db"

anvi-display-pan -p "/analysis_Anvio7/Pangenomic_Results/Ctu/PAN_Ctu-PAN.db" -g "/analysis_Anvio7/Pangenomic_Results/Coryne-GENOMES.db"
```

If you are using Windows and it cannot display the interactive page then try copy and pasting either 127.0.0.1:8080 or <http://localhost:8080/> into your browser. Do not use sudo in the front of the command either.

Once it opens in the browser, you can adjust image settings and click the "DRAW" icon to view it. For help using the interactive interface see <http://merenlab.org/2016/02/27/the-anvio-interactive-interface/>

To manually reorder the layers to match phylogenomic trees per species, click and drag the names. The order names is inverse to the layers in the concentric diagram. When all the layers were re-ordered to match the phylogenomic tree we clicked the draw button and saved the svg file.

## Summarizing the updated pangenome files by bin collection

We can create tables with a summary of the pangenomes using the program `anvi-summarize`.

You will need to know the collection name to use on the `-C` flag in the `anvi-summarize` command. This is an example of how to see the collections and bins you have for a pangenome db file from anvi'o.

```{bash, eval=FALSE}
anvi-show-collections-and-bins -p "/analysis_Anvio7/Pangenomic_Results/Cac/PAN_Cpr-PAN.db" --debug 
```

```{bash, eval = FALSE}
anvi-summarize -p "/analysis_Anvio7/Pangenomic_Results/Cps/PAN_Cps-PAN.db" \
               -g "/analysis_Anvio7/Pangenomic_Results/Coryne-GENOMES.db" \
               -o "/analysis_Anvio7" \
               -C CorevsAccessory
               
anvi-summarize -p "analysis_Anvio7/Pangenomic_Results/Cpr/PAN_Cpr-PAN.db" \
               -g "/analysis_Anvio7/Pangenomic_Results/Coryne-GENOMES.db" \
               -o "/analysis_Anvio7" \
               -C CorevsAccessory
               
anvi-summarize -p "/analysis_Anvio7/Pangenomic_Results/Ctu/PAN_Ctu-PAN.db" \
               -g "/analysis_Anvio7/Pangenomic_Results/Coryne-GENOMES.db" \
               -o "/analysis_Anvio7" \
               -C CorevsAccessory
               
anvi-summarize -p "/analysis_Anvio7/Pangenomic_Results/Cac/PAN_Cac-PAN.db" \
               -g "/analysis_Anvio7/Pangenomic_Results/Coryne-GENOMES.db" \
               -o "/analysis_Anvio7" \
               -C CorevsAccessory
```

The resulting summary folders contain a `gene_clusters_summary.txt.gz` file that links each gene to gene clusters, genomes, functions, layers, and bins selected from the interface.

### *C. propinquum* pangenome

| Pangenome Matrix | \# of Gene Clusters/Pangenome | Percentage                    | Gene Cluster occurs in \# of genomes |
|------------------|------------------|------------------|-------------------|
| Core             | 1824/3108                     | `r round(100*(1824/3108),1)`% | 19                                   |
| Single Copy Core | 1715/3108                     | `r round(100*(1715/3108),1)`% | 19                                   |
| Multicopy Core   | 109/3108                      | `r round(100*(109/3108),1)`%  | 19                                   |
| Soft Core        | 39/3108                       | `r round(100*(39/3108),1)`%   | 18                                   |
| Shell            | 689/3108                      | `r round(100*(689/3108),1)`%  | 3-17                                 |
| Cloud            | 556/3108                      | `r round(100*(556/3108),1)`%  | 1-2                                  |

### *C. pseudodiphtheriticum* pangenome

| Pangenome Matrix | \# of Gene Clusters/Pangenome | Percentage                    | Gene Cluster occurs in \# of genomes |
|------------------|------------------|------------------|-------------------|
| Core             | 1714/3590                     | `r round(100*(1714/3590),1)`% | 43                                   |
| Single Copy Core | 1488/3590                     | `r round(100*(1488/3590),1)`% | 43                                   |
| Multicopy Core   | 226/3590                      | `r round(100*(226/3590),1)`%  | 43                                   |
| Soft Core        | 46/3590                       | `r round(100*(46/3590),1)`%   | 40-42                                |
| Shell            | 972/3590                      | `r round(100*(972/3590),1)`%  | 3-39                                 |
| Cloud            | 858/3590                      | `r round(100*(858/3590),1)`%  | 1-2                                  |

### *C. pseudodiphtheriticum* pangenome after dropping Cps_090104

| Pangenome Matrix | \# of Gene Clusters/Pangenome | Percentage                    | Gene Cluster occurs in \# of genomes |
|------------------|------------------|------------------|-------------------|
| Core             | 1719/3550                     | `r round(100*(1719/3550),1)`% | 42                                   |
| Single Copy Core | 1598/3550                     | `r round(100*(1598/3550),1)`% | 42                                   |
| Multicopy Core   | 121/3550                      | `r round(100*(121/3550),1)`%  | 42                                   |
| Soft Core        | 40/3550                       | `r round(100*(40/3550),1)`%   | 39-41                                |
| Shell            | 967/3550                      | `r round(100*(967/3550),1)`%  | 3-38                                 |
| Cloud            | 824/3550                      | `r round(100*(824/3550),1)`%  | 1-2                                  |

### *C. accolens* pangenome

| Pangenome Matrix | \# of Gene Clusters/Pangenome | Percentage                    | Gene Cluster occurs in \# of genomes |
|------------------|------------------|------------------|-------------------|
| Core             | 1904/3427                     | `r round(100*(1904/3427),1)`% | 34                                   |
| Single Copy Core | 1734/3427                     | `r round(100*(1734/3427),1)`% | 34                                   |
| Multicopy Core   | 170/3427                      | `r round(100*(170/3427),1)`%  | 34                                   |
| Soft Core        | 77/3427                       | `r round(100*(77/3427),1)`%   | 32-33                                |
| Shell            | 676/3427                      | `r round(100*(676/3427),1)`%  | 3-31                                 |
| Cloud            | 770/3427                      | `r round(100*(770/3427),1)`%  | 1-2                                  |

### *C. accolens* pangenome after dropping Cac_49756

| Pangenome Matrix | \# of Gene Clusters/Pangenome | Percentage                    | Gene Cluster occurs in \# of genomes |
|------------------|------------------|------------------|-------------------|
| Core             | 1941/3307                     | `r round(100*(1941/3307),1)`% | 33                                   |
| Single Copy Core | 1782/3307                     | `r round(100*(1782/3307),1)`% | 33                                   |
| Multicopy Core   | 159/3307                      | `r round(100*(159/3307),1)`%  | 33                                   |
| Soft Core        | 41/3307                       | `r round(100*(41/3307),1)`%   | 31-32                                |
| Shell            | 673/3307                      | `r round(100*(673/33307),1)`% | 3-30                                 |
| Cloud            | 652/3307                      | `r round(100*(652/3307),1)`%  | 1-2                                  |

### *C. tuberculostearicum* pangenome

| Pangenome Matrix | \# of Gene Clusters/Pangenome | Percentage                    | Gene Cluster occurs in \# of genomes |
|------------------|------------------|------------------|-------------------|
| Core             | 1915/2907                     | `r round(100*(1915/2907),1)`% | 8                                    |
| Single Copy Core | 1860/2907                     | `r round(100*(1860/2907),1)`% | 8                                    |
| Multicopy Core   | 55/2907                       | `r round(100*(55/2907),1)`%   | 8                                    |
| Soft Core        | 47/2907                       | `r round(100*(47/2907),1)`%   | 7                                    |
| Shell            | 270/2907                      | `r round(100*(270/2907),1)`%  | 3-6                                  |
| Cloud            | 671/2907                      | `r round(100*(671/2907),1)`%  | 1-2                                  |

## Finding GC Content

Find GC Content for *C. propinquum*:

```{bash, eval=FALSE}
anvi-summarize -g /analysis_Anvio7/Pangenomic_Results/Coryne-GENOMES.db -p /analysis_Anvio7/Pangenomic_Results/Cpr/PAN_Cpr-PAN.db -C CorevsAccessory
```

Find GC Content for *C. pseudodiphtheriticum*:

```{bash, eval=FALSE}
anvi-summarize -g /analysis_Anvio7/Pangenomic_Results/Coryne-GENOMES.db -p /analysis_Anvio7/Pangenomic_Results/Cps/PAN_Cps-PAN.db -C CorevsAccessory
```

Find GC Content for *C. accolens*:

```{bash, eval=FALSE}
anvi-summarize -g /analysis_Anvio7/Pangenomic_Results/Coryne-GENOMES.db -p /analysis_Anvio7/Pangenomic_Results/Cac/PAN_Cac-PAN.db -C CorevsAccessory
```

Find GC Content for *C. tuberculostearicum*:

```{bash, eval=FALSE}
anvi-summarize -g /analysis_Anvio7/Pangenomic_Results/Coryne-GENOMES.db -p /analysis_Anvio7/Pangenomic_Results/Ctu/PAN_Ctu-PAN.db -C CorevsAccessory
```

# KEGG Enrichment Analysis

We then used a workflow from anvi'o (<https://merenlab.org/tutorials/infant-gut/#chapter-v-metabolism-prediction>) that allowed us to estimate the completeness of KEGG modules.

## Four *Corynebacterium* species.

We first did this in each of the four *Corynebacterium* pangenomes and in the combined pangenome for the 104 strains.

First we used the same text document that was a list of the genome files excluding *C. macingleyi*. This was used as input for the estimate metabolism step. It contains two columns, the first column, labeled "Name", is the name of the strain genome, the second column, labeled "Contigs_db_path", is the file name and location of the genome database file.

```{bash, eval = FALSE}
cd ./analysis_Anvio7

anvi-estimate-metabolism -e All_104_CorynesAnvio.txt -O Coryne_All --matrix-format

anvi-matrix-to-newick Coryne_All-completeness-MATRIX.txt
```

Dry run to get the profile db:

```{bash, eval = FALSE}
anvi-interactive -d Coryne_All-completeness-MATRIX.txt -p Coryne_All_metabolism_PROFILE.db --manual-mode --dry-run
```

Run :

```{bash, eval = FALSE}
anvi-interactive --manual-mode -d Coryne_All-completeness-MATRIX.txt -t Coryne_All-completeness-MATRIX.txt.newick -p Coryne_All_metabolism_PROFILE.db --title "Coryne_All Metabolism Heatmap"

anvi-export-state -s default -p Coryne_All_metabolism_PROFILE.db -o Coryne_All_metabolism_state.json

export ANVIO_MODULES_DB=`python -c "import anvio; import os; print(os.path.join(os.path.dirname(anvio.__file__), 'data/misc/KEGG/MODULES.db'))"`
echo -e > dpig+Coryne-No_Mac_modules_info.txt
sqlite3 $ANVIO_MODULES_DB "select module, data_value from kegg_modules where data_name='CLASS'" | sed 's/; /|/g' | tr '|' '\t' >> module_class.txt
sqlite3 $ANVIO_MODULES_DB "select module,data_value from kegg_modules where data_name='NAME'" | tr '|' '\t' > module_names.txt
paste module_class.txt <(cut -f 2 module_names.txt ) >> Coryne_All_modules_info.txt
rm module_names.txt module_class.txt

anvi-import-misc-data Coryne_All_modules_info.txt -p Coryne_All_metabolism_PROFILE.db -t items
```

To painlessly make your interactive visualizations look like the ones in the screenshots, you can import the following state file into your profile database:

```{bash, eval = FALSE}
anvi-import-state -p Coryne_All_metabolism_PROFILE.db -n default -s Coryne_All_metabolism_state.json
```

And then we can run the visualization command yet again to see the heatmap with labels:

```{bash, eval = FALSE}
anvi-interactive --manual-mode -d Coryne_All-completeness-MATRIX.txt -t Coryne_All-completeness-MATRIX.txt.newick -p Coryne_All_metabolism_PROFILE.db --title "Coryne_All Metabolism Heatmap"

anvi-estimate-metabolism -e All_104_CorynesAnvio.txt -O Coryne_All_metabolism --kegg-output-modes modules,kofam_hits
```

Need to use compute-metabolic-enrichment in anvio 7.1:

```{bash, eval = FALSE}
anvi-compute-metabolic-enrichment -M Coryne_All_metabolism_modules.txt -G Coryne_All_species_groups.txt -o Coryne_All_enriched_modules.txt

anvi-compute-metabolic-enrichment -M Coryne_All_metabolism_modules.txt -G Coryne_All_collection_groups.txt -o Coryne_All_enriched_coll_modules.txt
```

## *Corynebacterium* species vs *Dolosigranulum pigrum*

For this run, we follow the same protocol except we included 27 *D. pigrum* genomes and included them in the genome list. We will compare the five species.

```{bash, eval = FALSE}
cd ./Documents/aqr_metabol_comp

anvi-estimate-metabolism -e genome_list_Coryne_Dpi.txt -O Coryne_Dpi --matrix-format

anvi-matrix-to-newick Coryne_Dpi-completeness-MATRIX.txt
```

Dry run to get the profile db:

```{bash, eval = FALSE}
anvi-interactive -d Coryne_Dpi-completeness-MATRIX.txt -p Coryne_Dpi_metabolism_PROFILE.db --manual-mode --dry-run
```

Run :

```{bash, eval = FALSE}
anvi-interactive --manual-mode -d Coryne_Dpi-completeness-MATRIX.txt -t Coryne_Dpi-completeness-MATRIX.txt.newick -p Coryne_Dpi_metabolism_PROFILE.db --title "Coryne_Dpi Metabolism Heatmap"

anvi-export-state -s default -p Coryne_Dpi_metabolism_PROFILE.db -o Coryne_Dpi_metabolism_state.json

export ANVIO_MODULES_DB=`python -c "import anvio; import os; print(os.path.join(os.path.dirname(anvio.__file__), 'data/misc/KEGG/MODULES.db'))"`
echo -e > dpig+Coryne-No_Mac_modules_info.txt
sqlite3 $ANVIO_MODULES_DB "select module, data_value from kegg_modules where data_name='CLASS'" | sed 's/; /|/g' | tr '|' '\t' >> module_class.txt
sqlite3 $ANVIO_MODULES_DB "select module,data_value from kegg_modules where data_name='NAME'" | tr '|' '\t' > module_names.txt
paste module_class.txt <(cut -f 2 module_names.txt ) >> Coryne_Dpi_modules_info.txt
rm module_names.txt module_class.txt

anvi-import-misc-data Coryne_Dpi_modules_info.txt -p Coryne_Dpi_metabolism_PROFILE.db -t items
```

To painlessly make your interactive visualizations look like the ones in the screenshots, you can import the following state file into your profile database:

```{bash, eval = FALSE}
anvi-import-state -p Coryne_Dpi_metabolism_PROFILE.db -n default -s Coryne_Dpi_metabolism_state.json
```

And then we can run the visualization command yet again to see the heatmap with labels:

```{bash, eval = FALSE}
anvi-interactive --manual-mode -d Coryne_Dpi-completeness-MATRIX.txt -t Coryne_Dpi-completeness-MATRIX.txt.newick -p Coryne_Dpi_metabolism_PROFILE.db --title "Coryne_Dpi Metabolism Heatmap"

anvi-estimate-metabolism -e genome_list_Coryne_Dpi.txt -O Coryne_Dpi_metabolism --kegg-output-modes modules,kofam_hits
```

Need to use compute-metabolic-enrichment in anvio 7.1:

```{bash, eval = FALSE}
anvi-compute-metabolic-enrichment -M Coryne_Dpi_metabolism_modules.txt -G Coryne_Dpi_species_groups.txt -o Coryne_Dpi_enriched_modules.txt
```

## Intra-Species Geographic Comparison

We re-ran the same code but within each of the four *Corynebacterium* species. Instead of comparing metabolic capabilities of species we compared the geographic origin of the each genome: North America or Botswana. We made new categorization text files for each species accordingly as well as text files for each of the species groupings for their genome file locations.

### *C. propinquum*

```{bash, eval = FALSE}
cd ./Documents/aqr_metabol_comp

anvi-estimate-metabolism -e genome_list_Cpr.txt -O Cpr --matrix-format

anvi-matrix-to-newick Cpr-completeness-MATRIX.txt
```

Dry run to get the profile db:

```{bash, eval = FALSE}
anvi-interactive -d Cpr-completeness-MATRIX.txt -p Cpr_metabolism_PROFILE.db --manual-mode --dry-run
```

Run :

```{bash, eval = FALSE}
anvi-interactive --manual-mode -d Cpr-completeness-MATRIX.txt -t Cpr-completeness-MATRIX.txt.newick -p Cpr_metabolism_PROFILE.db --title "Cpr Metabolism Heatmap"

anvi-export-state -s default -p Cpr_metabolism_PROFILE.db -o Cpr_metabolism_state.json

export ANVIO_MODULES_DB=`python -c "import anvio; import os; print(os.path.join(os.path.dirname(anvio.__file__), 'data/misc/KEGG/MODULES.db'))"`
echo -e > Cpr_modules_info.txt
sqlite3 $ANVIO_MODULES_DB "select module, data_value from kegg_modules where data_name='CLASS'" | sed 's/; /|/g' | tr '|' '\t' >> module_class.txt
sqlite3 $ANVIO_MODULES_DB "select module,data_value from kegg_modules where data_name='NAME'" | tr '|' '\t' > module_names.txt
paste module_class.txt <(cut -f 2 module_names.txt ) >> Cpr_modules_info.txt
rm module_names.txt module_class.txt

anvi-import-misc-data Cpr_modules_info.txt -p Cpr_metabolism_PROFILE.db -t items
```

To painlessly make your interactive visualizations look like the ones in the screenshots, you can import the following state file into your profile database:

```{bash, eval = FALSE}
anvi-import-state -p Cpr_metabolism_PROFILE.db -n default -s Cpr_metabolism_state.json
```

And then we can run the visualization command yet again to see the heatmap with labels:

```{bash, eval = FALSE}
anvi-interactive --manual-mode -d Cpr-completeness-MATRIX.txt -t Cpr-completeness-MATRIX.txt.newick -p Cpr_metabolism_PROFILE.db --title "Cpr Metabolism Heatmap"

anvi-estimate-metabolism -e genome_list_Cpr.txt -O Cpr_metabolism --kegg-output-modes modules,kofam_hits
```

```{bash, eval = FALSE}
anvi-compute-metabolic-enrichment -M Cpr_metabolism_modules.txt -G Cpr_geo_groups.txt -o Cpr_enriched_modules.txt
```

### *C. pseudodiphtheriticum*

```{bash, eval = FALSE}
cd ./Documents/aqr_metabol_comp
-7.1
anvi-estimate-metabolism -e genome_list_Cps.txt -O Cps --matrix-format
anvi-matrix-to-newick Cps-completeness-MATRIX.txt
```

Dry run to get the profile db:

```{bash, eval = FALSE}
anvi-interactive -d Cps-completeness-MATRIX.txt -p Cps_metabolism_PROFILE.db --manual-mode --dry-run
```

Run :

```{bash, eval = FALSE}
anvi-interactive --manual-mode -d Cps-completeness-MATRIX.txt -t Cps-completeness-MATRIX.txt.newick -p Cps_metabolism_PROFILE.db --title "Cps Metabolism Heatmap"

anvi-export-state -s default -p Cps_metabolism_PROFILE.db -o Cps_metabolism_state.json

export ANVIO_MODULES_DB=`python -c "import anvio; import os; print(os.path.join(os.path.dirname(anvio.__file__), 'data/misc/KEGG/MODULES.db'))"`
echo -e > Cps_modules_info.txt
sqlite3 $ANVIO_MODULES_DB "select module, data_value from kegg_modules where data_name='CLASS'" | sed 's/; /|/g' | tr '|' '\t' >> module_class.txt
sqlite3 $ANVIO_MODULES_DB "select module,data_value from kegg_modules where data_name='NAME'" | tr '|' '\t' > module_names.txt
paste module_class.txt <(cut -f 2 module_names.txt ) >> Cps_modules_info.txt
rm module_names.txt module_class.txt

anvi-import-misc-data Cps_modules_info.txt -p Cps_metabolism_PROFILE.db -t items
```

To painlessly make your interactive visualizations look like the ones in the screenshots, you can import the following state file into your profile database:

```{bash, eval = FALSE}
anvi-import-state -p Cps_metabolism_PROFILE.db -n default -s Cps_metabolism_state.json
```

And then we can run the visualization command yet again to see the heatmap with labels:

```{bash, eval = FALSE}
anvi-interactive --manual-mode -d Cps-completeness-MATRIX.txt -t Cps-completeness-MATRIX.txt.newick -p Cps_metabolism_PROFILE.db --title "Cps Metabolism Heatmap"

anvi-estimate-metabolism -e genome_list_Cps.txt -O Cps_metabolism --kegg-output-modes modules,kofam_hits
```

```{bash, eval = FALSE}
anvi-compute-metabolic-enrichment -M Cps_metabolism_modules.txt -G Cps_geo_groups.txt -o Cps_enriched_modules.txt
```

### *C. accolens*

```{bash, eval = FALSE}
cd ./Documents/aqr_metabol_comp
anvi-estimate-metabolism -e genome_list_Cac.txt -O Cac --matrix-format

anvi-matrix-to-newick Cac-completeness-MATRIX.txt
```

Dry run to get the profile db:

```{bash, eval = FALSE}
anvi-interactive -d Cac-completeness-MATRIX.txt -p Cac_metabolism_PROFILE.db --manual-mode --dry-run
```

Run:

```{bash, eval = FALSE}
anvi-interactive --manual-mode -d Cac-completeness-MATRIX.txt -t Cac-completeness-MATRIX.txt.newick -p Cac_metabolism_PROFILE.db --title "Cac Metabolism Heatmap"

anvi-export-state -s default -p Cac_metabolism_PROFILE.db -o Cac_metabolism_state.json

export ANVIO_MODULES_DB=`python -c "import anvio; import os; print(os.path.join(os.path.dirname(anvio.__file__), 'data/misc/KEGG/MODULES.db'))"`
echo -e > Cac_modules_info.txt
sqlite3 $ANVIO_MODULES_DB "select module, data_value from kegg_modules where data_name='CLASS'" | sed 's/; /|/g' | tr '|' '\t' >> module_class.txt
sqlite3 $ANVIO_MODULES_DB "select module,data_value from kegg_modules where data_name='NAME'" | tr '|' '\t' > module_names.txt
paste module_class.txt <(cut -f 2 module_names.txt ) >> Cac_modules_info.txt
rm module_names.txt module_class.txt

anvi-import-misc-data Cac_modules_info.txt -p Cac_metabolism_PROFILE.db -t items
```

To painlessly make your interactive visualizations look like the ones in the screenshots, you can import the following state file into your profile database:

```{bash, eval = FALSE}
anvi-import-state -p Cac_metabolism_PROFILE.db -n default -s Cac_metabolism_state.json
```

And then we can run the visualization command yet again to see the heatmap with labels:

```{bash, eval = FALSE}
anvi-interactive --manual-mode -d Cac-completeness-MATRIX.txt -t Cac-completeness-MATRIX.txt.newick -p Cac_metabolism_PROFILE.db --title "Cac Metabolism Heatmap"

anvi-estimate-metabolism -e genome_list_Cac.txt -O Cac_metabolism --kegg-output-modes modules,kofam_hits
```

```{bash, eval = FALSE}
anvi-compute-metabolic-enrichment -M Cac_metabolism_modules.txt -G Cac_geo_groups.txt -o Cac_enriched_modules.txt
```

### *C. tuberculostearicum*

```{bash, eval = FALSE}
cd ./Documents/aqr_metabol_comp
anvi-estimate-metabolism -e genome_list_Ctu.txt -O Ctu --matrix-format
anvi-matrix-to-newick Ctu-completeness-MATRIX.txt
```

Dry run to get the profile db:

```{bash, eval = FALSE}
anvi-interactive -d Ctu-completeness-MATRIX.txt -p Ctu_metabolism_PROFILE.db --manual-mode --dry-run
anvi-interactive --manual-mode -d Ctu-completeness-MATRIX.txt -t Ctu-completeness-MATRIX.txt.newick -p Ctu_metabolism_PROFILE.db --title "Ctu Metabolism Heatmap"

anvi-export-state -s default -p Ctu_metabolism_PROFILE.db -o Ctu_metabolism_state.json

export ANVIO_MODULES_DB=`python -c "import anvio; import os; print(os.path.join(os.path.dirname(anvio.__file__), 'data/misc/KEGG/MODULES.db'))"`
echo -e > Ctu_modules_info.txt
sqlite3 $ANVIO_MODULES_DB "select module, data_value from kegg_modules where data_name='CLASS'" | sed 's/; /|/g' | tr '|' '\t' >> module_class.txt
sqlite3 $ANVIO_MODULES_DB "select module,data_value from kegg_modules where data_name='NAME'" | tr '|' '\t' > module_names.txt
paste module_class.txt <(cut -f 2 module_names.txt ) >> Ctu_modules_info.txt
rm module_names.txt module_class.txt

anvi-import-misc-data Ctu_modules_info.txt -p Ctu_metabolism_PROFILE.db -t items
```

To painlessly make your interactive visualizations look like the ones in the screenshots, you can import the following state file into your profile database:

```{bash, eval = FALSE}
anvi-import-state -p Ctu_metabolism_PROFILE.db -n default -s Ctu_metabolism_state.json
```

And then we can run the visualization command yet again to see the heatmap with labels:

```{bash, eval = FALSE}
anvi-interactive --manual-mode -d Ctu-completeness-MATRIX.txt -t Ctu-completeness-MATRIX.txt.newick -p Ctu_metabolism_PROFILE.db --title "Ctu Metabolism Heatmap"

anvi-estimate-metabolism -e genome_list_Ctu.txt -O Ctu_metabolism --kegg-output-modes modules,kofam_hits
```

```{bash, eval = FALSE}
anvi-compute-metabolic-enrichment -M Ctu_metabolism_modules.txt -G Ctu_geo_groups.txt -o Ctu_enriched_modules.txt
```

## Generate the Distribution Plot in Figure 5

We took the text file outputs with metabolism_modules from the previous steps on individual species runs, including one for D. pigrum, and converted them into xlsx files before reading them into R.

```{R, eval = FALSE}
# read in libraries
library(readxl)
library(ggplot2)
 
# read in module files of each species independent KEGG run after saving text file outputs as xlsx files
cac <- read_excel("Cac_metabolism_modules.xlsx")
cpr <- read_excel("Cpr_metabolism_modules.xlsx")
cps <- read_excel("Cps_metabolism_modules.xlsx")
ctu <- read_excel("Ctu_metabolism_modules.xlsx")
dpi <- read_excel("Dpi_metabolism_modules.xlsx")

# remove warning column
cac_2 <- subset(cac, select = -c(db_name, warnings))
cpr_2 <- subset(cpr, select = -c(db_name, warnings))
cps_2 <- subset(cps, select = -c(db_name, warnings))
ctu_2 <- subset(ctu, select = -c(db_name, warnings))
dpi_2 <- subset(dpi, select = -c(db_name))

# filter out to complete module completion > 0.75
cac_filt <- subset(cac_2, module_completeness >= 0.75)
cpr_filt <- subset(cpr_2, module_completeness >= 0.75)
cps_filt <- subset(cps_2, module_completeness >= 0.75)
ctu_filt <- subset(ctu_2, module_completeness >= 0.75)
dpi_filt <- subset(dpi_2, module_completeness >= 0.75)

# add factor for species
cac_filt$species <- "C. accolens"
cpr_filt$species <- "C. propinquum"
cps_filt$species <- "C. pseudodiphtheriticum"
ctu_filt$species <- "C. tuberculostearicum"
dpi_filt$species <- "D. pigrum"

# combine into 1 data array
combo <- rbind(cac_filt, cpr_filt, cps_filt, ctu_filt, dpi_filt)

# Generate a bar graph of distribution 

x axis is species and y axis is # of complete modules to make up a module category
# filter out repeated modules here
cac_filt2 <- unique(subset(cac_filt, select = c(kegg_module, module_name, module_class, module_category, module_subcategory, species)))
cpr_filt2 <- unique(subset(cpr_filt, select = c(kegg_module, module_name, module_class, module_category, module_subcategory, species)))
cps_filt2 <- unique(subset(cps_filt, select = c(kegg_module, module_name, module_class, module_category, module_subcategory, species)))
ctu_filt2 <- unique(subset(ctu_filt, select = c(kegg_module, module_name, module_class, module_category, module_subcategory, species)))
dpi_filt2 <- unique(subset(dpi_filt, select = c(kegg_module, module_name, module_class, module_category, module_subcategory, species)))

combo2 <- rbind(cpr_filt2, cps_filt2, cac_filt2, ctu_filt2, dpi_filt2)
library("writexl")
write_xlsx(combo2, "combo2.xlsx")

plot3 = ggplot(combo2) + geom_bar(aes(fill = module_category, x = factor(species, level = c("C. propinquum", "C. pseudodiphtheriticum", "C. accolens", "C. tuberculostearicum", "D. pigrum"))))+ scale_fill_brewer(palette = "Set2") + theme_bw() + theme(panel.border = element_blank(), panel.grid.major = element_blank(), panel.grid.minor = element_blank(), axis.line = element_line(colour = "black")) + theme(legend.position = "bottom") + theme(aspect.ratio = 1) 
```

<img src="images/Department-of-Molecular-Virology-&amp;-Microbiologyy-Horz-GRAY.png" align="left" width="200" height="90"/>

------------------------------------------------------------------------

------------------------------------------------------------------------

------------------------------------------------------------------------

------------------------------------------------------------------------

<center>

<h3>[REFERENCES]{.underline}</h3>

</center>
