---
title: "SupplementalMethods_anvi'o"
author:
  - Tommy Tran, (KLemonLab) Tommy.Tran@bcm.edu
  - Ari Roberts, (KLemonLab) Ari.Roberts@bcm.edu
  - Isabel Escapa, isabel.fernandez.escapa@gmail.com
output:
  rmdformats::robobook:
    use_bookdown: true
    code_folding: show
  github_document:
    toc: true
    toc_depth: 4
bibliography: referencesAnvio.bib
---

```{r klippy, echo=FALSE, include=TRUE}
klippy::klippy(c('r', 'bash'))
```

# anvi'o v7

This notebook contains analyses using anvi'o version 7.1.2 [@Eren2021]. If you want to know more in-depth information about anvi'o you can go to their official online site here: <https://anvio.org/>

# Import functional annotations

## COG annotation

First you need to download and set up the NCBI's Clusters of Orthologous Groups database using `anvi-setup-ncbi-cogs`: this is something you will do only once.

Now the `anvi-run-ncbi-cogs` command can be used to annotate the .db genome files against the NCBI's COGs database [@Tatusov2001]: It uses DIAMOND [@Buchfink2015] "fast" as default, but it is recommended to use the "sensitive" option. This will take considerably much longer. Blastp can also be used instead of DIAMOND.

```{bash, eval = FALSE}
#conda activate anvio-7.1.2
path_f="analysis_Anvio7_1/Contigs_db"

for file in $path_f/*.db; do
    anvi-run-ncbi-cogs -T 8 --sensitive -c $file;
done
```

## KEGG annotation

First you need to download and set up the KEGG KOfam database using `anvi-setup-kegg-kofams`: this is something you will do only once.

The program `anvi-run-kegg-kofams` annotates genes in a given anvi'o contigs database with KEGG Orthology (KO) numbers via hits to the KEGG KOfam database [@Kanehisa2000].

```{bash, eval = FALSE}
#conda activate anvio-7.1.2
path_f="analysis_Anvio7_1/Contigs_db"

for file in $path_f/*.db; do
    anvi-run-kegg-kofams -T 8 --log-bitscores -c $file;
done
```

The option `--log-bitscores` records the bit scores values. They were saved to the main folder for the repo, so I move them to a folder:

```{bash, eval = FALSE}
mkdir "analysis_Anvio7_1/KEGGbitscores"
mv *_bitscores.txt analysis_Anvio7_1/KEGGbitscores/
```

## PFAM annotation

First you need to download and set up the PFAM database using `anvi-setup-pfams`: this is something you will do only once.

The program `anvi-run-kegg-pfam` annotates genes in a given anvi'o contigs database with the PFAM database [@Mistry2021].

```{bash, eval = FALSE}
#conda activate anvio-7.1.2
path_f="analysis_Anvio7_1/Contigs_db"

for file in $path_f/*.db; do
    anvi-run-pfams -T 8 -c $file;
done
```

## Running HMMs

When you run the following command the occurrence of bacterial single-copy genes across your contigs is all added to your contigs database [@Eddy2011].

```{bash, eval = FALSE}
#conda activate anvio-7.1.2
path_f="analysis_Anvio7_1/Contigs_db"

for file in $path_f/*.db; do
    anvi-run-hmms -T 8 -c $file --just-do-it;
done
```

## Annotation Summaries

This generates basic stats on the annotations:

```{bash, eval = FALSE}
#conda activate anvio-7.1.2
mkdir -p "analysis_Anvio7_1/Contigs_stats"

path_f="analysis_Anvio7_1/Contigs_db"
path_o="analysis_Anvio7_1/Contigs_stats"

for file in $path_f/*.db; do
    FILENAME=`basename ${file%.*}`
    anvi-display-contigs-stats $file --report-as-text -o $path_o/stats_$FILENAME.txt;
done
```

This provides basic info on the contigs databases:

```{bash, eval = FALSE}
#conda activate anvio-7.1.2
path_f="analysis_Anvio7_1/Contigs_db"

for file in $path_f/*.db; do
    anvi-db-info $file;
done
```

Terminal output saved to `analysis_Anvio7_1/anvi-db-info_log.txt`.

# PANGENOME ANALYSIS

## Generating an anvi'o genomes storage

The program `anvi-gen-genomes-storage` requires you to provide names and locations of the genomes to be included in the genome storage database in a .txt file. We created the All_107_CorynesAnvio.txt file with the renamed file names and paths based on the info on `GENOMES/METADATA_107taxa.csv`. This file includes the name anvi'o will use internally to identify each genome. I added the prefix "RefSeq\_" to the reference genomes because some of them start with a digit and anvi'o only allow names that start with an ASCII letter.

We create 1 database with the 107 genomes that can be used to run the pangenomic analysis with all the strains at once or by selecting subsets of strains (i.e. by species, or by grouping related species together).

```{bash, eval = FALSE}
#conda activate anvio-7.1
mkdir -p "analysis_Anvio7_1/Pangenomic_Results"

anvi-gen-genomes-storage -e "analysis_Anvio7_1/All_107_CorynesAnvio.txt" -o "analysis_Anvio7_1/Pangenomic_Results/Coryne-GENOMES.db" --gene-caller Prodigal
```

This provides basic info on the generated genomes storage database:

```{bash, eval = FALSE}
anvi-db-info "analysis_Anvio7_1/Pangenomic_Results/Coryne-GENOMES.db"
```

# Make Species Genome Lists

We had to make new anvi'o environment with python 3.7 since the step below would not work unless we did that. This could be related to the updates with anvio 7.1 conda create -y --name anvio-7.1.2 python=3.7

# Running the pangenomic analsyses

```{bash, eval = FALSE}
#conda activate anvio-7.1
anvi-pan-genome -g analysis_Anvio7_1/Pangenomic_Results/Coryne-GENOMES.db -G Cacc.txt -o analysis_Anvio7_1/Pangenomic_Results/Cacc -n PAN_Cacc --use-ncbi-blast --mcl-inflation 10  --num-threads 8

anvi-pan-genome -g analysis_Anvio7_1/Pangenomic_Results/Coryne-GENOMES.db -G Cpro.txt -o analysis_Anvio7_1/Pangenomic_Results/Cpro -n PAN_Cpro --use-ncbi-blast --mcl-inflation 10  --num-threads 8

anvi-pan-genome -g analysis_Anvio7_1/Pangenomic_Results/Coryne-GENOMES.db -G Ctub.txt -o analysis_Anvio7_1/Pangenomic_Results/Ctub -n PAN_Ctub --use-ncbi-blast --mcl-inflation 10  --num-threads 8 

anvi-pan-genome -g analysis_Anvio7_1/Pangenomic_Results/Coryne-GENOMES.db -G Cpsd.txt -o analysis_Anvio7_1/Pangenomic_Results/Cpsd -n PAN_Cpsd --use-ncbi-blast --mcl-inflation 10  --num-threads 8
```

## Displaying the pangenomes

This will display the pangenome in your default internet browser. Anvi'o says google chrome is the most compatible with the anvi''o interactive interface. You need to pick one of the `-p` inputs for each PAN.db file:

```{bash, eval = FALSE}
#conda activate anvio-7
anvi-display-pan -p "/Users/klemonlab/Documents/CorPGA_rerun_aqr/analysis_Anvio7_1/Pangenomic_Results/Cpsd/PAN_Cpsd-PAN.db" -g "/Users/klemonlab/Documents/CorPGA_rerun_aqr/analysis_Anvio7_1/Pangenomic_Results/Coryne-GENOMES.db"

anvi-display-pan -p "/Users/klemonlab/Documents/CorPGA_rerun_aqr/analysis_Anvio7_1/Pangenomic_Results/Cpro/PAN_Cpro-PAN.db" -g "/Users/klemonlab/Documents/CorPGA_rerun_aqr/analysis_Anvio7_1/Pangenomic_Results/Coryne-GENOMES.db"

anvi-display-pan -p "/Users/klemonlab/Documents/CorPGA_rerun_aqr/analysis_Anvio7_1/Pangenomic_Results/Cacc/PAN_Cacc-PAN.db" -g "/Users/klemonlab/Documents/CorPGA_rerun_aqr/analysis_Anvio7_1/Pangenomic_Results/Coryne-GENOMES.db"

anvi-display-pan -p "/Users/klemonlab/Documents/CorPGA_rerun_aqr/analysis_Anvio7_1/Pangenomic_Results/Ctub/PAN_Ctub-PAN.db" -g "/Users/klemonlab/Documents/CorPGA_rerun_aqr/analysis_Anvio7_1/Pangenomic_Results/Coryne-GENOMES.db"
```

If you are using Windows and it cannot display the interactive page then try copy and pasting either 127.0.0.1:8080 or <http://localhost:8080/> into your browser. Do not use sudo in the front of the command either.

Once it opens in the browser, you can adjust image settings and click the "DRAW" icon to view it. For help using the interactive interface see <http://merenlab.org/2016/02/27/the-anvio-interactive-interface/>

## Summarizing the updated pangenome files by bin collection

We can create tables with a summary of the pangenomes using the program `anvi-summarize`:

```{bash, eval = FALSE}
#conda activate anvio-7
anvi-summarize -p "/Users/klemonlab/Documents/CorPGA_rerun_aqr/analysis_Anvio7_1/Pangenomic_Results/Cpsd/PAN_Cpsd-PAN.db" \
               -g "/Users/klemonlab/Documents/CorPGA_rerun_aqr/analysis_Anvio7_1/Pangenomic_Results/Coryne-GENOMES.db" \
               -o "/Users/klemonlab/Downloads/Cps_PAN_SUMMARY" \
               -C CorevsAccessory
               
anvi-summarize -p "/Users/klemonlab/Documents/CorPGA_rerun_aqr/analysis_Anvio7_1/Pangenomic_Results/Cpro/PAN_Cpro-PAN.db" \
               -g "/Users/klemonlab/Documents/CorPGA_rerun_aqr/analysis_Anvio7_1/Pangenomic_Results/Coryne-GENOMES.db" \
               -o "/Users/klemonlab/Downloads/Cpr_PAN_SUMMARY" \
               -C CorevsAccessory
               
anvi-summarize -p "/Users/klemonlab/Documents/CorPGA_rerun_aqr/analysis_Anvio7_1/Pangenomic_Results/Ctub/PAN_Ctub-PAN.db" \
               -g "/Users/klemonlab/Documents/CorPGA_rerun_aqr/analysis_Anvio7_1/Pangenomic_Results/Coryne-GENOMES.db" \
               -o "/Users/klemonlab/Downloads/Ctu_PAN_SUMMARY" \
               -C CorevsAccessory
               
anvi-summarize -p "/Users/klemonlab/Documents/CorPGA_rerun_aqr/analysis_Anvio7_1/Pangenomic_Results/Cacc/PAN_Cacc-PAN.db" \
               -g "/Users/klemonlab/Documents/CorPGA_rerun_aqr/analysis_Anvio7_1/Pangenomic_Results/Coryne-GENOMES.db" \
               -o "/Users/klemonlab/Downloads/Cac_PAN_SUMMARY" \
               -C CorevsAccessory
```

The resulting summary folders contain a `gene_clusters_summary.txt.gz` file that links each gene to gene clusters, genomes, functions, layers, and bins selected from the interface.

For consistency with other parts of the analysis the colors used for each species are:

-   "red" or "#FF0000" for Cpsd
-   "darkorange" or "#FF8C00" for Cpro
-   "darkblue" or "#0000ff" for Ctub
-   "darkviolet" or "#9400D3" for Cacc
-   "deeppink" or "#FF1493" for Cmac

## Manually Reorder the Layers to Match Phylogenomic Trees per Species

Click and drag the names. The order names is inverse to the layers in the concentric diagram. When all the layers were re-ordered to match the phylogenomic tree we clicked the draw button and saved the svg file.

Darker Hex color assignments for PANGGOLiN

persistant - 5e0000

shell - 7400a6

cloud - 054ca8

Hex color assignments for GET_HOMOGOLOUES:

Core - d18f24

SC_Core - a30303

Soft_Core - fc4635

Shell - c64dfa

Cloud - 358bfc

### *C. propinquum* pangenome

| Pangenome Matrix | \# of Gene Clusters/Pangenome | Percentage                    | Gene Cluster occurs in \# of genomes |
|-----------------|-----------------|-----------------|---------------------|
| Core             | 1824/3108                     | `r round(100*(1824/3108),1)`% | 19                                   |
| Single Copy Core | 1715/3108                     | `r round(100*(1715/3108),1)`% | 19                                   |
| Multicopy Core   | 109/3108                      | `r round(100*(109/3108),1)`%  | 19                                   |
| Soft Core        | 39/3108                       | `r round(100*(39/3108),1)`%   | 18                                   |
| Shell            | 689/3108                      | `r round(100*(689/3108),1)`%  | 3-17                                 |
| Cloud            | 556/3108                      | `r round(100*(556/3108),1)`%  | 1-2                                  |

### *C. pseudodiphtheriticum* pangenome

| Pangenome Matrix | \# of Gene Clusters/Pangenome | Percentage                    | Gene Cluster occurs in \# of genomes |
|-----------------|-----------------|-----------------|---------------------|
| Core             | 1714/3590                     | `r round(100*(1714/3590),1)`% | 43                                   |
| Single Copy Core | 1488/3590                     | `r round(100*(1488/3590),1)`% | 43                                   |
| Multicopy Core   | 226/3590                      | `r round(100*(226/3590),1)`%  | 43                                   |
| Soft Core        | 46/3590                       | `r round(100*(46/3590),1)`%   | 40-42                                |
| Shell            | 972/3590                      | `r round(100*(972/3590),1)`%  | 3-39                                 |
| Cloud            | 858/3590                      | `r round(100*(858/3590),1)`%  | 1-2                                  |

### *C. accolens* pangenome

| Pangenome Matrix | \# of Gene Clusters/Pangenome | Percentage                    | Gene Cluster occurs in \# of genomes |
|-----------------|-----------------|-----------------|---------------------|
| Core             | 1904/3427                     | `r round(100*(1904/3427),1)`% | 34                                   |
| Single Copy Core | 1734/3427                     | `r round(100*(1734/3427),1)`% | 34                                   |
| Multicopy Core   | 170/3427                      | `r round(100*(170/3427),1)`%  | 34                                   |
| Soft Core        | 77/3427                       | `r round(100*(77/3427),1)`%   | 32-33                                |
| Shell            | 676/3427                      | `r round(100*(676/3427),1)`%  | 3-31                                 |
| Cloud            | 770/3427                      | `r round(100*(770/3427),1)`%  | 1-2                                  |

### *C. tuberculostearicum* pangenome

| Pangenome Matrix | \# of Gene Clusters/Pangenome | Percentage                    | Gene Cluster occurs in \# of genomes |
|-----------------|-----------------|-----------------|---------------------|
| Core             | 1915/2907                     | `r round(100*(1915/2907),1)`% | 8                                    |
| Single Copy Core | 1860/2907                     | `r round(100*(1860/2907),1)`% | 8                                    |
| Multicopy Core   | 55/2907                       | `r round(100*(55/2907),1)`%   | 8                                    |
| Soft Core        | 47/2907                       | `r round(100*(47/2907),1)`%   | 7                                    |
| Shell            | 270/2907                      | `r round(100*(270/2907),1)`%  | 3-6                                  |
| Cloud            | 671/2907                      | `r round(100*(671/2907),1)`%  | 1-2                                  |

# Finding GC Content

You will need to know the collection to assign for the `-C` flag in the `anvi-summarize` command. This is an example of how to see the collections and bins you have for a pangenome db file from anvi'o.

```{bash, eval=FALSE}
anvi-show-collections-and-bins -p "/Users/klemonlab/Documents/CorPGA_rerun_aqr/analysis_Anvio7_1/Pangenomic_Results/Cacc/PAN_Cpro-PAN.db" --debug 
```

Find GC Content for *C. propinquum*:

```{bash, eval=FALSE}
# conda activate Anvio
anvi-summarize -g /Users/klemonlab/Documents/CorPGA_rerun_aqr/analysis_Anvio7_1/Pangenomic_Results/Coryne-GENOMES.db -p /Users/klemonlab/Documents/CorPGA_rerun_aqr/analysis_Anvio7_1/Pangenomic_Results/Cpro/PAN_Cpro-PAN.db -C CorevsAccessory
```

Find GC Content for *C. pseudodiphtheriticum*:

```{bash, eval=FALSE}
# conda activate Anvio
anvi-summarize -g /Users/klemonlab/Documents/CorPGA_rerun_aqr/analysis_Anvio7_1/Pangenomic_Results/Coryne-GENOMES.db -p /Users/klemonlab/Documents/CorPGA_rerun_aqr/analysis_Anvio7_1/Pangenomic_Results/Cpse/PAN_Cpse-PAN.db -C CorevsAccessory
```

Find GC Content for *C. accolens*:

```{bash, eval=FALSE}
# conda activate Anvio
anvi-summarize -g /Users/klemonlab/Documents/CorPGA_rerun_aqr/analysis_Anvio7_1/Pangenomic_Results/Coryne-GENOMES.db -p /Users/klemonlab/Documents/CorPGA_rerun_aqr/analysis_Anvio7_1/Pangenomic_Results/Cacc/PAN_Cacc-PAN.db -C CorevsAccessory
```

Find GC Content for *C. tuberculostearicum*:

```{bash, eval=FALSE}
# conda activate Anvio
anvi-summarize -g /Users/klemonlab/Documents/CorPGA_rerun_aqr/analysis_Anvio7_1/Pangenomic_Results/Coryne-GENOMES.db -p /Users/klemonlab/Documents/CorPGA_rerun_aqr/analysis_Anvio7_1/Pangenomic_Results/Ctub/PAN_Ctub-PAN.db -C CorevsAccessory
```

# anvi'o KEGG Analysis Pipeline

```{bash, eval = FALSE}
cd ./Documents/CorPGA_rerun_aqr/analysis_Anvio7_1
conda activate anvio-7.1.2
anvi-estimate-metabolism -e genome_list_Coryne_No_Mac.txt -O Coryne_No_Mac --matrix-format

anvi-matrix-to-newick Coryne_No_Mac-completeness-MATRIX.txt

# dry run to get the profile db:
anvi-interactive -d Coryne_No_Mac-completeness-MATRIX.txt -p Coryne_No_Mac_metabolism_PROFILE.db --manual-mode --dry-run

# run for reals:
anvi-interactive --manual-mode -d Coryne_No_Mac-completeness-MATRIX.txt -t Coryne_No_Mac-completeness-MATRIX.txt.newick -p Coryne_No_Mac_metabolism_PROFILE.db --title "Coryne_No_Mac Metabolism Heatmap"

anvi-export-state -s default -p Coryne_No_Mac_metabolism_PROFILE.db -o Coryne_No_Mac_metabolism_state.json


export ANVIO_MODULES_DB=`python -c "import anvio; import os; print(os.path.join(os.path.dirname(anvio.__file__), 'data/misc/KEGG/MODULES.db'))"`
echo -e > dpig+Coryne-No_Mac_modules_info.txt
sqlite3 $ANVIO_MODULES_DB "select module, data_value from kegg_modules where data_name='CLASS'" | sed 's/; /|/g' | tr '|' '\t' >> module_class.txt
sqlite3 $ANVIO_MODULES_DB "select module,data_value from kegg_modules where data_name='NAME'" | tr '|' '\t' > module_names.txt
paste module_class.txt <(cut -f 2 module_names.txt ) >> Coryne_No_Mac_modules_info.txt
rm module_names.txt module_class.txt


anvi-import-misc-data Coryne_No_Mac_modules_info.txt -p Coryne_No_Mac_metabolism_PROFILE.db -t items

# To painlessly make your interactive visualizations look like the ones in the screenshots, you can import the following state file into your profile database:
anvi-import-state -p Coryne_No_Mac_metabolism_PROFILE.db -n default -s Coryne_No_Mac_metabolism_state.json

# And then we can run the visualization command yet again to see the heatmap with labels.
anvi-interactive --manual-mode -d Coryne_No_Mac-completeness-MATRIX.txt -t Coryne_No_Mac-completeness-MATRIX.txt.newick -p Coryne_No_Mac_metabolism_PROFILE.db --title "Coryne_No_Mac Metabolism Heatmap"

anvi-estimate-metabolism -e genome_list_Coryne_No_Mac.txt -O Coryne_No_Mac_metabolism --kegg-output-modes modules,kofam_hits

# need to usse compute-metabolic-enrichment in anvio 7.1
anvi-compute-metabolic-enrichment -M Coryne_No_Mac_metabolism_modules.txt -G Coryne_No_Mac_species_groups.txt -o Coryne_No_Mac_enriched_modules.txt

anvi-compute-metabolic-enrichment -M Coryne_No_Mac_metabolism_modules.txt -G Coryne_No_Mac_collection_groups.txt -o Coryne_No_Mac_enriched_coll_modules.txt
```


# corpga vs dpi
```{bash, eval = FALSE}
cd ./Documents/aqr_metabol_comp
conda activate anvio-7.1.2
anvi-estimate-metabolism -e genome_list_Coryne_Dpi.txt -O Coryne_Dpi --matrix-format

anvi-matrix-to-newick Coryne_Dpi-completeness-MATRIX.txt

# dry run to get the profile db:
anvi-interactive -d Coryne_Dpi-completeness-MATRIX.txt -p Coryne_Dpi_metabolism_PROFILE.db --manual-mode --dry-run

# run for reals:
anvi-interactive --manual-mode -d Coryne_Dpi-completeness-MATRIX.txt -t Coryne_Dpi-completeness-MATRIX.txt.newick -p Coryne_Dpi_metabolism_PROFILE.db --title "Coryne_Dpi Metabolism Heatmap"

anvi-export-state -s default -p Coryne_Dpi_metabolism_PROFILE.db -o Coryne_Dpi_metabolism_state.json


export ANVIO_MODULES_DB=`python -c "import anvio; import os; print(os.path.join(os.path.dirname(anvio.__file__), 'data/misc/KEGG/MODULES.db'))"`
echo -e > dpig+Coryne-No_Mac_modules_info.txt
sqlite3 $ANVIO_MODULES_DB "select module, data_value from kegg_modules where data_name='CLASS'" | sed 's/; /|/g' | tr '|' '\t' >> module_class.txt
sqlite3 $ANVIO_MODULES_DB "select module,data_value from kegg_modules where data_name='NAME'" | tr '|' '\t' > module_names.txt
paste module_class.txt <(cut -f 2 module_names.txt ) >> Coryne_Dpi_modules_info.txt
rm module_names.txt module_class.txt


anvi-import-misc-data Coryne_Dpi_modules_info.txt -p Coryne_Dpi_metabolism_PROFILE.db -t items

# To painlessly make your interactive visualizations look like the ones in the screenshots, you can import the following state file into your profile database:
anvi-import-state -p Coryne_Dpi_metabolism_PROFILE.db -n default -s Coryne_Dpi_metabolism_state.json

# And then we can run the visualization command yet again to see the heatmap with labels.
anvi-interactive --manual-mode -d Coryne_Dpi-completeness-MATRIX.txt -t Coryne_Dpi-completeness-MATRIX.txt.newick -p Coryne_Dpi_metabolism_PROFILE.db --title "Coryne_Dpi Metabolism Heatmap"

anvi-estimate-metabolism -e genome_list_Coryne_Dpi.txt -O Coryne_Dpi_metabolism --kegg-output-modes modules,kofam_hits

# need to usse compute-metabolic-enrichment in anvio 7.1
anvi-compute-metabolic-enrichment -M Coryne_Dpi_metabolism_modules.txt -G Coryne_Dpi_species_groups.txt -o Coryne_Dpi_enriched_modules.txt
```


# within species geographic comparison
## Cacc
```{bash, eval = FALSE}
cd ./Documents/aqr_metabol_comp
conda activate anvio-7.1.2
anvi-estimate-metabolism -e genome_list_Cacc.txt -O Cacc --matrix-format

anvi-matrix-to-newick Cacc-completeness-MATRIX.txt

# dry run to get the profile db:
anvi-interactive -d Cacc-completeness-MATRIX.txt -p Cacc_metabolism_PROFILE.db --manual-mode --dry-run

# run for reals:
anvi-interactive --manual-mode -d Cacc-completeness-MATRIX.txt -t Cacc-completeness-MATRIX.txt.newick -p Cacc_metabolism_PROFILE.db --title "Cacc Metabolism Heatmap"

anvi-export-state -s default -p Cacc_metabolism_PROFILE.db -o Cacc_metabolism_state.json

export ANVIO_MODULES_DB=`python -c "import anvio; import os; print(os.path.join(os.path.dirname(anvio.__file__), 'data/misc/KEGG/MODULES.db'))"`
echo -e > Cacc_modules_info.txt
sqlite3 $ANVIO_MODULES_DB "select module, data_value from kegg_modules where data_name='CLASS'" | sed 's/; /|/g' | tr '|' '\t' >> module_class.txt
sqlite3 $ANVIO_MODULES_DB "select module,data_value from kegg_modules where data_name='NAME'" | tr '|' '\t' > module_names.txt
paste module_class.txt <(cut -f 2 module_names.txt ) >> Cacc_modules_info.txt
rm module_names.txt module_class.txt

anvi-import-misc-data Cacc_modules_info.txt -p Cacc_metabolism_PROFILE.db -t items

# To painlessly make your interactive visualizations look like the ones in the screenshots, you can import the following state file into your profile database:
anvi-import-state -p Cacc_metabolism_PROFILE.db -n default -s Cacc_metabolism_state.json

# And then we can run the visualization command yet again to see the heatmap with labels.
anvi-interactive --manual-mode -d Cacc-completeness-MATRIX.txt -t Cacc-completeness-MATRIX.txt.newick -p Cacc_metabolism_PROFILE.db --title "Cacc Metabolism Heatmap"

anvi-estimate-metabolism -e genome_list_Cacc.txt -O Cacc_metabolism --kegg-output-modes modules,kofam_hits

# need to usse compute-metabolic-enrichment in anvio 7.1
anvi-compute-metabolic-enrichment -M Cacc_metabolism_modules.txt -G Cacc_geo_groups.txt -o Cacc_enriched_modules.txt
```
## Cprop
```{bash, eval = FALSE}
cd ./Documents/aqr_metabol_comp
conda activate anvio-7.1.2
anvi-estimate-metabolism -e genome_list_Cprop.txt -O Cprop --matrix-format

anvi-matrix-to-newick Cprop-completeness-MATRIX.txt

# dry run to get the profile db:
anvi-interactive -d Cprop-completeness-MATRIX.txt -p Cprop_metabolism_PROFILE.db --manual-mode --dry-run

# run for reals:
anvi-interactive --manual-mode -d Cprop-completeness-MATRIX.txt -t Cprop-completeness-MATRIX.txt.newick -p Cprop_metabolism_PROFILE.db --title "Cprop Metabolism Heatmap"

anvi-export-state -s default -p Cprop_metabolism_PROFILE.db -o Cprop_metabolism_state.json

export ANVIO_MODULES_DB=`python -c "import anvio; import os; print(os.path.join(os.path.dirname(anvio.__file__), 'data/misc/KEGG/MODULES.db'))"`
echo -e > Cprop_modules_info.txt
sqlite3 $ANVIO_MODULES_DB "select module, data_value from kegg_modules where data_name='CLASS'" | sed 's/; /|/g' | tr '|' '\t' >> module_class.txt
sqlite3 $ANVIO_MODULES_DB "select module,data_value from kegg_modules where data_name='NAME'" | tr '|' '\t' > module_names.txt
paste module_class.txt <(cut -f 2 module_names.txt ) >> Cprop_modules_info.txt
rm module_names.txt module_class.txt

anvi-import-misc-data Cprop_modules_info.txt -p Cprop_metabolism_PROFILE.db -t items

# To painlessly make your interactive visualizations look like the ones in the screenshots, you can import the following state file into your profile database:
anvi-import-state -p Cprop_metabolism_PROFILE.db -n default -s Cprop_metabolism_state.json

# And then we can run the visualization command yet again to see the heatmap with labels.
anvi-interactive --manual-mode -d Cprop-completeness-MATRIX.txt -t Cprop-completeness-MATRIX.txt.newick -p Cprop_metabolism_PROFILE.db --title "Cprop Metabolism Heatmap"

anvi-estimate-metabolism -e genome_list_Cprop.txt -O Cprop_metabolism --kegg-output-modes modules,kofam_hits

# need to usse compute-metabolic-enrichment in anvio 7.1
anvi-compute-metabolic-enrichment -M Cprop_metabolism_modules.txt -G Cprop_geo_groups.txt -o Cprop_enriched_modules.txt
```
## Cpsd
```{bash, eval = FALSE}
cd ./Documents/aqr_metabol_comp
conda activate anvio-7.1.2
anvi-estimate-metabolism -e genome_list_Cpsd.txt -O Cpsd --matrix-format
anvi-matrix-to-newick Cpsd-completeness-MATRIX.txt

# dry run to get the profile db:
anvi-interactive -d Cpsd-completeness-MATRIX.txt -p Cpsd_metabolism_PROFILE.db --manual-mode --dry-run

# run for reals:
anvi-interactive --manual-mode -d Cpsd-completeness-MATRIX.txt -t Cpsd-completeness-MATRIX.txt.newick -p Cpsd_metabolism_PROFILE.db --title "Cpsd Metabolism Heatmap"

anvi-export-state -s default -p Cpsd_metabolism_PROFILE.db -o Cpsd_metabolism_state.json

export ANVIO_MODULES_DB=`python -c "import anvio; import os; print(os.path.join(os.path.dirname(anvio.__file__), 'data/misc/KEGG/MODULES.db'))"`
echo -e > Cpsd_modules_info.txt
sqlite3 $ANVIO_MODULES_DB "select module, data_value from kegg_modules where data_name='CLASS'" | sed 's/; /|/g' | tr '|' '\t' >> module_class.txt
sqlite3 $ANVIO_MODULES_DB "select module,data_value from kegg_modules where data_name='NAME'" | tr '|' '\t' > module_names.txt
paste module_class.txt <(cut -f 2 module_names.txt ) >> Cpsd_modules_info.txt
rm module_names.txt module_class.txt

anvi-import-misc-data Cpsd_modules_info.txt -p Cpsd_metabolism_PROFILE.db -t items

# To painlessly make your interactive visualizations look like the ones in the screenshots, you can import the following state file into your profile database:
anvi-import-state -p Cpsd_metabolism_PROFILE.db -n default -s Cpsd_metabolism_state.json

# And then we can run the visualization command yet again to see the heatmap with labels.
anvi-interactive --manual-mode -d Cpsd-completeness-MATRIX.txt -t Cpsd-completeness-MATRIX.txt.newick -p Cpsd_metabolism_PROFILE.db --title "Cpsd Metabolism Heatmap"

anvi-estimate-metabolism -e genome_list_Cpsd.txt -O Cpsd_metabolism --kegg-output-modes modules,kofam_hits

# need to usse compute-metabolic-enrichment in anvio 7.1
anvi-compute-metabolic-enrichment -M Cpsd_metabolism_modules.txt -G Cpsd_geo_groups.txt -o Cpsd_enriched_modules.txt
```
## Ctub
```{bash, eval = FALSE}
cd ./Documents/aqr_metabol_comp
conda activate anvio-7.1.2
anvi-estimate-metabolism -e genome_list_Ctub.txt -O Ctub --matrix-format
anvi-matrix-to-newick Ctub-completeness-MATRIX.txt

# dry run to get the profile db:
anvi-interactive -d Ctub-completeness-MATRIX.txt -p Ctub_metabolism_PROFILE.db --manual-mode --dry-run
anvi-interactive --manual-mode -d Ctub-completeness-MATRIX.txt -t Ctub-completeness-MATRIX.txt.newick -p Ctub_metabolism_PROFILE.db --title "Ctub Metabolism Heatmap"

anvi-export-state -s default -p Ctub_metabolism_PROFILE.db -o Ctub_metabolism_state.json

export ANVIO_MODULES_DB=`python -c "import anvio; import os; print(os.path.join(os.path.dirname(anvio.__file__), 'data/misc/KEGG/MODULES.db'))"`
echo -e > Ctub_modules_info.txt
sqlite3 $ANVIO_MODULES_DB "select module, data_value from kegg_modules where data_name='CLASS'" | sed 's/; /|/g' | tr '|' '\t' >> module_class.txt
sqlite3 $ANVIO_MODULES_DB "select module,data_value from kegg_modules where data_name='NAME'" | tr '|' '\t' > module_names.txt
paste module_class.txt <(cut -f 2 module_names.txt ) >> Ctub_modules_info.txt
rm module_names.txt module_class.txt

anvi-import-misc-data Ctub_modules_info.txt -p Ctub_metabolism_PROFILE.db -t items

# To painlessly make your interactive visualizations look like the ones in the screenshots, you can import the following state file into your profile database:
anvi-import-state -p Ctub_metabolism_PROFILE.db -n default -s Ctub_metabolism_state.json

# And then we can run the visualization command yet again to see the heatmap with labels.
anvi-interactive --manual-mode -d Ctub-completeness-MATRIX.txt -t Ctub-completeness-MATRIX.txt.newick -p Ctub_metabolism_PROFILE.db --title "Ctub Metabolism Heatmap"

anvi-estimate-metabolism -e genome_list_Ctub.txt -O Ctub_metabolism --kegg-output-modes modules,kofam_hits

# need to usse compute-metabolic-enrichment in anvio 7.1
anvi-compute-metabolic-enrichment -M Ctub_metabolism_modules.txt -G Ctub_geo_groups.txt -o Ctub_enriched_modules.txt
```
# Make the distribution plot:
```{R, eval = FALSE}
# read in libraries
library(readxl)
library(ggplot2)
 
# read in module files of each species
cacc <- read_excel("/Users/klemonlab/Documents/aqr_metabol_comp/Cacc_metabolism_modules.xlsx")
cpro <- read_excel("/Users/klemonlab/Documents/aqr_metabol_comp/Cprop_metabolism_modules.xlsx")
cpsd <- read_excel("/Users/klemonlab/Documents/aqr_metabol_comp/Cpsd_metabolism_modules.xlsx")
ctub <- read_excel("/Users/klemonlab/Documents/aqr_metabol_comp/Ctub_metabolism_modules.xlsx")
ckef <- read_excel("/Users/klemonlab/Documents/aqr_metabol_comp/Ckef_metabolism_modules.xlsx")
dpig <- read_excel("/Users/klemonlab/Documents/aqr_metabol_comp/Pigrum_metabolism_modules.xlsx")

# remove warning column
cacc_2 <- subset(cacc, select = -c(db_name, warnings))
cpro_2 <- subset(cpro, select = -c(db_name, warnings))
cpsd_2 <- subset(cpsd, select = -c(db_name, warnings))
ctub_2 <- subset(ctub, select = -c(db_name, warnings))
ckef_2 <- subset(ckef, select = -c(db_name, warnings))
dpig_2 <- subset(dpig, select = -c(db_name))

# filter out to complete module completion > 0.75
cacc_filt <- subset(cacc_2, module_completeness >= 0.75)
cpro_filt <- subset(cpro_2, module_completeness >= 0.75)
cpsd_filt <- subset(cpsd_2, module_completeness >= 0.75)
ctub_filt <- subset(ctub_2, module_completeness >= 0.75)
ckef_filt <- subset(ckef_2, module_completeness >= 0.75)
dpig_filt <- subset(dpig_2, module_completeness >= 0.75)

# add factor for speices
cacc_filt$species <- "C. accolens"
cpro_filt$species <- "C. propinquum"
cpsd_filt$species <- "C. pseudodiphtheriticum"
ctub_filt$species <- "C. tuberculostearicum"
ckef_filt$species <- "C. kefirresidentii"
dpig_filt$species <- "D. pigrum"


# combine into 1 data array
big_boi <- rbind(cacc_filt, cpro_filt, cpsd_filt, ctub_filt, dpig_filt)
big_boi_kef <- rbind(cacc_filt, cpro_filt, cpsd_filt, ctub_filt, dpig_filt, ckef_filt)
```

# make bar graph of distribution: x axis is species y axis is # categorical make up of compelte modules
```{R, eval = FALSE}
# filter out repeated modules here
cacc_filt2 <- unique(subset(cacc_filt, select = c(kegg_module, module_name, module_class, module_category, module_subcategory, species)))
cpro_filt2 <- unique(subset(cpro_filt, select = c(kegg_module, module_name, module_class, module_category, module_subcategory, species)))
cpsd_filt2 <- unique(subset(cpsd_filt, select = c(kegg_module, module_name, module_class, module_category, module_subcategory, species)))
ctub_filt2 <- unique(subset(ctub_filt, select = c(kegg_module, module_name, module_class, module_category, module_subcategory, species)))
dpig_filt2 <- unique(subset(dpig_filt, select = c(kegg_module, module_name, module_class, module_category, module_subcategory, species)))

big_boi2 <- rbind(cpro_filt2, cpsd_filt2, cacc_filt2, ctub_filt2, dpig_filt2)
library("writexl")
write_xlsx(big_boi2, "big_boi2.xlsx")

plot3 = ggplot(big_boi2) + geom_bar(aes(fill = module_category, x = factor(species, level = c("C. propinquum", "C. pseudodiphtheriticum", "C. accolens", "C. tuberculostearicum", "D. pigrum"))))+ scale_fill_brewer(palette = "Set2") + theme_bw() + theme(panel.border = element_blank(), panel.grid.major = element_blank(), panel.grid.minor = element_blank(), axis.line = element_line(colour = "black")) + theme(legend.position = "bottom") + theme(aspect.ratio = 1) 
```
<img src="images/Department-of-Molecular-Virology-&amp;-Microbiologyy-Horz-GRAY.png" align="left" width="200" height="90"/>

------------------------------------------------------------------------

------------------------------------------------------------------------

------------------------------------------------------------------------

------------------------------------------------------------------------

<center>

<h3>[REFERENCES]{.underline}</h3>

</center>
