---
title: "Supplemental Methods: Prokka Annotations"
author: 
- Tommy Tran, (KLemonLab) Tommy.Tran@bcm.edu
- Isabel Escapa, (KLemonLab) Isabel.FernandezEscapa@bcm.edu
output:
  rmdformats::robobook:
    use_bookdown: true
    code_folding: show
bibliography: referencesAnnotations.bib
---

```{r klippy, echo=FALSE, include=TRUE}
klippy::klippy(c('r', 'bash'))
```

We used [Prokka v1.14.6](https://github.com/tseemann/prokka) [@Seemann2014] to annotate 107 *Corynebacterium* strain genomes. We annotated the genomes in two different ways for proper compatibility and strain labeling with both GET_HOMOLOGUES and anvi'o.

More information about Prokka can be found here: <https://github.com/tseemann/prokka>

# Prokka annotation for GET_HOMOLOGUES

This step annotates all the .fasta files in the selected input folder (`path_i`) and places all the output annotated files in the output folder (`path_o`). Output files headers get updated with --genus 'Corynebacterium' --species 'sp' and --strain based on the file name. We used default parameters, including gene recognition and translation initiation site identification with Prodigal [@Hyatt2010].

```{bash, eval=FALSE}
mkdir -p "analysis_GH/Prokka_out"
path_i="Strain_genomes"
path_o="analysis_GH/Prokka_out"

for file in $path_i/*.f*; do
    FILENAME=`basename ${file%.*}`
    prokka --prefix $FILENAME --outdir $path_o --genus 'Corynebacterium' --species 'sp' --strain $FILENAME --centre X --compliant --cpus 0 --force $file; 
done
```

# Prokka annotation for anvi'o

More information about importing Prokka annotations into anvi'o can be found here: <https://merenlab.org/2017/05/18/working-with-prokka/#note-for-the-pangenomics-workflow>

## Fasta reformatting

Before the annotation step with Prokka we need to reformat the .fasta files using `anvi-script-reformat-fasta`. This script creates .fasta files with simplified deflines and also by using --seq-type NT prevents downstream errors with "characters that are not any of A, C, T, G, N, a, c, t, g, n."

```{bash, eval=FALSE}
#conda activate anvio-dev

path_i="data/genomes"
path_o="data/Anvio8/Reformatted"
mkdir -p "$path_o"

for file in $path_i/*.f*; do
    FILENAME=`basename ${file%.*}`
    anvi-script-reformat-fasta -o $path_o/$FILENAME.fa --min-len 0 --simplify-names $file --seq-type NT; 
done
```

## Prokka annotation

This step repeats the Prokka annotation using the anvio reformatted .fasta files. Output files headers get updated with --genus --species and --strain based on the info in the genomes list.

```{bash, eval=FALSE}
#conda activate Prokka

csv_file="data/genome_lists/CorPGA_GenomeList_v01a.csv"
path_i="data/Anvio8/Reformatted"
path_o="data/Anvio8/Prokka_out"
mkdir -p "$path_o"

while IFS=',' read -r name genus species; do
    if [[ "$name" != "name" ]]; then  # Skip the header
        prokka --prefix "$name" --outdir "$path_o" --genus "$genus" --species "$species" --strain "$name" --cpus 0 --force "$path_i/$name.fa"
    fi
done < "$csv_file"
```

## Parsing .gff files

This step is to parse Prokka annotated genomes to import both the external Prodigal gene calls and functions independently into anvi'o. The input (`path_i`) is the annotation in GFF3 format and outputs (`path_o`) are tab-delimited text files, one for gene calls (`calls_*.txt`) and one for annotations (`annot_*.txt`). This is done with the script `gff_parser.py` described in this [tutorial](https://merenlab.org/2017/05/18/working-with-prokka/).

```{bash, eval = FALSE}
#conda activate gffutils

path_i="data/Anvio8/Prokka_out"
path_o="data/Anvio8/Parsed_prokka"
mkdir -p "$path_o"

for file in $path_i/*.gff; do
    FILENAME=`basename ${file%.*}`
    python scripts/gff_parser.py $file \
    --gene-calls $path_o/calls_$FILENAME.txt \
    --annotation $path_o/annot_$FILENAME.txt;
done
```

## Generating contigs databases

In this step the reformatted fasta files (`path_i`) and the external gene calls (`calls_*.txt`) from Prokka (`path_e`) get imported to generate anvi’o contigs databases (`path_o`). Initially we got a lot of early stop codon errors. Therefore, we add the –ignore-internal-stop-codons flag.

```{bash, eval = FALSE}
#conda activate anvio-7.1
mkdir -p "data/Anvio8/Contigs_db"
path_i="data/Anvio8/Reformatted"
path_e="data/Anvio8/Parsed_prokka"
path_o="data/Anvio8/Contigs_db"

for file in $path_i/*.fna; do
    FILENAME=`basename ${file%.*}`
    anvi-gen-contigs-database -f $file \
                              -o $path_o/$FILENAME.db \
                              --external-gene-calls $path_e/calls_$FILENAME.txt \
                              --ignore-internal-stop-codons \
                              -n $FILENAME;
done
```

## Importing Prokka functional annotation

Finally, the external functional annotations (`annot_*.txt`) from Prokka (`path_e`) get imported into the Anvi’o contigs databases (`path_i`).

```{bash, eval = FALSE}
#conda activate anvio-7.1
path_i="data/Anvio8/Contigs_db"
path_e="data/Anvio8/Parsed_prokka"

for file in $path_i/*.db; do
    FILENAME=`basename ${file%.*}`
    anvi-import-functions -c $file \
                          -i $path_e/annot_$FILENAME.txt
      
done
```

<img src="images/Department-of-Molecular-Virology-&amp;-Microbiologyy-Horz-GRAY.png" align="left" width="200" height="90"/>

------------------------------------------------------------------------

------------------------------------------------------------------------

------------------------------------------------------------------------

------------------------------------------------------------------------

<center>

<h3>[REFERENCES]{.underline}</h3>

</center>
