# Metabolism Exploration {.unnumbered}

```{r}
#| echo: FALSE
#| message: FALSE
library(tidyverse)
```

## Data input

```{r}
annot_Cgl <- read_delim("data/Anvio8/Parsed_NCBI/annot_GCF_000011325.1.txt", delim = "\t", escape_double = FALSE, trim_ws = TRUE)
annot_Cgl$genome_name <- "Cgl_ATCC_13032"

Summary <- read_delim("data/Anvio8/Pangenomic_Analysis/All4CorRefs/Summaries/PAN_All4CorRefs_gene_clusters_summary.txt.gz", show_col_types = FALSE)
Summary = left_join(Summary,annot_Cgl, by = join_by(genome_name, gene_callers_id))

Modules <- read_delim("data/Anvio8/Metabolic_Analysis/CorPGA_modules.txt", show_col_types = FALSE, col_types = cols(unique_enzymes_hit_counts = col_character(), gene_caller_ids_in_module = col_character()))
KOHits <- read_delim("data/Anvio8/Metabolic_Analysis/CorPGA_hits.txt", show_col_types = FALSE)

Enrich_Mod_4Cor <- read_delim("data/Anvio8/Metabolic_Analysis/Enrichment_Modules/CorPGA_enriched_modules_stepwise_1_4Cor.txt", show_col_types = FALSE)
Enrich_KO_4Cor_Filtered <- read_delim("data/Anvio8/Metabolic_Analysis/Enrichment_KOs/CorPGA_enriched_KOs_4Cor_Filtered.csv", show_col_types = FALSE)

Enrich_Mod_CorDpi <- read_delim("data/Anvio8/Metabolic_Analysis/Enrichment_Modules/CorPGA_enriched_modules_stepwise_1_CorDpi.txt", show_col_types = FALSE)
Enrich_KO_CorDpi_Filtered <- read_delim("data/Anvio8/Metabolic_Analysis/Enrichment_KOs/CorPGA_enriched_KOs_CorDpi_Filtered.csv", show_col_types = FALSE)

Enrich_Mod_Cps <- read_delim("data/Anvio8/Metabolic_Analysis/Enrichment_Modules/CorPGA_enriched_modules_stepwise_1_Cps.txt", show_col_types = FALSE)
Enrich_KO_Cps_Filtered <- read_delim("data/Anvio8/Metabolic_Analysis/Enrichment_KOs/CorPGA_enriched_KOs_Cps_Filtered.csv", show_col_types = FALSE)

Enrich_Mod_References <- read_delim("data/Anvio8/Metabolic_Analysis/Enrichment_Modules/CorPGA_enriched_modules_stepwise_1_References.txt", show_col_types = FALSE)
Enrich_KO_References_Filtered <- read_delim("data/Anvio8/Metabolic_Analysis/Enrichment_KOs/CorPGA_enriched_KOs_References_Filtered.csv", show_col_types = FALSE)
```

Directory for outputs:
```{r}
dir <- "data/Anvio8/Metabolic_Analysis/Pathway_Exploration"
dir.create(file.path(dir), recursive = TRUE, showWarnings=FALSE)
```

Species info:
```{r}
#species_values <- c("Cgl", "Ctu", "Cac", "Cpr", "Cps")
species_values <- c("Cpr", "Cps", "Cac", "Ctu", "Cgl", "Cdi", "Csi", "Ckr", "Cam", "Can", "Sau", "Sep", "Spn", "Dpi")
num_genomes_values <- c("1", "8", "33", "19", "42")
species_num_genomes_df <- data.frame(species = species_values, num_genomes = as.numeric(num_genomes_values))
```

Modules info:
```{r}
M00854 <- c("K00963", "K12447", "K22920", "K00693", "K00750", "K13679", "K16150", "K16153", "K20812", "K00975", "K00703", "K13679", "K16148", "K00700", "K16149")
M00855 <- c("K00688", "K16153", "K01196", "K00705", "K22451", "K02438", "K01200", "K01835", "K15778", "K15779")
M00854_M00855 <- c("K00963", "K12447", "K22920", "K00693", "K00750", "K13679", "K16150", "K16153", "K20812", "K00975", "K00703", "K13679", "K16148", "K00700", "K16149", "K00688", "K16153", "K01196", "K00705", "K22451", "K02438", "K01200", "K01835", "K15778", "K15779")
```

## Exploration of the KEGG Results

### Average completion score for a list of modules for each species

```{r}
# Complete the modules dataset with rows for strains with completely missing modules and change NA to = 0
Modules <- Modules %>%  
  complete(db_name, nesting(module_category, module_subcategory, module)) %>%  
  mutate(stepwise_module_completeness = ifelse(is.na(stepwise_module_completeness), 0, stepwise_module_completeness))
```


```{r, warning=FALSE}
# Function to calculate average module completion for a given Module and Species

average_Module_completion <- function(df, M, S) {
  result <- df %>%
    filter(grepl(S, db_name)) %>%
    filter(module %in% M) %>%
    summarise(mean = round(mean(stepwise_module_completeness),2))

  return(result$mean)
}

# Example use
#average_Module_completion(Modules, "M00026", "Cpr")

# List of ModuleIDs
ModuleList <- c("M00026", "M00022", "M00023", "M00024", "M00025", "M00019", "M00570", "M00432", "M00020", "M00021", "M00018", "M00017", "M00016", "M00525", "M00526", "M00527", "M00028", "M00844", "M00015", "M00530", "M00529", "M00804", "M00973", "M00176", "M00616")

# List of species
species_values <- c("Cpr", "Cps", "Cac", "Ctu", "Cgl", "Cdi", "Csi", "Ckr", "Cam", "Can", "Sau", "Sep", "Spn", "Dpi")

# Generate all combinations of ModuleIDs and species
combinations <- expand.grid(ModuleID = ModuleList, Species = species_values)

# Use purrr::pmap_dbl to apply the function to each combination and store the results in a data frame
result_df <- pmap_dfr(combinations, ~ tibble(ModuleID = ..1, Species = ..2, 
                                             Average_Completion = average_Module_completion(Modules, ..1, ..2)))

# Pivot the result data frame to have ModuleIDs as rows and species as columns
final_df_modules <- result_df %>%
  pivot_wider(names_from = Species, values_from = Average_Completion)
```

Table with average_stepwise_module_completeness for all "Amino acid metabolism" modules. 
```{r}
AA_Modules <- Modules %>%
  filter(module_category=="Amino acid metabolism") %>%
  mutate(species_name = str_extract(db_name, "^[^_]+")) %>%
  group_by(module_subcategory, module, module_name, species_name) %>%
  summarise(average_stepwise_module_completeness = round(mean(stepwise_module_completeness), 2))

AA_Modules$species_name <- factor(AA_Modules$species_name, levels=species_values)
AA_Modules <- AA_Modules %>% arrange(species_name)

AA_ModulesWide <- AA_Modules %>%
  pivot_wider(names_from = species_name, values_from = average_stepwise_module_completeness, values_fill = 0)

```

### Number of genomes for each species a list of KOs is present

```{r, warning=FALSE}
# Function to calculate average module completion for a given Module and Species

number_genomes_KO_is_present <- function(df, KO, species) {
  result <- df %>%
    filter(grepl(species, db_name)) %>%
    filter(enzyme %in% KO) %>%
    distinct(genome_name) 

  return(nrow(result))
}

# Example use
#number_genomes_KO_is_present(KOHits, "K00957", "Cpr")

# List of ModuleIDs
KOList <- c("K00817", "K01693", "K00821", "K00600", "K00262", "K01915", "K00265", "K00266", "K00259", "K14260", "K01953", "K01425", "K01424", "K01956", "K01955", "K01430", "K01429", "K01428", "K03187", "K03188", "K03189", "K03190", "K11959", "K11960", "K11962", "K03320")

KOList <- c("K07090", "K03795", "K00956", "K00957", "K00390", "K00392", "K00528")

# List of species
species_values <- c("Cpr", "Cps", "Cac", "Ctu", "Cgl", "Cdi", "Csi", "Ckr", "Cam", "Can", "Sau", "Sep", "Spn", "Dpi")

# Generate all combinations of ModuleIDs and species
combinations <- expand.grid(KOID = KOList, Species = species_values)

# Use purrr::pmap_dbl to apply the function to each combination and store the results in a data frame
result_df <- pmap_dfr(combinations, ~ tibble(KOID = ..1, Species = ..2, 
                                             Genome_Number = number_genomes_KO_is_present(KOHits, ..1, ..2)))

# Pivot the result data frame to have ModuleIDs as rows and species as columns
final_df_KOs <- result_df %>%
  pivot_wider(names_from = Species, values_from = Genome_Number)
```

```{r}
KOList <- c("K07090", "K03795", "K00956", "K00957", "K00390", "K00392", "K00528")

KOselection <- Summary %>%
  filter(KOfam_ACC %in% KOList) %>%
  arrange(genome_name,gene_callers_id) %>%
  mutate(KO_GC_id=paste(KOfam_ACC, gene_cluster_id, gene_callers_id, sep="_")) %>%
  select(KO_GC_id, genome_name, gene_callers_id)

KOselectionWide <- spread(KOselection, genome_name, gene_callers_id)
KOselectionWide$KO_GC <- sub("_[^_]+$", "", KOselectionWide$KO_GC_id)

KOselectionWide_summary <- KOselectionWide %>%
  select(-KO_GC_id) %>%
  group_by(KO_GC) %>%
  summarise(across(everything(), ~ if (any(!is.na(.))) toString(na.omit(.)) else NA_character_)) %>%
  arrange(desc(Cgl_ATCC_13032)) %>% 
  separate(KO_GC, into = c("KO", "GC"), sep = "_", extra = "merge")

write_csv(KOselectionWide_summary, file.path(dir, "SummaryARS.csv"))
```


### Number of genomes for each species a included in a list of GCs

```{r, warning=FALSE}
# Function to calculate average module completion for a given Module and Species

number_genomes_GC_is_present <- function(df, GC, species) {
  result <- df %>%
    filter(grepl(species, genome_name)) %>%
    filter(gene_cluster_id %in% GC) %>%
    distinct(genome_name) 

  return(nrow(result))
}

# Example use
#number_genomes_GC_is_present(Summary, "GC_00004683", "Cpr")

# List of ModuleIDs
GCList <- c("GC_00000647", "GC_00000128", "GC_00000502","GC_00000644","GC_00000383", "GC_00000103", "GC_00000073", "GC_00000214", "GC_00001022", "GC_00003183", "GC_00005127", "GC_00005070", "GC_00000924", "GC_00000496", "GC_00004638", "GC_00000603", "GC_00000590", "GC_00000872", "GC_00006661", "GC_00009187", "GC_00000510", "GC_00000411", "GC_00000767", "GC_00008014", "GC_00009615", "GC_00006414", "GC_00010430", "GC_00009898", "GC_00011478", "GC_00001796", "GC_00004683", "GC_00001514", "GC_00002481", "GC_00002732", "GC_00007871", "GC_00001774", "GC_00001775", "GC_00001792", "GC_00007530", "GC_00001800", "GC_00007211", "GC_00001799", "GC_00009638", "GC_00008645", "GC_00001781", "GC_00007018", "GC_00006274", "GC_00001805", "GC_00008808", "GC_00007968")

GCList <- c("GC_00001076", "GC_00008677", "GC_00001061", "GC_00001058", "GC_00002998", "GC_00000686", "GC_00000017")


# List of species
species_values <- c("Cpr", "Cps", "Cac", "Ctu", "Cgl", "Cdi", "Csi", "Ckr", "Cam")

# Generate all combinations of ModuleIDs and species
combinations <- expand.grid(GCID = GCList, Species = species_values)

# Use purrr::pmap_dbl to apply the function to each combination and store the results in a data frame
result_df <- pmap_dfr(combinations, ~ tibble(GCID = ..1, Species = ..2, 
                                             GC_Number = number_genomes_GC_is_present(Summary, ..1, ..2)))

# Pivot the result data frame to have ModuleIDs as rows and species as columns
final_df_GCs <- result_df %>%
  pivot_wider(names_from = Species, values_from = GC_Number)
```



### Proportion of genomes in a species a given KO is present in

```{r, warning=FALSE}
is_KO_present_in_all <- function(df, KO, species, num_genomes) {
  result <- df %>%
    filter(grepl(species, db_name)) %>%
    filter(enzyme %in% KO) %>%
    distinct(genome_name) 

  return(nrow(result)/num_genomes)
  #return(result)
}

KO <- c("K00957")

is_KO_present_in_all(KOHits, KO, "Cac", 33)
is_KO_present_in_all(KOHits, KO, "Cpr", 19)
is_KO_present_in_all(KOHits, KO, "Cps", 42)
is_KO_present_in_all(KOHits, KO, "Ctu", 8)
is_KO_present_in_all(KOHits, KO, "Cgl", 1)

is_KO_present_in_all(KOHits, KO, "Dpi", 27)
```

### What is the average number copies a KO has for all species?

It can take a lists of KOs, like the elements of a module. Not very useful since the same gene can have more than one KO assignment. But It can be used to see the values of several KOs at once.

```{r}
# Define a function to calculate KO_copy_average
KO_copy_average <- function(df, KO, species, num_genomes) {
  result <- df %>%
    filter(grepl(species, db_name)) %>%
    filter(enzyme %in% KO)
  return(nrow(result) / num_genomes)
}

# Define a function to calculate KO counts and write to CSV
calculate_and_write_KO_counts <- function(df, dir, species_values, num_genomes_values, KO_values) {
  result_df <- data.frame(KO = KO_values)
  
  for (species in species_values) {
    # Create a column name for the result based on species
    col_name <- species
    
    # Calculate KO_copy_average for each KO value
    result <- sapply(KO_values, function(KO) {
      KO_copy_average(df, KO, species, species_num_genomes_df$num_genomes[species_num_genomes_df$species == species])
    })
    
    # Add the result to the data frame
    result_df[col_name] <- round(result, 2)
  }
  
  # Construct the output file name based on the KO_values list
  output_file <- paste("CorPGA_KO_copies_", deparse(substitute(KO_values)), ".csv", sep = "")
  
  # Write the data frame into a file
  write_csv(result_df, file.path(dir, output_file))
}

# Call the function to calculate KO counts and write to CSVs
calculate_and_write_KO_counts(KOHits, dir, species_values, num_genomes_values, M00854)
calculate_and_write_KO_counts(KOHits, dir, species_values, num_genomes_values, M00855)
```

### What is the number of unique genes that code for all steps in a module for all species?

```{r}
gene_copy_average <- function(df, KO, species, num_genomes) {
  result <- df %>%
    filter(grepl(species, db_name)) %>%
    filter(enzyme %in% KO) %>%
    group_by(gene_caller_id) %>%
    summarize(enzyme_definition = paste(enzyme_definition, collapse = ", "),
              enzyme = paste(enzyme, collapse = ", ")) 
    return(nrow(result) / num_genomes)
}

# Define a function to calculate gene counts and write to CSV
calculate_and_write_gene_counts <- function(df, dir, species_values, num_genomes_values, KO_values) {
  result_df <- data.frame(Module = deparse(substitute(KO_values)))
  
  for (species in species_values) {
    # Create a column name for the result based on species
    col_name <- species
    
    # Calculate KO_copy_average for each KO value
    result <- gene_copy_average(df, KO_values, species, species_num_genomes_df$num_genomes[species_num_genomes_df$species == species])
    
    # Add the result to the data frame
    result_df[col_name] <- round(result, 2)
  }
  
  # Construct the output file name based on the KO_values list
  output_file <- paste("CorPGA_gene_copies_", deparse(substitute(KO_values)), ".csv", sep = "")
  
  # Write the data frame into a file
  write_csv(result_df, file.path(dir, output_file))
}

# Call the function to calculate KO counts and write to CSVs
calculate_and_write_gene_counts(KOHits, dir, species_values, num_genomes_values, M00854)
calculate_and_write_gene_counts(KOHits, dir, species_values, num_genomes_values, M00855)
calculate_and_write_gene_counts(KOHits, dir, species_values, num_genomes_values, M00854_M00855)
```

### What is the number of unique genes that code for all steps in a module for Cgl linked to the original annotation?

```{r}
unique_genes_for_Module <- function(df, KO_values, species, annot, calls) {
  result <- df %>%
    filter(grepl(species, db_name)) %>%
    filter(enzyme %in% KO_values) %>%
    group_by(gene_caller_id) %>%
    summarize(enzyme_definition = paste(enzyme_definition, collapse = ", "),
              enzyme = paste(enzyme, collapse = ", "))
  
  result <- left_join(result, annot, join_by(gene_caller_id == gene_callers_id))
  result <- left_join(result, calls, join_by(gene_caller_id == gene_callers_id))
  result <- result %>% 
    select(-source.x, -source.y, -version)
    
  output_file <- paste(substitute(species), "_unique_genes_", deparse(substitute(KO_values)), ".csv", sep = "")
  
  # Write the data frame into a file
  write_csv(result, file.path(dir, output_file))
}

annot_GCF_000011325.1 <- read_delim("data/Anvio8/Parsed_NCBI/annot_GCF_000011325.1.txt", show_col_types = FALSE)
calls_GCF_000011325.1 <- read_delim("data/Anvio8/Parsed_NCBI/calls_GCF_000011325.1.txt", show_col_types = FALSE)

unique_genes_for_Module(KOHits, M00854, "Cgl_ATCC_13032", annot_GCF_000011325.1, calls_GCF_000011325.1)
unique_genes_for_Module(KOHits, M00855, "Cgl_ATCC_13032", annot_GCF_000011325.1, calls_GCF_000011325.1)
unique_genes_for_Module(KOHits, M00854_M00855, "Cgl_ATCC_13032", annot_GCF_000011325.1, calls_GCF_000011325.1)
```

## Blast Validations

```{bash}
#| eval: FALSE

path_o="data/Anvio8/Metabolic_Analysis/Blast_Validations/Databases"
mkdir -p "$path_o"

path_i="data/Anvio8/Prokka_out/"
./scripts/process_files_forblastDB.sh $path_i .fna $path_o "NovCor_FullSequence"
./scripts/process_files_forblastDB.sh $path_i .ffn $path_o "NovCor_Annotated_nt"
./scripts/process_files_forblastDB.sh $path_i .faa $path_o "NovCor_Annotated_aa"

path_i="data/Anvio8/Parsed_NCBI/"
./scripts/process_files_forblastDB.sh $path_i .fa $path_o "NCBIStrains_FullSequence"
```

The directory "data/Anvio8/Metabolic_Analysis/NovCor_v03/Blast_Validations/Databases" had to be moved out of Box for the makeblastdb command to work

```{bash}
#| eval: FALSE

#cd <local/path/Blast_Validations/Databases>

makeblastdb -in NovCor_FullSequence -title "NovCor_FullSequence" -input_type fasta -parse_seqids -dbtype 'nucl'
makeblastdb -in NovCor_Annotated_nt -title "NovCor_Annotated_nt" -input_type fasta -parse_seqids -dbtype 'nucl'
makeblastdb -in NovCor_Annotated_aa -title "NovCor_Annotated_aa" -input_type fasta -parse_seqids -dbtype 'prot'

makeblastdb -in NCBIStrains_FullSequence -title "NCBIStrains_FullSequence" -input_type fasta -parse_seqids -dbtype 'nucl'
```

```{bash}
#| eval: FALSE

path_o="data/Anvio8/Metabolic_Analysis/Blast_Validations"

tblastn -db $path_o/Databases/NovCor_FullSequence \
        -query $path_o/Cgl_gdhA.faa \
        -out $path_o/Cgl_gdhA_vs_NovCor_FullSequence.csv \
        -outfmt "10 std qcovs delim=,"
        
tblastn -db $path_o/Databases/NovCor_Annotated_nt \
        -query $path_o/Cgl_gdhA.faa \
        -out $path_o/Cgl_gdhA_vs_NovCor_Annotated_nt.csv \
        -outfmt "10 std qcovs delim=,"       

blastp -db $path_o/Databases/NovCor_Annotated_aa \
        -query $path_o/Cgl_gdhA.faa \
        -out $path_o/Cgl_gdhA_vs_NovCor_Annotated_aa.csv \
        -outfmt "10 std qcovs delim=,"  
        
tblastn -db $path_o/Databases/NCBIStrains_FullSequence \
        -query $path_o/Cgl_gdhA.faa \
        -out $path_o/Cgl_gdhA_NCBIStrains_vs_FullSequence.csv \
        -outfmt "10 std qcovs delim=,"
        
        
tblastn -db $path_o/Databases/NovCor_FullSequence \
        -query $path_o/Cgl_urease.faa \
        -out $path_o/Cgl_urease_vs_NovCor_FullSequence.csv \
        -outfmt "10 std qcovs delim=,"
        
tblastn -db $path_o/Databases/NovCor_Annotated_nt \
        -query $path_o/Cgl_urease.faa \
        -out $path_o/Cgl_urease_vs_NovCor_Annotated_nt.csv \
        -outfmt "10 std qcovs delim=,"       

blastp -db $path_o/Databases/NovCor_Annotated_aa \
        -query $path_o/Cgl_urease.faa \
        -out $path_o/Cgl_urease_vs_NovCor_Annotated_aa.csv \
        -outfmt "10 std qcovs delim=,"  
        
tblastn -db $path_o/Databases/NCBIStrains_FullSequence \
        -query $path_o/Cgl_urease.faa \
        -out $path_o/Cgl_urease_NCBIStrains_vs_FullSequence.csv \
        -outfmt "10 std qcovs delim=,"
```

```{r}
Cgl_gdhA_vs_NovCor_FullSequence <- read_csv("data/Anvio8/Metabolic_Analysis/Blast_Validations/Cgl_gdhA_vs_NovCor_FullSequence.csv", show_col_types = FALSE, col_names = c("qaccver","saccver","pident","length","mismatch","gapopen","qstart","qend","sstart","send","evalue","bitscore","qcovs"))
Cgl_gdhA_vs_NovCor_Annotated_nt <- read_csv("data/Anvio8/Metabolic_Analysis/Blast_Validations/Cgl_gdhA_vs_NovCor_Annotated_nt.csv", show_col_types = FALSE, col_names = c("qaccver","saccver","pident","length","mismatch","gapopen","qstart","qend","sstart","send","evalue","bitscore","qcovs"))
Cgl_gdhA_vs_NovCor_Annotated_aa <- read_csv("data/Anvio8/Metabolic_Analysis/Blast_Validations/Cgl_gdhA_vs_NovCor_Annotated_aa.csv", show_col_types = FALSE, col_names = c("qaccver","saccver","pident","length","mismatch","gapopen","qstart","qend","sstart","send","evalue","bitscore","qcovs"))
Cgl_gdhA_vs_NovCor_FullSequence <- read_csv("data/Anvio8/Metabolic_Analysis/Blast_Validations/Cgl_gdhA_vs_NovCor_FullSequence.csv", show_col_types = FALSE, col_names = c("qaccver","saccver","pident","length","mismatch","gapopen","qstart","qend","sstart","send","evalue","bitscore","qcovs"))

Cgl_urease_vs_NovCor_FullSequence <- read_csv("data/Anvio8/Metabolic_Analysis/Blast_Validations/Cgl_urease_vs_NovCor_FullSequence.csv", show_col_types = FALSE, col_names = c("qaccver","saccver","pident","length","mismatch","gapopen","qstart","qend","sstart","send","evalue","bitscore","qcovs"))
Cgl_urease_vs_NovCor_Annotated_nt <- read_csv("data/Anvio8/Metabolic_Analysis/Blast_Validations/Cgl_urease_vs_NovCor_Annotated_nt.csv", show_col_types = FALSE, col_names = c("qaccver","saccver","pident","length","mismatch","gapopen","qstart","qend","sstart","send","evalue","bitscore","qcovs"))
Cgl_urease_vs_NovCor_Annotated_aa <- read_csv("data/Anvio8/Metabolic_Analysis/Blast_Validations/Cgl_urease_vs_NovCor_Annotated_aa.csv", show_col_types = FALSE, col_names = c("qaccver","saccver","pident","length","mismatch","gapopen","qstart","qend","sstart","send","evalue","bitscore","qcovs"))
Cgl_urease_vs_NovCor_FullSequence <- read_csv("data/Anvio8/Metabolic_Analysis/Blast_Validations/Cgl_urease_vs_NovCor_FullSequence.csv", show_col_types = FALSE, col_names = c("qaccver","saccver","pident","length","mismatch","gapopen","qstart","qend","sstart","send","evalue","bitscore","qcovs"))
```
